---
title: "resistanceModelNotesAndy"
output: html_document
---

Notes on insecticide resistance model for liverpool.
andy south 27/1/15

Timelines and deliverables
**Days 1 to 12:** 
Read and assimilate the Curtis paper cited above. Run the LSTM code written by Levick, replicate the results from the Curtis paper, and prepare the code for sensitivity analysis by firstly refactoring into an R package and creating a Github repository. Identify appropriate parameter ranges and any rules that should be applied while sampling (e.g. male exposure to insecticide within homes should not be higher than for the more endophilic females).  Run sensitivity analysis and present the results as classification trees showing what parameter combinations favour which deployment strategy.  

**Deliverables:**  the provision of sensitivity analyses and classification trees that will allow us to publish the model in a peer-reviewed journal.  

**Days 13 to 25.**  This component is to use the full model rather than the sub-model used to replicate the Curtis results. Liaise with IVCC to identify the policy question they consider the most pressing and the likely end users of the simulation package. Perform sensitivity analysis on the full model as above. Attempt to put a user-friendly interface onto the simulation package using the “Shiny R” package.  

**Deliverables:** sensitivity analysis of the complete model and a reasoned consideration (or delivery) of the feasibility of distributing the package with a graphical interface for use by users with no computing experience.  



**27/1/15**  7.5hrs

```
#working version of Beths code
#run it outside of this project because it changes wd
source("C:\\Dropbox\\Ian and Andy\\andy\\malaria\\Beth code\\malaria_code_beth.r")
```

Editing `malaria_code_andy.r`, indenting functions etc.

Much of the script from line 464 is setting up parameters for example runs.

input[a,b] : a=parameter number and b=scenario number

464         set input params
1058-2716   run model
2718        Actions needing full results.list (e.g. curtis plots use multiple scenarios) 

edited bookmarks to navigate around the file in RStudio.

**9/3/15** 4-5 1hr


**10/3/15** 4.15 - 7.15 3hrs

```{r}
listFunctions <- function(filename) {
  temp.env <- new.env()
  sys.source(filename, envir = temp.env)
  functions <- lsf.str(envir=temp.env)
  rm(temp.env)
  return(functions)
}

#listFunctions("C:\\Dropbox\\Ian and Andy\\andy\\malaria\\Beth code\\malaria_code_beth.r")
```

So there are only 11 functions.

allele.freq : function (mat)  
curtis_f1 : function (nrelaxmat, relaxmat, gencol, r1col)  
curtis_f2 : function (combmat, bmat, amat, gencol, r1col, r2col)  
curtis_ld : function (resultsmat, relaxedmat, gencol, ldcol)  
haplotype : function (mat)  
HW : function (P, mat)  
linkage : function (mat)  
make.genotypemat : function (P_1, P_2)  
make.matrix : function (mat, rnames)  
singlealleleFrequency : function (locus, max_gen, results.list, input)  
timetoFifty : function (locus, max_gen, results.list, input


I can create an initial github repo for the files from Beth (&this notes file)
That means I will have version control for my initial restructuring. Later I may want to create a new repo for the reformatted code as a package.

What to call the first repo ? resistance

1. created repo on github
1. created RStudio project from github repo
+ To get ssh push working.
+ RStudio Tools, Shell
+ git remote set-url origin git@github.com:AndySouth/resistance.git

**11/3/15** 9.30-12.30 3hrs 2.15-6.45 7.5hrs

moved my dropbox folder back to C.

Plan
1. move Beths functions out into a single file for each in an R folder
1. edit Beths commments into roxygen format
1. only later think about changing functions and documentation to follow best practice

Changed plan slightly by appending plot onto start of plot functions.

To source files in the R folder
`lapply(dir("R"),function(x) source(paste0("R//",x)))`

**12/3/15** 9.30-14 4.5hrs 3-5 6-7 7.5hrs

The model generates 3 matrices : results, genotype, fitness
1. Results : freq of R allele at each loci in each sex and linkage disequilibrium of R allele in each sex, per generation
2. Genotype : frequencies of each of the ten genotypes, per generation
3. Fitness : fitness scores of each genotype/niche combination (table 4. of Main Document)

When a file is input with multiple scenarios, each of the three matrices is stored in a list, where scenario number gives position of the matrix in the list.

results.list : 
fitness.list : 
genotype.list : 

Can I put these into a single list so that they can be returned from a single function ?

done ~ put input object creation into a function
But there is still a fair bit of code that reads parameter values out of the input matrix and into named variables.

I might be able to restructure these collections of single variables into arrays.
e.g. for the exposure levels of m&f

```{r}
#to create an array a[sex][locus1][locus2]
sex <- c("F","M")
locus1 <- c("0","A","a")
locus2 <- c("0","B","b")
dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','a','b']
#to check that male exposures total 1
sum(a['M',,])
```


```
  ## Exposure levels of males and females to each insecticide niche ##
  # males
  a.m_00 <- input[8,i]
  
  a.m_a0 <- input[9,i]
  a.m_A0 <- input[10,i]
  
  a.m_0b <- input[11,i]
  a.m_0B <- input[12,i]
  
  a.m_ab <- input[13,i]
  a.m_AB <- input[14,i]
  
  a.m_Ab <- input[15,i]
  a.m_aB <- input[16,i]
  
  #a.m <- sum(a.m_00, a.m_a0, a.m_A0, a.m_0b, a.m_0B, a.m_ab, a.m_AB, a.m_Ab, a.m_aB)
  #if ( a.m != 1 ){		 
  #	print( paste("Error in male exposures: must total one: ", a.m) )
  	
  #	}else{
  #		print( paste( "Male exposures total 1: ", a.m ))
  #		}
  		
  
  # females
  a.f_00 <- input[17,i]
  
  a.f_a0 <- input[18,i]
  a.f_A0 <- input[19,i]
  
  a.f_0b <- input[20,i]
  a.f_0B <- input[21,i]
  
  a.f_ab <- input[22,i]
  a.f_AB <- input[23,i]
  
  a.f_Ab <- input[24,i]
  a.f_aB <- input[25,i]
```

I should check that what we are doing fits in with what others are doing. We may be able to capitalise on existing tools.

[popgen on CRAN](http://cran.r-project.org/web/packages/popgen/popgen.pdf)
seems a bit old and doesn't do a huge amount, no vignette.

[gstudio](http://cran.r-project.org/web/packages/gstudio/) looks more promising. e.g. it defines locus objects.
[gstudio documentation](http://dyerlab.github.io/gstudio/)
A locus object is set like this, but I'm not sure if this would help us.
It does have some stuff on drift at end of help.
```
require(gstudio)
loc <- locus(c("C", "A"))
loc
```

pegas provides functions for the analysis of allelic data and of haplotype data from DNA sequences.
It requires and complements two other R-packages: ape and adegenet.
[pegas](http://ape-package.ird.fr/pegas.html)
[pegas data structures](http://ape-package.ird.fr/pegas/DefinitionDataClassesPegas.pdf)
includes a class called loci. : An object of class "loci" is a data frame where rows represent individuals
and columns are loci and optional additional variables.
Again I can't quite see how it would be useful for us.

There is a [HardyWeinberg](http://cran.r-project.org/web/packages/HardyWeinberg/vignettes/HardyWeinberg.pdf) package on CRAN. "The HardyWeinberg package consists of a set of tools for analyzing diallelic
genetic markers, and is particularly focused on the graphical representation of their (dis)equilibrium condition in various ways."

There is a CRAN [genetics task view](http://cran.r-project.org/web/views/Genetics.html).

The [genetics package](http://cran.r-project.org/web/packages/genetics/genetics.pdf) which also has a locus class, seems to be more detailed, about chromosomes etc. and probably not of use to us.

#### Structure of the results matrices

```{r}
max_gen <- 2 #just as example

# Set up results matrix - prints overall freq of R and S allele per locus per sex, LD and overall allele freq (i.e. 1)
results <- matrix ( nrow = max_gen, ncol = 11 )
colnames( results ) <- c( "Gen", "m.R1", "m.R2", "m.LD","f.R1", "f.R2", "f.LD", "M", "F", "dprime", "r2" )

# set up fitness by niche matrix - records fitness scores for each niche for each genotype
fitness <- matrix ( nrow = 10, ncol = 9, c(rep(0,90)))
colnames(fitness) <- c( "-,-", "a,-", "A,-", "b,-", "B,-", "a,b", "A,B", "A,b", "a,B" )
rownames(fitness) <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
						"RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
						"RR1SS2", "RR1RS2", "RR1RR2")

# set up genotype matrix - records frequencies of each of the 9 two locus genotypes each generation
genotype <- matrix( nrow=max_gen, ncol=11 )
colnames(genotype) <- c("gen", "SS1SS2", "SS2RS2", "SS1RR2", 
						"RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
						"RR1SS2", "RR1RS2", "RR1RR2")


```

Somehow it seems I've broken it, I'm getting NAs in results matrix.

The input files are different (but same structure) between Beths version that works and my new version ...
calibration is 100 in my failing version and 1012 in Beths, i think this is because I'm reading in the csv when I shouldn't.

I seem not to get to line 889 in my version
  	# male
  	f.m.SS1SS2 <- genotype.freq[1,]

Then suddenly it seemed to start working when I put a browser in.
It's something tricky around line 711, if it doesn't get to 889 it doesn't fill in the results matrix ??

Aha! It seems to have been this without the comments before the tilde's that stoppped it from working
```
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 		
  ### Loop to run the model from the initial conditions generated above ####
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

**13/3/15** 9.15-12.15 3hrs

A reproducible version of yesterdays bug.

```
x<-1
~
for(i in 1:10) x<-x+1
cat(x)
```
[R bug reporting](https://bugs.r-project.org/bugzilla3/enter_bug.cgi) 

1. can I create a list of the 3 lists for results ?
Yes
Replace this  
results.list <- list()			
fitness.list <- list()
genotype.list <- list()	
with  
listOut <- list( results=list(), fitness=list(), genotype=list() )

Then find & replace  
results.list with listOut$results  
fitness.list with listOut$fitness
genotype.list with listOut$genotype


**16/3/15** mon 18.30-19.30 1hr
cut out lines 92 to 1740 into a runModel() function, later can subdivide it up

**17/3/15** tues to liverpool 8.30-5.30 9hrs

Assuming that resistance to each insecticide was encoded at one locus respectively, and that these can either have a homozygous susceptible (SS), heterozygous (RS) or homozygous resistant (RR) genotypes at each locus, there are then ten possible genotypes (including cis & trans forms of the double heterozygous).

We consider two insecticides (A and B) at three possible levels; absent, low(ab) and high(AB), resulting in 9 niches;

Locus1 relates to resistance for insecticide A, and locus2 for B. Locus2 does not effect fitness under exposure to A.

Males and females can be exposed differently.

Curtis states that if resistance to one insecticide is present at very low levels in the population, then resistance will rise slower if this insecticide is used on its own, rather than in a combination with a more established insecticide. 

This is the important Curtis statement “the use of a mixture where the initial gene frequencies are unequal leads to more rapid increase in the frequency of the rarer of the genes”

A key aspect of Curtis’ prediction was that this relationship would be determined by an increasing association between the presence of the two resistance alleles, measured as high levels of linkage disequilibrium.

The results given above seem to be in contrast with Curtis’ assumption, and suggest that in fact using the insecticides in combination can slow the spread of resistance.

My summary :
Curtis suggests resistance spreads more quickly with 2 insecticides, due to  increasing association between the alleles. Beths work suggests that combination can slow spread of resistance. 

Good figure 1 in manuscript :
The rate of spread of IR depends on whether the resistance mutation is dominant, semi-dominant, or recessive.

**aha! I've only just understood this bit**
**At high concentrations only the RR individuals survive so resistance is recessive; as concentrations decline some RS mosquitoes survive making resistance semi-dominant; at low concentration both RR and RS mosquitoes survive, making resistance dominant.** 
**so main problems of resistance arise when insecticide concentrations decline**

From manuscript about scenarios to run:
**row in params csv**

Insecticide coverage for females: 0.1, 0.2, 0.6, 0.8, 0.9 (n=5) **9 & 15**  
Insecticide coverage for males: as for females, or half that of females (n=2) **18 & 24**  
‘New’ insecticide resistance starting freq: 0.0001, 0.001, 0.01 (n=3) **6**  
‘Old’ insecticide resistance starting freq: 0.05, 0.1, 0.2, 0.5, 0.8 (n=5) **7**  
Resistance heterozygous fitness: 0.5, 0.75. 1    (n=3 for each insecticide) **?28 & 30**  
Sensitive heterozygous fitness:0, 0.2, 0.4 (n=3 for each insecticide) **?27 & 29**  
Resistance allele dominance: 0, 0.2, 0.5, 0.9, 1 for each locus (n=5 for each locus) **32, 35 do to start**  
Fitness costs: 0, 0.05, 0.1 (n=3) **43&44**  
Fitness dominance: 0, 0.5, 1 (n=3) **33&36**  

BUT now we are going to do it by sampling from a uniform distribution before doing decision trees. Then you can progressively add results to a file.

I could start by creating a document with the results at the extremes of the parameter space.

Send Ian & Beth devtools installation instructions and a document.

Approx. 60,000 combinations. Too many???  

would latin cube hypersampling be better?  

The simulation output was the ratio of  the time (in generation) for resistance  to the ‘new’ insecticide  to reach a given allele threshold frequency when the ‘new’ insecticide was deployed as combination, compared to the time taken to reach the same threshold if deployed on its own; a ratio >1 indicates that combination is better. Three threshold frequencies were investigated: 0.1, 0.25, 0.5.

These ratios was then analysed by regression to find out how the input parameters alter the ratio, and hence under what circumstance combination would be best/worst. Classification tress can also be done.

The intention was not to consider particular plausible deployment situations, but was to systematically investigate a wide range of parameter space to determine whether CD was better than single-insecticide deployment at reducing the rate at which the minor resistance form spreads.
 
Show my progress :

1. github repo
1. move functions to their own files and structure documentation
1. move parts of main file into functions
1. combine 3 outputs to a single list so that can be returned from a function
1. nearly able to create a package with structured documentation etc.

Questions for Beth & Ian

1. Curtis Table1 vi that shows resistance spreading faster with mixture.
1. is coverage same as exposure ? yes
1. what subset of parameters would they like me to vary to start
1. locations of sensitivity params in params.csv
1. what outputs are wanted for sensitivity scenarios ? num generations until resistance reaches threshold (0.1, 0.25 & 0.5) Curtis just used 0.5.
 
Talking with Ian & Beth

Sequential use.
 
For some scenarios, we'll need to run with one insecticide and then stop and start other. 
 
When does mixture fail, until one resistance allele reaches critical point, then continue using 2nd insecticide until resistance for that one reaches critical point too. 
 
Could create a version of timetoFifty that accepts threshold param.
 
Ian has created list of defaults in most recent manuscript.
 
The Curtis version is going to be simpler. Because not all param variations are included.

**18/3/15** weds in liverpool 8.30-6 9.5hrs

Ian created `suggested_sensitivity_analysis.doc`
 
I created a variables table in there with Ians parameter ranges.

My poor paraphrasing of what I think Ian said about the sequential scenarios. The starting conditions from after one insecticide has 'failed' are the same as starting from scratch for the other insecticide because the insecticides to do effect selection on the other locus.
 
moving `malaria_code_andy.r` into a function resistanceMaster()
exposed a couple of variable scoping problems.

plus now the graphs look different and I get :
Warning message: In data.matrix(mat) : NAs introduced by coercion

this is from createInputMatrix()
input <- make.matrix(input, input$Input)
 
this might be something to do with the location of "input.parameters.csv" which I need to sort ...
 
devtools commands for setting up a package :
create requires that the directory doesn't exist yet; it will be created. 
setup assumes an existing directory from which it will infer the package name.
 
probably need to set rstudio rg to false because mine is already a rs project.

setup(rstudio=FALSE)
 
Then in RStudio I had to set project options, build tools to package
(so maybe I should have done just done devtools::setup())

but on build :
ERROR: The build directory does not contain a DESCRIPTION
file so cannot be built as a package.
Build directory: C:/rsprojects/resistance

but there is a DESCRIPTION there ???

devtools::load_all(".")
Error: Line starting 'person("Bethany", "L ...' is malformed!
sorted, just needed tab before 2ry authors
 
package does build now
 
 
potential refactoring of code, if i think carefully, using arrays I should be able to replace this :
```
    if(calibration==103){		## no selection calibration
        ## male
        # SS1
        fs.m.SS1SS2 <- f.m.SS1SS2
        fs.m.SS1RS2 <- f.m.SS1RS2
        fs.m.SS1RR2 <- f.m.SS1RR2
        # RS1
        fs.m.RS1SS2 <- f.m.RS1SS2 
        fs.m.RS1RS2_cis <- f.m.RS1RS2_cis
        fs.m.RS1RS2_trans <- f.m.RS1RS2_trans
        fs.m.RS1RR2 <- f.m.RS1RR2
        # RR2 
        fs.m.RR1SS2 <- f.m.RR1SS2
        fs.m.RR1RS2 <- f.m.RR1RS2
        fs.m.RR1RR2 <- f.m.RR1RR2
        
        ## female
        # SS1
        fs.f.SS1SS2 <- f.f.SS1SS2
        fs.f.SS1RS2 <- f.f.SS1RS2
        fs.f.SS1RR2 <- f.f.SS1RR2
        # RS1
        fs.f.RS1SS2 <- f.f.RS1SS2 
        fs.f.RS1RS2_cis <- f.f.RS1RS2_cis
        fs.f.RS1RS2_trans <- f.f.RS1RS2_trans
        fs.f.RS1RR2 <- f.f.RS1RR2
        # RR2 
        fs.f.RR1SS2 <- f.f.RR1SS2
        fs.f.RR1RS2 <- f.f.RR1RS2
        fs.f.RR1RR2 <- f.f.RR1RR2
```

with something like :
`fs <- f`

where arrays are set up as :
```{r}
#to create an array a[sex][locus1][locus2]
namesLoci <- c('SS','RS','RR')
locus1 <- paste0(namesLoci,'1')
locus2 <- paste0(namesLoci,'2')
sex <- c("F","M")

dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','SS1','RR2']
#to check that males total 1
sum(a['M',,])
```

Just need to work out how best to deal with cis & trans.
This would be a less pleasing alternative
```{r}

namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')

#adding cis & trans
namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2cis','RS1RS2trans','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')

sex <- c("F","M")

dimnames1 <- list( sex=sex, namesLoci=namesLoci )

dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','SS1SS2']
#to check that males total 1
sum(a['M',])
str(a)
```

**19/3/15** fri 9.45-14 4.25hrs 3.30-7 3.5hrs 7.75hrs
 
checking whether I can run without the input file. Fixed now.

```{r}
#require(devtools)    
#install_github('AndySouth/resistance') 
library(resistance)
resistanceMaster(params.csv = FALSE)
``` 

This is what the fitness scores that can be saved for each scenario as *two-locus_fitness-scores.csv look like :
```
	-,-	a,-	A,-	-,b	-,B	a,b	A,B	A,b	a,B
SS1SS2	1	0	0	0	0	0	0	0	0
SS1RS2	1	0	0	0	0	0	1.86E-05	0	0
SS1RR2	1	0	0	0	0	0	0.1161	0	0
RS1SS2	1	0	0	0	0	0	0	0	0
RS1RS2	1	0	0	0	0	0	2.13E-05	0	0
RS1RR2	1	0	0	0	0	0	0.132913	0	0
RR1SS2	1	0	0	0	0	0	0	0	0
RR1RS2	1	0	0	0	0	0	3.44E-05	0	0
RR1RR2	1	0	0	0	0	0	0.215	0	0
```

Starting resistSimple() to be able to run single simple Scenarios.


**20/3/15** sat 10.15-2.15 4hrs

[recent popgen hack](https://github.com/NESCent/r-popgen-hackathon/wiki)
Doesn't seem that useful for this project.

Created setInputOneScenario to enable setting inputs for a single scenario
+ sensible defaults
+ later will allow customised scenarios e.g. setInputCurtisFig1
+ will reduce code volume and repetition in createInputMatrix

Good progress :
check out e.g.
`resistSimple(P_1=0.01,P_2=0.3)`


**23/3/15** mon 9.45-11.45 2hrs 5-8 5hrs 

created a modified plot allele freq function to be able to see overlapping lines

Default settings :
```{r}
resistSimple()
```

Reducing dominance coefficient of locus1 from 1 shifts curves to right & separates locus 1 & 2
```{r}
resistSimple(h.RS1_A0=0.17)
```

But then setting dominance coefficient of locus2 to anything less than 0.7 grounds the curves (actually by pushing them further to right).
```{r}
resistSimple(h.RS1_A0=0.17,h.RS2_0B=0.7)
```

Reducing the selection coefficient for locus1 in A (from 1) pushes curves to right, but strangely doesn't separate locus 1 & 2 ??
```{r}
resistSimple(s.RR1_A0=0.17)
```

Put variable names into the suggested sensitivity analysis table.
 
Thinking about creating a shiny app(s) within the resistance package.
[Advice from RStudio](http://rstudio.github.io/shiny/tutorial/#deployment-local)

Put your Shiny application directory under the package’s inst directory, then create and export a function that contains something like this:

`shiny::runApp(system.file('appdir', package='packagename'))`

I might be able to use something like [this](http://shiny.rstudio.com/gallery/plot-plus-three-columns.html) to put inputs in columns.

Or perhaps better to keep it to a single column to start ?

Starting frequency of resistance
Locus1
Locus2

Exposure to insecticide (same for M&F & in Curtis for both insecticides)

Fitness of susceptibles in presence of insecticide.
Insecticide1
Insecticide2

Dominance of resistance
Locus1
Locus2

Selective advantage of resistance
Locus1
Locus2

It could perhaps be 2 columns, Locus1 & Locus2

Or to fit better with widget titles, 2 rows (l1&2) with 5 columns.

**24/3/15** tue 9.45-2.15 4.5hrs

created function to run shiny UI from package, and add it to readme

there was a bug that dominance had no effect, because I was setting `h.RS1_00` rather than `h.RS1_A0 & h.RS1_0B` 

Cool! Seems to be working well now.

Sent message to Ian & Beth to test installing from Github and the shinyUI.

**25/3/15** weds 12.30-13.45 1.25 2.30-4 1.5 2.75

created a time record  

Barbosa paper methods: 
The analysis was performed using R & package lhs. It does not
allow for the specification of each variable distribution
beforehand, so sampling was performed assuming a uniform
distribution. Once the sample was generated, the
uniform sample from a column (variable) could be transformed
to the required distribution (Table 3) by using
quantile functions (using the qtriangle comand in R).

A data set of 3,000 replications was generated, with random
parameter sets. Ten replicates of this procedure were performed as suggested
in [24] to investigate the predictive precision of model
using LHS as the sampling method. This was achieved
by analysing each replicate separately and verifying that
results were consistent across ten replicates.

...

Results included a counter-intuitive outcome that the
inclusion of a synergist could lead to an increase in the
rate of the spread of resistance (i.e. y > 1). Further investigation of this result was pursed by performing a logistic regression with a binary dependent variable (1 if y > 1 and 0 if y < 1), therefore quantifying how changes in the
parameters values affect the odds of getting the unexpected
outcome y > 1.

...

Classification trees are used to predict membership of
cases in the classes of a categorical dependent variable
(1 if y > 1 or 0 if y < 1) from their input parameters
and were implemented using an algorithm that grows a
binary tree [27]. At each internal node in the tree, a test
is applied to the input parameters to identify the binary
distinction which gives the most information about the
class membership.

These are the references that refer to the classification trees :
27. Breiman L: Classification and Regression Trees. Belmont (Calif.): Wadsworth International Group; 1984: 358.
28. Therneau T, Atkinson EJ: An introduction to recursive partitioning using the RPART routines 2011:1–67. [r.789695.n4.nabble.com/attachment/3209029/0/zed.pdf].

seems that [recursive partitioning using rpart](http://r.789695.n4.nabble.com/attachment/3209029/0/zed.pdf) is the R software used.

[rpart package](http://cran.r-project.org/web/packages/rpart/index.html)
Recursive partitioning for classification, regression and survival trees. An implementation of most of the functionality of the 1984 book by Breiman et al.

[rpart intro vignette](http://cran.r-project.org/web/packages/rpart/vignettes/longintro.pdf) updated version of the reference.


First example from the vignette (p10). This works on a binary response yes/no. method = 'class' is for categorical.
```{r}
library(rpart)
#uses stagec, a dataframe
str(stagec)
#creates a factor for response var to improve labelling
progstat <- factor(stagec$pgstat, levels = 0:1, labels = c("No", "Prog"))
#fit the model by adding the columns
cfit <- rpart(progstat ~ age + eet + g2 + grade + gleason + ploidy, data = stagec, method = 'class')
plot(cfit) #plots tree
text(cfit) #labels tree
```

So that seems to suggest that I'll need to output a dataframe with one row per simulation, columns for outputs (generations until resistance reached) and inputs. Will need to run it by Ian exactly what the outputs are.

p25 in vignette has an example with multiple categories.
Seems there are methods that work on continuous data too, which might be our case ...

**30/3/15** mon 10.30 - 11.30 1hr

RSTMH digital methods conference.
KDR insecticide resistance gene.
Turning 'big' data about how resistance spreads into epidemiological models and then into DSS about control and eradication is very complex. 
Malaria genomics: tracking a diverse and evolving parasite population
Professor Dominic Kwiatkowski, University of Oxford and Wellcome Trust Sanger Institute
Paul Ben Allistere, Panoptes web app
https://www.malariagen.net/apps/ag1000g/phase1-AR2/index.html#genomebrowser

**2/4/15** 1hr

emails from Ian & Marlize Coleman about insecticide resistance gaming.

Ian : Marlize Coleman has funding from B&MGF to build computer games of insecticide deployment policies and wants to tie them to the outputs of various IR models.

Marlize : At the moment **we need someone to look at current mathematical models that have any relevance to insecticide resistance management and determine if there is any potential to use them, or just components of the model/s, to create scenarios for a IR game**. We planning to start development of the game in July so we need to get a game scope sorted pretty soon. It will also be important for this person to communicate and discuss these scenarios/simulations with the game development team to ensure that they have a good grasp of the backend mathematical detail.

We want to expand on this version by adding a bit more complexity in the backend and we’d like to create a bit more flexibility e.g. choice of vectors, disease prevalence/incidence, interventions, specific learning outcomes of interest etc.

If you are interested, we can discuss more detail. Probably 3 months full time or **6 months 2-3 days a week**. You might be better placed to tell me how much effort will be required. Perhaps more time in the first few months and less during game development for assistance to developers.

Me to Ian : Will be good to talk to you about potential IR models after I've spoken to Marlize. Do you imagine that a simplified version of Curtis might be useful for her ?

Ian : I definitely think the 2locus Curtis model should be included. I don't think it necessarily needs to be simple. We may restrict the user to certain parameter values then preload the results from all possible permutations. We can discuss later. BW Ian


**13/4/15** mon 9.30 - 13.30 4hrs 2.30-6 3.5hrs 7.5hrs

looking at new serious gaming project ... & can offset these hours on that

publish shinyCurtis1 to shinyapps.io to enable sharing with Marlize
using publish button from RStudio worked well
It gives this warning on installation, but then does work from web.
Warning message:
In FUN(c("shiny", "resistance")[[2L]], ...) :
  Package 'resistance' not available in repository or locally

https://andysouth.shinyapps.io/shinyCurtis1/

Options :
1. run R code from within the game
1. translate model modules into the game language
1. save model outputs as input files that can modify game behaviour

Points to discuss with Marlize :

* I am mostly committed until end of June, but might be able to shuffle. If we know what they want from the design, we could work on the model behind to produce this after the game design has been started.
* might they want to run actual model code or translate it ?
* I don't know how much info there is out there on the effect on resistance of different insecticide applications, but am happy to research in discussion with Ian, and produce working versions.
* I would be a good intermediary between the biologists and the coders, as I'm somewhere in between myself !

Imperial malaria models :
http://www1.imperial.ac.uk/resources/27F31FE6-4CA6-4C97-85BD-58B342BC2CDA/usingthemalariatoolssoftwareforesp.pdf
http://www1.imperial.ac.uk/malariamodelling/toolsdata/tools/
On download I got an error msg saying : "The program can't start because VCOMP120.DLL is missing from your computer. Try reinstalling the program to fix" 

http://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1000324
Reducing Plasmodium falciparum Malaria Transmission in Africa: A Model-Based Evaluation of Intervention Strategies

emailed them to see about the error


**14/4/15** tue 11.45 - 14.15 2.5hrs 3-5.30 2.5 5hrs  

can I create a plotting method that can accept multiple scenarios ?
initially it could just plot all scenarios as additional lines in the existing colours for locus & sex
use `plotallele.freq.andy`
might need to modify to accept more than one matrix of results outputs  
I thought tricky bit might be to set plot bounds based on multiple lines (but actually the y is already hardcoded from 0-1 for allele frequency). Might need to be a bit careful with x because different scenarios can have different numbers of generations.

done : fixed why `resistanceMaster()` currently fails. I suspect due to input scenario calibration combination, it used to work ??

Error in 1:ddt_cutoff : result would be too long a vector
In addition: Warning messages:
1: In data.matrix(mat) : NAs introduced by coercion
2: In min(which((amat[, r1col]) > 0.5)) :
  no non-missing arguments to min; returning Inf

the error comes from `plotcurtis_f2()`
I suspect it's because the input.parameters.csv isn't the one compatible with producing the curtis plot.
`plotcurtis_f2()` needs results matrices from 3 scenarios: 1 hch, 2 ddt, 3 combined. Yet the current csv I have in the package seems to have 7 scenarios
Input,Curtis Fig 2 Comb - calibration,DDT high,DDT high,DDT high,HCH high,HCH high,HCH high  
Why did it work before then ??
Aha! it's because in `createInputMatrix()` this sets up the 3 Curtis scenarios if an input file is not specified :
  `if( !params.csv && calibration == 1012 ){
    input <- matrix( ncol=3, nrow=52 )
    colnames(input) <- c("B (HCH)", "A (DDT)","Combination")`
So maybe I should have no input csv as the default.

setting params.csv=FALSE as the default for resistanceMaster() sorted it.

Back to `plotallele.freq.andy`, I might be able easily to modify it to accept a list of matrices for multiple scenarios, as well as it's current single scenario mode.

this gives me the number of generations per scenario
`sapply(listOut$results,nrow)`
70  70 160
`max_gen <- max( sapply(listOut$results,nrow) )`
 
done : create a plotting method that can accept multiple scenarios 
 
now try to use it to plot a parameter range.
To run multiple scenarios one way would be to call `setInputOneScenario()` repeatedly to create single scenarios & then rbind these together into an input object that I can pass to `runModel()`

```{r}

input <- NULL
vals <- seq(0,1,0.2) #create a range of input values
for(i in vals)
{
 inputOneScenario <- setInputOneScenario( h.RS1_A0=i )  
 input <- cbind(input, inputOneScenario)
}

listOut <- runModel(input)
plotallele.freq.andy(listOut)

```
This is getting close to allowing the user to specify an input parameter and what range of that parameter you want to plot. I could then potentially create a UI allowing the user to do that.
Would it be useful to put the above code into a function ? 
Actually the above code is very close to what I would need for the sensitivity analysis.

I could create a `setInputMultiScenario()` function that accepts a vector of values for a param, and creates an input object with all other params set to their default. It could maybe even accept multiple args, but then would need to return the n*n scenarios. Also this is going away from the way that Ian suggested doing it by smapling from a distribution, maybe I should try that route first ?

In that case do I want to randomly change all params at once ?
Also might be good to get it to output a file at the end of each 100 runs, so that if it does crash during a long run, everything is not lost. These could just be RDA files saved from the listOut object.
Do I want it to output listOut() each time or will I process so that the outputs are smaller ?
(Might be better to keep the raw outputs to save any confusion later.)
 
The sensitivity analysis function may be called a bit like this : 
 
```
#specify ranges for parameters, if a range is not specified the default will be used
#could offer option to setother params at diff fixed values
#this would only accomodate uniform distributions, OK for now
#think about random seeds later
#be careful about some params which must total one
input <- setInputSensiScenarios( n=100, h.RS1_A0=c(0,1) )
```
 
Am I reinventing the wheel here, can anything else help me with this ?
 
**15/4/15** wed 1.30-2.30 1hrs 3.45-6.15 2.5 3.5hrs

setInputSensiScenarios
function to create sensitivity analysis scenarios, probably by creating a large input object.

**16/4/15** thurs 10-14 4hrs 2.30-5 2.5 6.5hrs

To get input params into 1 row per scenario I can transpose, but the result is a matrix and has no names.
But I need to find a way of getting column names set.

```{r}
input <- setInputSensiScenarios( nScenarios=3, P_1=c(0,0.5), P_2=c(0,0.5) )
tinput <- t(input)
str(tinput)
#num [1:3, 1:52] 100 100 100 100 100 100 1 1 1 0 ...
```

putting names into `setInputOneScenario()` fixed this.

Having a problem with sensiAnalysis1()
Error in eval(..1) : ..1 used in an incorrect context, no ... to look in

This may be because setInputSensiScenarios() expects a range for each argument, whereas setInputOneScenario expects a single value for the same named args. (but I don't really think it is this)

Also stackoverflow pointed that using c for a varname (and there is one in resistance) is dangerous because R can expect the function.

replaced c with recomb_rate for "Recombination Rate" to avoid R conflicts with c() 
previous ... error still remained

Aha! i think problem is that I specify defaults in setInputSensiScenarios when I don't need to (and that stops them making it into the ...)

cool yes, getting there.

**22/4/15** jury service 11.30 - 12.30 1 hr

```
out <- sensiAn1( nScenarios = 10, h.RS1_A0=c(0.1,1))
```


**24/4/15** 4-5 1hr

**27/4/15** 9.15-13.15 4hrs 1.45-2.45 5hrs

trying to test rpart in sensiAn1
getting :
cfit <- rpart(res ~ h.RS1_A0 + h.RS2_0B, data = forCT, method = 'class')
Error in cbind(yval2, yprob, nodeprob) : 
  number of rows of matrices must match (see arg 2)

This was because there were some NAs in the results that can from where a resistance frequency of 0.5 was not reached.
so I temp converted them to 999

Now get :
Error in plot.rpart(cfit) : fit is not a tree, just a root

Is this because I haven't used enough samples or because it doesn't work with just 2 predictors ?

Trying with more samples, still get same error :

`inputAndResults <- sensiAn1( 100, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1) )`

more samples and more inputs I do get a tree
`inputAndResults <- sensiAn1(500, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))`


Some of susanas code for finessing the tree :
tree->rpart(outcome ~ h_n+ h_o+h_I+s_o+s_I+φ_o+φ_I+α_o+α_I+α_fn+α_mo+α_mI+α_mn+z+beta_f+β_m ,method='class',data=working_example,control=rpart.control(minsplit=50))

# control=rpart.control(minsplit=50)
# minsplit	the minimum number of observations that must exist in a node in order for a split to be attempted.
# then prune the tree to avoid overfitting the data
pruned_tree<- prune(tree, cp=tree$cptable[which.min(tree$cptable[,"xerror"]),"CP"])


**28/4/15** 10-14.30 4.5hrs 3-6 7.5hrs

reading Ians recent manuscript update.
I think this from p13 is key on what I need to do with the analysis : 

** The simulation output was the ratio of  the time (in generation) for resistance  to the ‘new’ insecticide  to reach a given allele threshold frequency when the ‘new’ insecticide was deployed as combination, compared to the time taken to reach the same threshold if deployed on its own; a ratio >1 indicates that combination is better. Five threshold frequencies were investigated: 0.05, 0.1, 0.2, 0.5, 0.8. **

In order to do that I'd need to repeat each scenario for a single & combined insecticide treatment. Single is just the new rarer insecticide, Combined also has an older insecticide to which initial resistance is more common.

Searching for the Curtis paper. Cited by 181 papers according to Google.
https://www.google.co.uk/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=%22theoretical%20models%20of%20the%20use%20of%20insecticide%20mixtures%22

https://books.google.co.uk/books?id=ooHjBwAAQBAJ&pg=PA177&lpg=PA177&dq=%22theoretical+models+of+the+use+of+insecticide+mixtures%22&source=bl&ots=b3CPTQc5LL&sig=FCB5MCvX1xUjtfgog1z_dq1pH1Y&hl=en&sa=X&ei=bXw_VbriFoWpsAGnvYG4Ag&ved=0CEEQ6AEwBA#v=onepage&q=Curtis&f=false

Pesticide Resistance in Arthropods (1990) edited by Richard Roush, Bruce E. Tabashnik

these 2 chapters look useful :
Roush, R. T., and J. C. Daly. 1990. The role of population genetics in resistance research
and management, pp. 97-152.In: Pesticide resistance in arthropods. R. T. Roush and
B. E. tabashnik (Eds.) Chapman and Hall, New York.

Tabashnik, B. E. 1990. Modeling and evaluation of resistance management tactics, pp.
153-182. In: Pesticide resistance in arthropods. R. T. Roush and B. E. Tabashnik
(Eds.) Chapman and Hall, New York.

page 131 talks about insecticide mixtures. Refers to some modelling by Roush (1989a)

Roush, R. T. 1989. Designing resistance management programs: How can you choose?
Pestic. Sci. 26:423-441.

p467 in Insect Resistance Management: Biology, Economics, and Prediction edited by David W. Onstad (2013)
seems to have some simulations addressing similar issues to us.

https://books.google.co.uk/books?id=6hp384ZH0_kC&pg=PA453&dq=%22Modeling+and+evaluation+of+resistance+management+tactics%22&source=gbs_toc_r&cad=4#v=onepage&q=%22Modeling%20and%20evaluation%20of%20resistance%20management%20tactics%22&f=false


Now have sorted passing user args to rpart in sensiAn1(). It does produce reasonable looking trees just by e.g.

`inputAndResults <- sensiAn1(100, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))`

**29/4/15** 10.30-12.30 2hrs 3.15-5.15 4hrs

e.g. W doesn't have cis/trans
I could refactor
```{r}
        # relaxed selection fitnesses
        ## Males
        W.m.SS1SS2 <- 0.1 
        W.m.SS1RS2 <- 0.1
        W.m.SS1RR2 <- 0.1
        
        W.m.RS1SS2 <- 0.1
        W.m.RS1RS2 <- 0.1  
        W.m.RS1RR2 <- 0.1  
        
        W.m.RR1SS2 <- 0.1 
        W.m.RR1RS2 <- 0.1 
        W.m.RR1RR2 <- 0.1  
        
        ## Female
        W.f.SS1SS2 <- 0.1
        W.f.SS1RS2 <- 0.1
        W.f.SS1RR2 <- 0.1
        
        W.f.RS1SS2 <- 0.1
        W.f.RS1RS2 <- 0.1 
        W.f.RS1RR2 <- 0.1 
        
        W.f.RR1SS2 <- 0.1
        W.f.RR1RS2 <- 0.1 
        W.f.RR1RR2 <- 0.1
# to :
#to create an array a[sex][locus1][locus2]
namesLoci <- c('SS','RS','RR')
locus1 <- paste0(namesLoci,'1')
locus2 <- paste0(namesLoci,'2')
sex <- c("f","m")

dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
W <- array(0,dim=dim1, dimnames=dimnames1)

#to fill al elements as above
W[] <- 0.1
#check length
length(W)
#to access one element
W['m','SS1','RR2']
#to check that males total 1
sum(W['m',,])
```        

working on refactoring variables to arrays, made flexible createArray() function.
good progress
started runModel2() to contain the refactoring

**30/4/15** 9.45-

Need to be careful about different W fitness variables
`W.RR1_00 : [locus][niche]` single locus (Wlocus)
`W.RR1SS2_0b :   [locus1][locus2][niche]` niche (Wniche)
`W.m.SS1SS2 : [sex][locus1][locus2]` sex (Windiv)

Once I have array refactoring working by .
```{r}

    a <- createArray( sex=c('m','f'), niche1=c('0','a','A'), niche2=c('0','b','B') )
    Wniche <- createArray( locus1 = c('SS1','RS1','RR1'), locus2 = c('SS2','RS2','RR2'), niche1=c('0','a','A'), niche2=c('0','b','B') )    
    Windiv <- createArray( sex=c('m','f'), locus1 = c('SS1','RS1','RR1'), locus2 = c('SS2','RS2','RR2') )
```
I should be able to replace 
```
    W.m.SS1SS2 <- (a.m_00 * W.SS1SS2_00) + 
      (a.m_a0 * W.SS1SS2_a0) + (a.m_A0 * W.SS1SS2_A0) + 
      (a.m_0b * W.SS1SS2_0b) + (a.m_0B * W.SS1SS2_0B) + 
      (a.m_ab * W.SS1SS2_ab) + (a.m_AB * W.SS1SS2_AB) + 
      (a.m_Ab * W.SS1SS2_Ab) + (a.m_aB * W.SS1SS2_aB) 
```
with
```{r}
    Windiv['m','SS1','SS2'] <- sum( a['m',,] * Wniche['SS1','SS2',,])
```
To prove that the above works
```{r}
    #set constant
    a['m',,] <- 2 
    #set variable
    Wniche['SS1','SS2',,] <- 1:9
    Wniche['SS1','SS2',,]
    
    #showing that each element gets multiplied by 2
    a['m',,] * Wniche['SS1','SS2',,]
    
    #show result of sum
    sum( a['m',,] * Wniche['SS1','SS2',,])
    
    Windiv['m','SS1','SS2'] <- sum( a['m',,] * Wniche['SS1','SS2',,])
    
```
Gives a result of 90 for just mSS1SS2 which is correct
``` Windiv
, , locus2 = SS2

   locus1
sex SS1 RS1 RR1
  m  90   0   0
  f   0   0   0

, , locus2 = RS2

   locus1
sex SS1 RS1 RR1
  m   0   0   0
  f   0   0   0

, , locus2 = RR2

   locus1
sex SS1 RS1 RR1
  m   0   0   0
  f   0   0   0
```

Or indeed might be able to replace the whole 120 lines of code from 510-630 with :
No these give: 
```
#Error in a[, , ] * Wniche[, , , ] : non-conformable arrays
    Windiv['m',,] <- sum( a['m',,] * Wniche[,,,])
    Windiv['f',,] <- sum( a['f',,] * Wniche[,,,])
    #or even just
    Windiv[,,] <- sum( a[,,] * Wniche[,,,]) 
```
Just needs a little more thought ...
It's because a needs to be applied repeatedly to each genotype.
Maybe worry about that later ..

Want to develop a sequential testing process, so that I can compare runModel() and runModel2().
Maybe I could use testthat for it ?

Just need to do this to start :
`devtools::use_testthat()`

see testRefactoring() which does things like this :

```
input <- setInputOneScenario()
tst <- runModel(input)
tst2 <- runModel2(input)
input <- setInputSensiScenarios(20, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))
tst <- runModel2(input, produce.plots=FALSE)
tst2 <- runModel(input, produce.plots=FALSE)
identical(tst,tst2)
```



 
### TO DO  

1. is there a way I can easily run single & combined deployments for the same scenario ?
* is it `niche_00` and `niche_AB`
* or in `a.m_AB` insecticide exposure male hi1 hi2
* do I also need to change "Starting frequency of resistance" ?
* perhaps I can set up a simple example input to send to Beth ?
* or ask Beth to send me the input.csv used to produce tables 8 & 9 in 'Brief_full-genotype-tables.pdf'
* maybe comparing the files in full-dom_CD.csv & SD in C:\Dropbox\Levick et al IR manuscript will help me.

1. before May 15th move Spark and Glimmer shiny apps
1. I should have included unit testing on my CV

1. go through the paper and check which bits relate to which in the code. Maybe create a table ? Could I create the table from R ? but actually much of the functionality is still in runModel() and needs to be put into functions.

1. Maybe create a model diagram perhaps from diagrammer ?

1. maybe separate out classification trees from sensiAn1

1. in susana's script the 'outcome' is 0/1 if used in our case that would seem to lose a lot of information about the outcomes ? Even timeToFifty would be better for us.

1. write a function to calculate timeToThreshold for all scenarios based on listOut, allow user to specify which locus.

1. maybe add option in setInputSensiScenarios to save input object as a csv or rda.

1. add some defensive checks, e.g. for sensible input variables

1. submit travel claim 

1. create a vignette from first shiny app documenting how the model works

1. finish tidying documentation of reading in input vars in createInputMatrix()

1. the curtis fig2 plot is produced from >1 scenario, do I want to get it into 1 'reactive' scenario
1. edit/refactor runModel carefully
1. check out the part about running sequential insecticide scenarios

1. spend some time developing my understanding of how the model works
1. submit tilde bug report


### suggestion from Beth
could provide a .csv file with the Curtis scenarios set up
and then the need to set the calibration number and params.csv in the R script can be removed altogether

### How best to refactor variables to arrays ?
Refactoring to arrays will reduce code volume, reduce risk of mistyped variable names and allow objects to be passed between functions.
cis/trans makes it slightly tricky, but not all variables have cis/trans
only used for genotype frequencies (f) and in output matrices fitness & genotype 
in the wild in Beths code :
```
fs.m.RS1RS2_cis <- f.m.RS1RS2_cis
fs.m.RS1RS2_trans <- f.m.RS1RS2_trans
```







### variable naming conventions
**SS1, RS1, RR1** homozygous susceptible, heterozygous or homozygous resistant genotypes at each locus.   
With 2 loci there are ten possible genotypes (including cis & trans forms of the double heterozygous).  

** A, B, a, b, -** 2 insecticides at 3 levels; low(ab), high(AB), absent(--) resulting in 9 niches.  

Locus1 relates to resistance for insecticide A, and locus2 for B. Locus2 does not effect exposure to A.  

```
colnames(fitness) <- c( "-,-", "a,-", "A,-", "b,-", "B,-", "a,b", "A,B", "A,b", "a,B" )
rownames(fitness) <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
                        "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                        "RR1SS2", "RR1RS2", "RR1RR2")
```
#### inputs
** a ** exposure to insecticide
** phi ** fitness of one locus (baseline)
** W.SS1_00 & W.SS2_00** fitness of SS in no insecticide
** h ** dominance coefficient
** s ** selection coefficient
** z ** fitness cost of resistance allele in no insecticide
** niche_ ** insecticide niche toggle

#### calculated
** W ** fitnesses by niche and sex
** f ** genotype frequency
** fs ** genotype frequency following selection
** G ** gametes

Different W fitness variables
`W.RR1_00 : [locus][niche]` single locus
`W.RR1SS2_0b :   [locus1][locus2][niche]` niche
`W.m.SS1SS2 : [sex][locus1][locus2]` sex



Question.
What if `a.m_A0 > 0` but `niche_A0 == 0` ?
The ifelse clauses allow niches to be toggled on/off, i.e. a fitness can be given for A0
but if only the niche A,B is toggled on, the fitness scores for A0 and Ab will be set to 0






