---
title: "resistanceModelNotesAndy"
output: html_document
---

Notes on insecticide resistance model for liverpool.
andy south 27/1/15

Timelines and deliverables
**Days 1 to 12:** 
Read and assimilate the Curtis paper cited above. Run the LSTM code written by Levick, replicate the results from the Curtis paper, and prepare the code for sensitivity analysis by firstly refactoring into an R package and creating a Github repository. Identify appropriate parameter ranges and any rules that should be applied while sampling (e.g. male exposure to insecticide within homes should not be higher than for the more endophilic females).  Run sensitivity analysis and present the results as classification trees showing what parameter combinations favour which deployment strategy.  

**Deliverables:**  the provision of sensitivity analyses and classification trees that will allow us to publish the model in a peer-reviewed journal.  

**Days 13 to 25.**  This component is to use the full model rather than the sub-model used to replicate the Curtis results. Liaise with IVCC to identify the policy question they consider the most pressing and the likely end users of the simulation package. Perform sensitivity analysis on the full model as above. Attempt to put a user-friendly interface onto the simulation package using the “Shiny R” package.  

**Deliverables:** sensitivity analysis of the complete model and a reasoned consideration (or delivery) of the feasibility of distributing the package with a graphical interface for use by users with no computing experience.  



**27/1/15**  7.5hrs

```
#working version of Beths code
#run it outside of this project because it changes wd
source("C:\\Dropbox\\Ian and Andy\\andy\\malaria\\Beth code\\malaria_code_beth.r")
```

Editing `malaria_code_andy.r`, indenting functions etc.

Much of the script from line 464 is setting up parameters for example runs.

input[a,b] : a=parameter number and b=scenario number

464         set input params
1058-2716   run model
2718        Actions needing full results.list (e.g. curtis plots use multiple scenarios) 

edited bookmarks to navigate around the file in RStudio.

**9/3/15** 4-5 1hr


**10/3/15** 4.15 - 7.15 3hrs

```{r}
listFunctions <- function(filename) {
  temp.env <- new.env()
  sys.source(filename, envir = temp.env)
  functions <- lsf.str(envir=temp.env)
  rm(temp.env)
  return(functions)
}

#listFunctions("C:\\Dropbox\\Ian and Andy\\andy\\malaria\\Beth code\\malaria_code_beth.r")
```

So there are only 11 functions.

allele.freq : function (mat)  
curtis_f1 : function (nrelaxmat, relaxmat, gencol, r1col)  
curtis_f2 : function (combmat, bmat, amat, gencol, r1col, r2col)  
curtis_ld : function (resultsmat, relaxedmat, gencol, ldcol)  
haplotype : function (mat)  
HW : function (P, mat)  
linkage : function (mat)  
make.genotypemat : function (P_1, P_2)  
make.matrix : function (mat, rnames)  
singlealleleFrequency : function (locus, max_gen, results.list, input)  
timetoFifty : function (locus, max_gen, results.list, input


I can create an initial github repo for the files from Beth (&this notes file)
That means I will have version control for my initial restructuring. Later I may want to create a new repo for the reformatted code as a package.

What to call the first repo ? resistance

1. created repo on github
1. created RStudio project from github repo
+ To get ssh push working.
+ RStudio Tools, Shell
+ git remote set-url origin git@github.com:AndySouth/resistance.git

**11/3/15** 9.30-12.30 3hrs 2.15-6.45 7.5hrs

moved my dropbox folder back to C.

Plan
1. move Beths functions out into a single file for each in an R folder
1. edit Beths commments into roxygen format
1. only later think about changing functions and documentation to follow best practice

Changed plan slightly by appending plot onto start of plot functions.

To source files in the R folder
`lapply(dir("R"),function(x) source(paste0("R//",x)))`

**12/3/15** 9.30-14 4.5hrs 3-5 6-7 7.5hrs

The model generates 3 matrices : results, genotype, fitness
1. Results : freq of R allele at each loci in each sex and linkage disequilibrium of R allele in each sex, per generation
2. Genotype : frequencies of each of the ten genotypes, per generation
3. Fitness : fitness scores of each genotype/niche combination (table 4. of Main Document)

When a file is input with multiple scenarios, each of the three matrices is stored in a list, where scenario number gives position of the matrix in the list.

results.list : 
fitness.list : 
genotype.list : 

Can I put these into a single list so that they can be returned from a single function ?

done ~ put input object creation into a function
But there is still a fair bit of code that reads parameter values out of the input matrix and into named variables.

I might be able to restructure these collections of single variables into arrays.
e.g. for the exposure levels of m&f

```{r}
#to create an array a[sex][locus1][locus2]
sex <- c("F","M")
locus1 <- c("0","A","a")
locus2 <- c("0","B","b")
dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','a','b']
#to check that male exposures total 1
sum(a['M',,])
```


```
  ## Exposure levels of males and females to each insecticide niche ##
  # males
  a.m_00 <- input[8,i]
  
  a.m_a0 <- input[9,i]
  a.m_A0 <- input[10,i]
  
  a.m_0b <- input[11,i]
  a.m_0B <- input[12,i]
  
  a.m_ab <- input[13,i]
  a.m_AB <- input[14,i]
  
  a.m_Ab <- input[15,i]
  a.m_aB <- input[16,i]
  
  #a.m <- sum(a.m_00, a.m_a0, a.m_A0, a.m_0b, a.m_0B, a.m_ab, a.m_AB, a.m_Ab, a.m_aB)
  #if ( a.m != 1 ){		 
  #	print( paste("Error in male exposures: must total one: ", a.m) )
  	
  #	}else{
  #		print( paste( "Male exposures total 1: ", a.m ))
  #		}
  		
  
  # females
  a.f_00 <- input[17,i]
  
  a.f_a0 <- input[18,i]
  a.f_A0 <- input[19,i]
  
  a.f_0b <- input[20,i]
  a.f_0B <- input[21,i]
  
  a.f_ab <- input[22,i]
  a.f_AB <- input[23,i]
  
  a.f_Ab <- input[24,i]
  a.f_aB <- input[25,i]
```

I should check that what we are doing fits in with what others are doing. We may be able to capitalise on existing tools.

[popgen on CRAN](http://cran.r-project.org/web/packages/popgen/popgen.pdf)
seems a bit old and doesn't do a huge amount, no vignette.

[gstudio](http://cran.r-project.org/web/packages/gstudio/) looks more promising. e.g. it defines locus objects.
[gstudio documentation](http://dyerlab.github.io/gstudio/)
A locus object is set like this, but I'm not sure if this would help us.
It does have some stuff on drift at end of help.
```
require(gstudio)
loc <- locus(c("C", "A"))
loc
```

pegas provides functions for the analysis of allelic data and of haplotype data from DNA sequences.
It requires and complements two other R-packages: ape and adegenet.
[pegas](http://ape-package.ird.fr/pegas.html)
[pegas data structures](http://ape-package.ird.fr/pegas/DefinitionDataClassesPegas.pdf)
includes a class called loci. : An object of class "loci" is a data frame where rows represent individuals
and columns are loci and optional additional variables.
Again I can't quite see how it would be useful for us.

There is a [HardyWeinberg](http://cran.r-project.org/web/packages/HardyWeinberg/vignettes/HardyWeinberg.pdf) package on CRAN. "The HardyWeinberg package consists of a set of tools for analyzing diallelic
genetic markers, and is particularly focused on the graphical representation of their (dis)equilibrium condition in various ways."

There is a CRAN [genetics task view](http://cran.r-project.org/web/views/Genetics.html).

The [genetics package](http://cran.r-project.org/web/packages/genetics/genetics.pdf) which also has a locus class, seems to be more detailed, about chromosomes etc. and probably not of use to us.

#### Structure of the results matrices

```{r}
max_gen <- 2 #just as example

# Set up results matrix - prints overall freq of R and S allele per locus per sex, LD and overall allele freq (i.e. 1)
results <- matrix ( nrow = max_gen, ncol = 11 )
colnames( results ) <- c( "Gen", "m.R1", "m.R2", "m.LD","f.R1", "f.R2", "f.LD", "M", "F", "dprime", "r2" )

# set up fitness by niche matrix - records fitness scores for each niche for each genotype
fitness <- matrix ( nrow = 10, ncol = 9, c(rep(0,90)))
colnames(fitness) <- c( "-,-", "a,-", "A,-", "b,-", "B,-", "a,b", "A,B", "A,b", "a,B" )
rownames(fitness) <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
						"RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
						"RR1SS2", "RR1RS2", "RR1RR2")

# set up genotype matrix - records frequencies of each of the 9 two locus genotypes each generation
genotype <- matrix( nrow=max_gen, ncol=11 )
colnames(genotype) <- c("gen", "SS1SS2", "SS2RS2", "SS1RR2", 
						"RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
						"RR1SS2", "RR1RS2", "RR1RR2")


```

Somehow it seems I've broken it, I'm getting NAs in results matrix.

The input files are different (but same structure) between Beths version that works and my new version ...
calibration is 100 in my failing version and 1012 in Beths, i think this is because I'm reading in the csv when I shouldn't.

I seem not to get to line 889 in my version
  	# male
  	f.m.SS1SS2 <- genotype.freq[1,]

Then suddenly it seemed to start working when I put a browser in.
It's something tricky around line 711, if it doesn't get to 889 it doesn't fill in the results matrix ??

Aha! It seems to have been this without the comments before the tilde's that stoppped it from working
```
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 		
  ### Loop to run the model from the initial conditions generated above ####
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

**13/3/15** 9.15-12.15 3hrs

A reproducible version of yesterdays bug.

```
x<-1
~
for(i in 1:10) x<-x+1
cat(x)
```
[R bug reporting](https://bugs.r-project.org/bugzilla3/enter_bug.cgi) 

1. can I create a list of the 3 lists for results ?
Yes
Replace this  
results.list <- list()			
fitness.list <- list()
genotype.list <- list()	
with  
listOut <- list( results=list(), fitness=list(), genotype=list() )

Then find & replace  
results.list with listOut$results  
fitness.list with listOut$fitness
genotype.list with listOut$genotype


**16/3/15** mon 18.30-19.30 1hr
cut out lines 92 to 1740 into a runModel() function, later can subdivide it up

**17/3/15** tues to liverpool 8.30-5.30 9hrs

Assuming that resistance to each insecticide was encoded at one locus respectively, and that these can either have a homozygous susceptible (SS), heterozygous (RS) or homozygous resistant (RR) genotypes at each locus, there are then ten possible genotypes (including cis & trans forms of the double heterozygous).

We consider two insecticides (A and B) at three possible levels; absent, low(ab) and high(AB), resulting in 9 niches;

Locus1 relates to resistance for insecticide A, and locus2 for B. Locus2 does not effect fitness under exposure to A.

Males and females can be exposed differently.

Curtis states that if resistance to one insecticide is present at very low levels in the population, then resistance will rise slower if this insecticide is used on its own, rather than in a combination with a more established insecticide. 

This is the important Curtis statement “the use of a mixture where the initial gene frequencies are unequal leads to more rapid increase in the frequency of the rarer of the genes”

A key aspect of Curtis’ prediction was that this relationship would be determined by an increasing association between the presence of the two resistance alleles, measured as high levels of linkage disequilibrium.

The results given above seem to be in contrast with Curtis’ assumption, and suggest that in fact using the insecticides in combination can slow the spread of resistance.

My summary :
Curtis suggests resistance spreads more quickly with 2 insecticides, due to  increasing association between the alleles. Beths work suggests that combination can slow spread of resistance. 

Good figure 1 in manuscript :
The rate of spread of IR depends on whether the resistance mutation is dominant, semi-dominant, or recessive.

**aha! I've only just understood this bit**
**At high concentrations only the RR individuals survive so resistance is recessive; as concentrations decline some RS mosquitoes survive making resistance semi-dominant; at low concentration both RR and RS mosquitoes survive, making resistance dominant.** 
**so main problems of resistance arise when insecticide concentrations decline**

From manuscript about scenarios to run:
**row in params csv**

Insecticide coverage for females: 0.1, 0.2, 0.6, 0.8, 0.9 (n=5) **9 & 15**  
Insecticide coverage for males: as for females, or half that of females (n=2) **18 & 24**  
‘New’ insecticide resistance starting freq: 0.0001, 0.001, 0.01 (n=3) **6**  
‘Old’ insecticide resistance starting freq: 0.05, 0.1, 0.2, 0.5, 0.8 (n=5) **7**  
Resistance heterozygous fitness: 0.5, 0.75. 1    (n=3 for each insecticide) **?28 & 30**  
Sensitive heterozygous fitness:0, 0.2, 0.4 (n=3 for each insecticide) **?27 & 29**  
Resistance allele dominance: 0, 0.2, 0.5, 0.9, 1 for each locus (n=5 for each locus) **32, 35 do to start**  
Fitness costs: 0, 0.05, 0.1 (n=3) **43&44**  
Fitness dominance: 0, 0.5, 1 (n=3) **33&36**  

BUT now we are going to do it by sampling from a uniform distribution before doing decision trees. Then you can progressively add results to a file.

I could start by creating a document with the results at the extremes of the parameter space.

Send Ian & Beth devtools installation instructions and a document.

Approx. 60,000 combinations. Too many???  

would latin cube hypersampling be better?  

The simulation output was the ratio of  the time (in generation) for resistance  to the ‘new’ insecticide  to reach a given allele threshold frequency when the ‘new’ insecticide was deployed as combination, compared to the time taken to reach the same threshold if deployed on its own; a ratio >1 indicates that combination is better. Three threshold frequencies were investigated: 0.1, 0.25, 0.5.

These ratios was then analysed by regression to find out how the input parameters alter the ratio, and hence under what circumstance combination would be best/worst. Classification tress can also be done.

The intention was not to consider particular plausible deployment situations, but was to systematically investigate a wide range of parameter space to determine whether CD was better than single-insecticide deployment at reducing the rate at which the minor resistance form spreads.
 
Show my progress :

1. github repo
1. move functions to their own files and structure documentation
1. move parts of main file into functions
1. combine 3 outputs to a single list so that can be returned from a function
1. nearly able to create a package with structured documentation etc.

Questions for Beth & Ian

1. Curtis Table1 vi that shows resistance spreading faster with mixture.
1. is coverage same as exposure ? yes
1. what subset of parameters would they like me to vary to start
1. locations of sensitivity params in params.csv
1. what outputs are wanted for sensitivity scenarios ? num generations until resistance reaches threshold (0.1, 0.25 & 0.5) Curtis just used 0.5.
 
Talking with Ian & Beth

Sequential use.
 
For some scenarios, we'll need to run with one insecticide and then stop and start other. 
 
When does mixture fail, until one resistance allele reaches critical point, then continue using 2nd insecticide until resistance for that one reaches critical point too. 
 
Could create a version of timetoFifty that accepts threshold param.
 
Ian has created list of defaults in most recent manuscript.
 
The Curtis version is going to be simpler. Because not all param variations are included.

**18/3/15** weds in liverpool 8.30-6 9.5hrs

Ian created `suggested_sensitivity_analysis.doc`
 
I created a variables table in there with Ians parameter ranges.

My poor paraphrasing of what I think Ian said about the sequential scenarios. The starting conditions from after one insecticide has 'failed' are the same as starting from scratch for the other insecticide because the insecticides to do effect selection on the other locus.
 
moving `malaria_code_andy.r` into a function resistanceMaster()
exposed a couple of variable scoping problems.

plus now the graphs look different and I get :
Warning message: In data.matrix(mat) : NAs introduced by coercion

this is from createInputMatrix()
input <- make.matrix(input, input$Input)
 
this might be something to do with the location of "input.parameters.csv" which I need to sort ...
 
devtools commands for setting up a package :
create requires that the directory doesn't exist yet; it will be created. 
setup assumes an existing directory from which it will infer the package name.
 
probably need to set rstudio rg to false because mine is already a rs project.

setup(rstudio=FALSE)
 
Then in RStudio I had to set project options, build tools to package
(so maybe I should have done just done devtools::setup())

but on build :
ERROR: The build directory does not contain a DESCRIPTION
file so cannot be built as a package.
Build directory: C:/rsprojects/resistance

but there is a DESCRIPTION there ???

devtools::load_all(".")
Error: Line starting 'person("Bethany", "L ...' is malformed!
sorted, just needed tab before 2ry authors
 
package does build now
 
 
potential refactoring of code, if i think carefully, using arrays I should be able to replace this :
```
    if(calibration==103){		## no selection calibration
        ## male
        # SS1
        fs.m.SS1SS2 <- f.m.SS1SS2
        fs.m.SS1RS2 <- f.m.SS1RS2
        fs.m.SS1RR2 <- f.m.SS1RR2
        # RS1
        fs.m.RS1SS2 <- f.m.RS1SS2 
        fs.m.RS1RS2_cis <- f.m.RS1RS2_cis
        fs.m.RS1RS2_trans <- f.m.RS1RS2_trans
        fs.m.RS1RR2 <- f.m.RS1RR2
        # RR2 
        fs.m.RR1SS2 <- f.m.RR1SS2
        fs.m.RR1RS2 <- f.m.RR1RS2
        fs.m.RR1RR2 <- f.m.RR1RR2
        
        ## female
        # SS1
        fs.f.SS1SS2 <- f.f.SS1SS2
        fs.f.SS1RS2 <- f.f.SS1RS2
        fs.f.SS1RR2 <- f.f.SS1RR2
        # RS1
        fs.f.RS1SS2 <- f.f.RS1SS2 
        fs.f.RS1RS2_cis <- f.f.RS1RS2_cis
        fs.f.RS1RS2_trans <- f.f.RS1RS2_trans
        fs.f.RS1RR2 <- f.f.RS1RR2
        # RR2 
        fs.f.RR1SS2 <- f.f.RR1SS2
        fs.f.RR1RS2 <- f.f.RR1RS2
        fs.f.RR1RR2 <- f.f.RR1RR2
```

with something like :
`fs <- f`

where arrays are set up as :
```{r}
#to create an array a[sex][locus1][locus2]
namesLoci <- c('SS','RS','RR')
locus1 <- paste0(namesLoci,'1')
locus2 <- paste0(namesLoci,'2')
sex <- c("F","M")

dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','SS1','RR2']
#to check that males total 1
sum(a['M',,])
```

Just need to work out how best to deal with cis & trans.
This would be a less pleasing alternative
```{r}

namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')

#adding cis & trans
namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2cis','RS1RS2trans','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')

sex <- c("F","M")

dimnames1 <- list( sex=sex, namesLoci=namesLoci )

dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','SS1SS2']
#to check that males total 1
sum(a['M',])
str(a)
```

**19/3/15** fri 9.45-14 4.25hrs 3.30-7 3.5hrs 7.75hrs
 
checking whether I can run without the input file. Fixed now.

```{r}
#require(devtools)    
#install_github('AndySouth/resistance') 
library(resistance)
resistanceMaster(params.csv = FALSE)
``` 

This is what the fitness scores that can be saved for each scenario as *two-locus_fitness-scores.csv look like :
```
	-,-	a,-	A,-	-,b	-,B	a,b	A,B	A,b	a,B
SS1SS2	1	0	0	0	0	0	0	0	0
SS1RS2	1	0	0	0	0	0	1.86E-05	0	0
SS1RR2	1	0	0	0	0	0	0.1161	0	0
RS1SS2	1	0	0	0	0	0	0	0	0
RS1RS2	1	0	0	0	0	0	2.13E-05	0	0
RS1RR2	1	0	0	0	0	0	0.132913	0	0
RR1SS2	1	0	0	0	0	0	0	0	0
RR1RS2	1	0	0	0	0	0	3.44E-05	0	0
RR1RR2	1	0	0	0	0	0	0.215	0	0
```

Starting resistSimple() to be able to run single simple Scenarios.


**20/3/15** sat 10.15-2.15 4hrs

[recent popgen hack](https://github.com/NESCent/r-popgen-hackathon/wiki)
Doesn't seem that useful for this project.

Created setInputOneScenario to enable setting inputs for a single scenario
+ sensible defaults
+ later will allow customised scenarios e.g. setInputCurtisFig1
+ will reduce code volume and repetition in createInputMatrix

Good progress :
check out e.g.
`resistSimple(P_1=0.01,P_2=0.3)`


**23/3/15** mon 9.45-11.45 2hrs 5-8 5hrs 

created a modified plot allele freq function to be able to see overlapping lines

Default settings :
```{r}
resistSimple()
```

Reducing dominance coefficient of locus1 from 1 shifts curves to right & separates locus 1 & 2
```{r}
resistSimple(h.RS1_A0=0.17)
```

But then setting dominance coefficient of locus2 to anything less than 0.7 grounds the curves (actually by pushing them further to right).
```{r}
resistSimple(h.RS1_A0=0.17,h.RS2_0B=0.7)
```

Reducing the selection coefficient for locus1 in A (from 1) pushes curves to right, but strangely doesn't separate locus 1 & 2 ??
```{r}
resistSimple(s.RR1_A0=0.17)
```

Put variable names into the suggested sensitivity analysis table.
 
Thinking about creating a shiny app(s) within the resistance package.
[Advice from RStudio](http://rstudio.github.io/shiny/tutorial/#deployment-local)

Put your Shiny application directory under the package’s inst directory, then create and export a function that contains something like this:

`shiny::runApp(system.file('appdir', package='packagename'))`

I might be able to use something like [this](http://shiny.rstudio.com/gallery/plot-plus-three-columns.html) to put inputs in columns.

Or perhaps better to keep it to a single column to start ?

Starting frequency of resistance
Locus1
Locus2

Exposure to insecticide (same for M&F & in Curtis for both insecticides)

Fitness of susceptibles in presence of insecticide.
Insecticide1
Insecticide2

Dominance of resistance
Locus1
Locus2

Selective advantage of resistance
Locus1
Locus2

It could perhaps be 2 columns, Locus1 & Locus2

Or to fit better with widget titles, 2 rows (l1&2) with 5 columns.

**24/3/15** tue 9.45-2.15 4.5hrs

created function to run shiny UI from package, and add it to readme

there was a bug that dominance had no effect, because I was setting `h.RS1_00` rather than `h.RS1_A0 & h.RS1_0B` 

Cool! Seems to be working well now.

Sent message to Ian & Beth to test installing from Github and the shinyUI.

**25/3/15** weds 12.30-13.45 1.25 2.30-4 1.5 2.75

created a time record  

Barbosa paper methods: 
The analysis was performed using R & package lhs. It does not
allow for the specification of each variable distribution
beforehand, so sampling was performed assuming a uniform
distribution. Once the sample was generated, the
uniform sample from a column (variable) could be transformed
to the required distribution (Table 3) by using
quantile functions (using the qtriangle comand in R).

A data set of 3,000 replications was generated, with random
parameter sets. Ten replicates of this procedure were performed as suggested
in [24] to investigate the predictive precision of model
using LHS as the sampling method. This was achieved
by analysing each replicate separately and verifying that
results were consistent across ten replicates.

...

Results included a counter-intuitive outcome that the
inclusion of a synergist could lead to an increase in the
rate of the spread of resistance (i.e. y > 1). Further investigation of this result was pursed by performing a logistic regression with a binary dependent variable (1 if y > 1 and 0 if y < 1), therefore quantifying how changes in the
parameters values affect the odds of getting the unexpected
outcome y > 1.

...

Classification trees are used to predict membership of
cases in the classes of a categorical dependent variable
(1 if y > 1 or 0 if y < 1) from their input parameters
and were implemented using an algorithm that grows a
binary tree [27]. At each internal node in the tree, a test
is applied to the input parameters to identify the binary
distinction which gives the most information about the
class membership.

These are the references that refer to the classification trees :
27. Breiman L: Classification and Regression Trees. Belmont (Calif.): Wadsworth International Group; 1984: 358.
28. Therneau T, Atkinson EJ: An introduction to recursive partitioning using the RPART routines 2011:1–67. [r.789695.n4.nabble.com/attachment/3209029/0/zed.pdf].

seems that [recursive partitioning using rpart](http://r.789695.n4.nabble.com/attachment/3209029/0/zed.pdf) is the R software used.

[rpart package](http://cran.r-project.org/web/packages/rpart/index.html)
Recursive partitioning for classification, regression and survival trees. An implementation of most of the functionality of the 1984 book by Breiman et al.

[rpart intro vignette](http://cran.r-project.org/web/packages/rpart/vignettes/longintro.pdf) updated version of the reference.


First example from the vignette (p10). This works on a binary response yes/no. method = 'class' is for categorical.
```{r}
library(rpart)
#uses stagec, a dataframe
str(stagec)
#creates a factor for response var to improve labelling
progstat <- factor(stagec$pgstat, levels = 0:1, labels = c("No", "Prog"))
#fit the model by adding the columns
cfit <- rpart(progstat ~ age + eet + g2 + grade + gleason + ploidy, data = stagec, method = 'class')
plot(cfit) #plots tree
text(cfit) #labels tree
```

So that seems to suggest that I'll need to output a dataframe with one row per simulation, columns for outputs (generations until resistance reached) and inputs. Will need to run it by Ian exactly what the outputs are.

p25 in vignette has an example with multiple categories.
Seems there are methods that work on continuous data too, which might be our case ...

**30/3/15** mon 10.30 - 11.30 1hr

RSTMH digital methods conference.
KDR insecticide resistance gene.
Turning 'big' data about how resistance spreads into epidemiological models and then into DSS about control and eradication is very complex. 
Malaria genomics: tracking a diverse and evolving parasite population
Professor Dominic Kwiatkowski, University of Oxford and Wellcome Trust Sanger Institute
Paul Ben Allistere, Panoptes web app
https://www.malariagen.net/apps/ag1000g/phase1-AR2/index.html#genomebrowser

**2/4/15** 1hr

emails from Ian & Marlize Coleman about insecticide resistance gaming.

Ian : Marlize Coleman has funding from B&MGF to build computer games of insecticide deployment policies and wants to tie them to the outputs of various IR models.

Marlize : At the moment **we need someone to look at current mathematical models that have any relevance to insecticide resistance management and determine if there is any potential to use them, or just components of the model/s, to create scenarios for a IR game**. We planning to start development of the game in July so we need to get a game scope sorted pretty soon. It will also be important for this person to communicate and discuss these scenarios/simulations with the game development team to ensure that they have a good grasp of the backend mathematical detail.

We want to expand on this version by adding a bit more complexity in the backend and we’d like to create a bit more flexibility e.g. choice of vectors, disease prevalence/incidence, interventions, specific learning outcomes of interest etc.

If you are interested, we can discuss more detail. Probably 3 months full time or **6 months 2-3 days a week**. You might be better placed to tell me how much effort will be required. Perhaps more time in the first few months and less during game development for assistance to developers.

Me to Ian : Will be good to talk to you about potential IR models after I've spoken to Marlize. Do you imagine that a simplified version of Curtis might be useful for her ?

Ian : I definitely think the 2locus Curtis model should be included. I don't think it necessarily needs to be simple. We may restrict the user to certain parameter values then preload the results from all possible permutations. We can discuss later. BW Ian


**13/4/15** mon 9.30 - 13.30 4hrs 2.30-6 3.5hrs 7.5hrs

looking at new serious gaming project ... & can offset these hours on that

publish shinyCurtis1 to shinyapps.io to enable sharing with Marlize
using publish button from RStudio worked well
It gives this warning on installation, but then does work from web.
Warning message:
In FUN(c("shiny", "resistance")[[2L]], ...) :
  Package 'resistance' not available in repository or locally

https://andysouth.shinyapps.io/shinyCurtis1/

Options :
1. run R code from within the game
1. translate model modules into the game language
1. save model outputs as input files that can modify game behaviour

Points to discuss with Marlize :

* I am mostly committed until end of June, but might be able to shuffle. If we know what they want from the design, we could work on the model behind to produce this after the game design has been started.
* might they want to run actual model code or translate it ?
* I don't know how much info there is out there on the effect on resistance of different insecticide applications, but am happy to research in discussion with Ian, and produce working versions.
* I would be a good intermediary between the biologists and the coders, as I'm somewhere in between myself !

Imperial malaria models :
http://www1.imperial.ac.uk/resources/27F31FE6-4CA6-4C97-85BD-58B342BC2CDA/usingthemalariatoolssoftwareforesp.pdf
http://www1.imperial.ac.uk/malariamodelling/toolsdata/tools/
On download I got an error msg saying : "The program can't start because VCOMP120.DLL is missing from your computer. Try reinstalling the program to fix" 

http://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1000324
Reducing Plasmodium falciparum Malaria Transmission in Africa: A Model-Based Evaluation of Intervention Strategies

emailed them to see about the error


**14/4/15** tue 11.45 - 14.15 2.5hrs 3-5.30 2.5 5hrs  

can I create a plotting method that can accept multiple scenarios ?
initially it could just plot all scenarios as additional lines in the existing colours for locus & sex
use `plotallele.freq.andy`
might need to modify to accept more than one matrix of results outputs  
I thought tricky bit might be to set plot bounds based on multiple lines (but actually the y is already hardcoded from 0-1 for allele frequency). Might need to be a bit careful with x because different scenarios can have different numbers of generations.

done : fixed why `resistanceMaster()` currently fails. I suspect due to input scenario calibration combination, it used to work ??

Error in 1:ddt_cutoff : result would be too long a vector
In addition: Warning messages:
1: In data.matrix(mat) : NAs introduced by coercion
2: In min(which((amat[, r1col]) > 0.5)) :
  no non-missing arguments to min; returning Inf

the error comes from `plotcurtis_f2()`
I suspect it's because the input.parameters.csv isn't the one compatible with producing the curtis plot.
`plotcurtis_f2()` needs results matrices from 3 scenarios: 1 hch, 2 ddt, 3 combined. Yet the current csv I have in the package seems to have 7 scenarios
Input,Curtis Fig 2 Comb - calibration,DDT high,DDT high,DDT high,HCH high,HCH high,HCH high  
Why did it work before then ??
Aha! it's because in `createInputMatrix()` this sets up the 3 Curtis scenarios if an input file is not specified :
  `if( !params.csv && calibration == 1012 ){
    input <- matrix( ncol=3, nrow=52 )
    colnames(input) <- c("B (HCH)", "A (DDT)","Combination")`
So maybe I should have no input csv as the default.

setting params.csv=FALSE as the default for resistanceMaster() sorted it.

Back to `plotallele.freq.andy`, I might be able easily to modify it to accept a list of matrices for multiple scenarios, as well as it's current single scenario mode.

this gives me the number of generations per scenario
`sapply(listOut$results,nrow)`
70  70 160
`max_gen <- max( sapply(listOut$results,nrow) )`
 
done : create a plotting method that can accept multiple scenarios 
 
now try to use it to plot a parameter range.
To run multiple scenarios one way would be to call `setInputOneScenario()` repeatedly to create single scenarios & then rbind these together into an input object that I can pass to `runModel()`

```{r}

input <- NULL
vals <- seq(0,1,0.2) #create a range of input values
for(i in vals)
{
 inputOneScenario <- setInputOneScenario( h.RS1_A0=i )  
 input <- cbind(input, inputOneScenario)
}

listOut <- runModel(input)
plotallele.freq.andy(listOut)

```
This is getting close to allowing the user to specify an input parameter and what range of that parameter you want to plot. I could then potentially create a UI allowing the user to do that.
Would it be useful to put the above code into a function ? 
Actually the above code is very close to what I would need for the sensitivity analysis.

I could create a `setInputMultiScenario()` function that accepts a vector of values for a param, and creates an input object with all other params set to their default. It could maybe even accept multiple args, but then would need to return the n*n scenarios. Also this is going away from the way that Ian suggested doing it by smapling from a distribution, maybe I should try that route first ?

In that case do I want to randomly change all params at once ?
Also might be good to get it to output a file at the end of each 100 runs, so that if it does crash during a long run, everything is not lost. These could just be RDA files saved from the listOut object.
Do I want it to output listOut() each time or will I process so that the outputs are smaller ?
(Might be better to keep the raw outputs to save any confusion later.)
 
The sensitivity analysis function may be called a bit like this : 
 
```
#specify ranges for parameters, if a range is not specified the default will be used
#could offer option to setother params at diff fixed values
#this would only accomodate uniform distributions, OK for now
#think about random seeds later
#be careful about some params which must total one
input <- setInputSensiScenarios( n=100, h.RS1_A0=c(0,1) )
```
 
Am I reinventing the wheel here, can anything else help me with this ?
 
**15/4/15** wed 1.30-2.30 1hrs 3.45-6.15 2.5 3.5hrs

setInputSensiScenarios
function to create sensitivity analysis scenarios, probably by creating a large input object.

**16/4/15** thurs 10-14 4hrs 2.30-5 2.5 6.5hrs

To get input params into 1 row per scenario I can transpose, but the result is a matrix and has no names.
But I need to find a way of getting column names set.

```{r}
input <- setInputSensiScenarios( nScenarios=3, P_1=c(0,0.5), P_2=c(0,0.5) )
tinput <- t(input)
str(tinput)
#num [1:3, 1:52] 100 100 100 100 100 100 1 1 1 0 ...
```

putting names into `setInputOneScenario()` fixed this.

Having a problem with sensiAnalysis1()
Error in eval(..1) : ..1 used in an incorrect context, no ... to look in

This may be because setInputSensiScenarios() expects a range for each argument, whereas setInputOneScenario expects a single value for the same named args. (but I don't really think it is this)

Also stackoverflow pointed that using c for a varname (and there is one in resistance) is dangerous because R can expect the function.

replaced c with recomb_rate for "Recombination Rate" to avoid R conflicts with c() 
previous ... error still remained

Aha! i think problem is that I specify defaults in setInputSensiScenarios when I don't need to (and that stops them making it into the ...)

cool yes, getting there.

**22/4/15** jury service 11.30 - 12.30 1 hr

```
out <- sensiAn1( nScenarios = 10, h.RS1_A0=c(0.1,1))
```


**24/4/15** 4-5 1hr

**27/4/15** 9.15-13.15 4hrs 1.45-2.45 5hrs

trying to test rpart in sensiAn1
getting :
cfit <- rpart(res ~ h.RS1_A0 + h.RS2_0B, data = forCT, method = 'class')
Error in cbind(yval2, yprob, nodeprob) : 
  number of rows of matrices must match (see arg 2)

This was because there were some NAs in the results that can from where a resistance frequency of 0.5 was not reached.
so I temp converted them to 999

Now get :
Error in plot.rpart(cfit) : fit is not a tree, just a root

Is this because I haven't used enough samples or because it doesn't work with just 2 predictors ?

Trying with more samples, still get same error :

`inputAndResults <- sensiAn1( 100, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1) )`

more samples and more inputs I do get a tree
`inputAndResults <- sensiAn1(500, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))`


Some of susanas code for finessing the tree :
tree->rpart(outcome ~ h_n+ h_o+h_I+s_o+s_I+φ_o+φ_I+α_o+α_I+α_fn+α_mo+α_mI+α_mn+z+beta_f+β_m ,method='class',data=working_example,control=rpart.control(minsplit=50))

# control=rpart.control(minsplit=50)
# minsplit	the minimum number of observations that must exist in a node in order for a split to be attempted.
# then prune the tree to avoid overfitting the data
pruned_tree<- prune(tree, cp=tree$cptable[which.min(tree$cptable[,"xerror"]),"CP"])


**28/4/15** 10-14.30 4.5hrs 3-6 7.5hrs

reading Ians recent manuscript update.
I think this from p13 is key on what I need to do with the analysis : 

** The simulation output was the ratio of  the time (in generation) for resistance  to the ‘new’ insecticide  to reach a given allele threshold frequency when the ‘new’ insecticide was deployed as combination, compared to the time taken to reach the same threshold if deployed on its own; a ratio >1 indicates that combination is better. Five threshold frequencies were investigated: 0.05, 0.1, 0.2, 0.5, 0.8. **

In order to do that I'd need to repeat each scenario for a single & combined insecticide treatment. Single is just the new rarer insecticide, Combined also has an older insecticide to which initial resistance is more common.

Searching for the Curtis paper. Cited by 181 papers according to Google.
https://www.google.co.uk/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=%22theoretical%20models%20of%20the%20use%20of%20insecticide%20mixtures%22

https://books.google.co.uk/books?id=ooHjBwAAQBAJ&pg=PA177&lpg=PA177&dq=%22theoretical+models+of+the+use+of+insecticide+mixtures%22&source=bl&ots=b3CPTQc5LL&sig=FCB5MCvX1xUjtfgog1z_dq1pH1Y&hl=en&sa=X&ei=bXw_VbriFoWpsAGnvYG4Ag&ved=0CEEQ6AEwBA#v=onepage&q=Curtis&f=false

Pesticide Resistance in Arthropods (1990) edited by Richard Roush, Bruce E. Tabashnik

these 2 chapters look useful :
Roush, R. T., and J. C. Daly. 1990. The role of population genetics in resistance research
and management, pp. 97-152.In: Pesticide resistance in arthropods. R. T. Roush and
B. E. tabashnik (Eds.) Chapman and Hall, New York.

Tabashnik, B. E. 1990. Modeling and evaluation of resistance management tactics, pp.
153-182. In: Pesticide resistance in arthropods. R. T. Roush and B. E. Tabashnik
(Eds.) Chapman and Hall, New York.

page 131 talks about insecticide mixtures. Refers to some modelling by Roush (1989a)

Roush, R. T. 1989. Designing resistance management programs: How can you choose?
Pestic. Sci. 26:423-441.

p467 in Insect Resistance Management: Biology, Economics, and Prediction edited by David W. Onstad (2013)
seems to have some simulations addressing similar issues to us.

https://books.google.co.uk/books?id=6hp384ZH0_kC&pg=PA453&dq=%22Modeling+and+evaluation+of+resistance+management+tactics%22&source=gbs_toc_r&cad=4#v=onepage&q=%22Modeling%20and%20evaluation%20of%20resistance%20management%20tactics%22&f=false


Now have sorted passing user args to rpart in sensiAn1(). It does produce reasonable looking trees just by e.g.

`inputAndResults <- sensiAn1(100, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))`

**29/4/15** 10.30-12.30 2hrs 3.15-5.15 4hrs

e.g. W doesn't have cis/trans
I could refactor
```{r}
        # relaxed selection fitnesses
        ## Males
        W.m.SS1SS2 <- 0.1 
        W.m.SS1RS2 <- 0.1
        W.m.SS1RR2 <- 0.1
        
        W.m.RS1SS2 <- 0.1
        W.m.RS1RS2 <- 0.1  
        W.m.RS1RR2 <- 0.1  
        
        W.m.RR1SS2 <- 0.1 
        W.m.RR1RS2 <- 0.1 
        W.m.RR1RR2 <- 0.1  
        
        ## Female
        W.f.SS1SS2 <- 0.1
        W.f.SS1RS2 <- 0.1
        W.f.SS1RR2 <- 0.1
        
        W.f.RS1SS2 <- 0.1
        W.f.RS1RS2 <- 0.1 
        W.f.RS1RR2 <- 0.1 
        
        W.f.RR1SS2 <- 0.1
        W.f.RR1RS2 <- 0.1 
        W.f.RR1RR2 <- 0.1
# to :
#to create an array a[sex][locus1][locus2]
namesLoci <- c('SS','RS','RR')
locus1 <- paste0(namesLoci,'1')
locus2 <- paste0(namesLoci,'2')
sex <- c("f","m")

dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
W <- array(0,dim=dim1, dimnames=dimnames1)

#to fill al elements as above
W[] <- 0.1
#check length
length(W)
#to access one element
W['m','SS1','RR2']
#to check that males total 1
sum(W['m',,])
```        

working on refactoring variables to arrays, made flexible createArray() function.
good progress
started runModel2() to contain the refactoring

**30/4/15** 9.45-13.30 3.75hrs 2.15-5.30 3.25hrs 6-7 8hrs 

Need to be careful about different W fitness variables
`W.RR1_00 : [locus,niche]` single locus (Wlocus)
`W.RR1SS2_0b :   [locus1, locus2, niche]` niche (Wniche)
`W.m.SS1SS2 : [sex, locus1, locus2]` sex (Windiv)

Once I have array refactoring working by .
```{r}

    a <- createArray( sex=c('m','f'), niche1=c('0','a','A'), niche2=c('0','b','B') )
    Wniche <- createArray( locus1 = c('SS1','RS1','RR1'), locus2 = c('SS2','RS2','RR2'), niche1=c('0','a','A'), niche2=c('0','b','B') )    
    Windiv <- createArray( sex=c('m','f'), locus1 = c('SS1','RS1','RR1'), locus2 = c('SS2','RS2','RR2') )
```
I should be able to replace 
```
    W.m.SS1SS2 <- (a.m_00 * W.SS1SS2_00) + 
      (a.m_a0 * W.SS1SS2_a0) + (a.m_A0 * W.SS1SS2_A0) + 
      (a.m_0b * W.SS1SS2_0b) + (a.m_0B * W.SS1SS2_0B) + 
      (a.m_ab * W.SS1SS2_ab) + (a.m_AB * W.SS1SS2_AB) + 
      (a.m_Ab * W.SS1SS2_Ab) + (a.m_aB * W.SS1SS2_aB) 
```
with
```{r}
    Windiv['m','SS1','SS2'] <- sum( a['m',,] * Wniche['SS1','SS2',,])
```
To prove that the above works
```{r}
    #set constant
    a['m',,] <- 2 
    #set variable
    Wniche['SS1','SS2',,] <- 1:9
    Wniche['SS1','SS2',,]
    
    #showing that each element gets multiplied by 2
    a['m',,] * Wniche['SS1','SS2',,]
    
    #show result of sum
    sum( a['m',,] * Wniche['SS1','SS2',,])
    
    Windiv['m','SS1','SS2'] <- sum( a['m',,] * Wniche['SS1','SS2',,])
    
```
Gives a result of 90 for just mSS1SS2 which is correct
``` Windiv
, , locus2 = SS2

   locus1
sex SS1 RS1 RR1
  m  90   0   0
  f   0   0   0

, , locus2 = RS2

   locus1
sex SS1 RS1 RR1
  m   0   0   0
  f   0   0   0

, , locus2 = RR2

   locus1
sex SS1 RS1 RR1
  m   0   0   0
  f   0   0   0
```

Or indeed might be able to replace the whole 120 lines of code from 510-630 with :
No these give: 
```
#Error in a[, , ] * Wniche[, , , ] : non-conformable arrays
    Windiv['m',,] <- sum( a['m',,] * Wniche[,,,])
    Windiv['f',,] <- sum( a['f',,] * Wniche[,,,])
    #or even just
    Windiv[,,] <- sum( a[,,] * Wniche[,,,]) 
```
Just needs a little more thought ...
It's because a needs to be applied repeatedly to each genotype.
Maybe worry about that later ..

Want to develop a sequential testing process, so that I can compare runModel() and runModel2().
Maybe I could use testthat for it ?

Just need to do this to start :
`devtools::use_testthat()`

see testRefactoring() which does things like this :

```
input <- setInputOneScenario()
tst <- runModel(input)
tst2 <- runModel2(input)
input <- setInputSensiScenarios(20, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))
tst <- runModel2(input, produce.plots=FALSE)
tst2 <- runModel(input, produce.plots=FALSE)
identical(tst,tst2)
```

So now I can keep developing runModel2() and use Ctrl T to test I'm not breaking it. 

Think, does my Wniche contain too many elements at 81 ?
i.e. does it contain impossible combinations ? I think I'm OK Beths code fills 9*9 groups of individual variables.

`W.RR1SS2_0b` ... `*81 3*3*3*3` to: Wniche[locus1, locus2, niche1, niche2] 

I may need to add a genotype option to createArray() to cope with f that needs cis & trans


If I convert niche_0B to niche[niche1,niche2] I can simplify further the calculation of 
Two genotype fitnesses in two insecticide Niche
by having a loop that goes through each niche.

Coolio! reduced 250 lines of code to ~ 20 with this
```
    ## Two genotype fitnesses in two insecticide Niche ##

    #!r to replace 250+ lines below
    for( niche1 in dimnames(Wniche)$niche1)
    {
      for( niche2 in dimnames(Wniche)$niche2)
      {
        #if this niche toggled off set fitness to 0
        if (niche[niche1,niche2] == 0)
        {
          Wniche[,,niche1,niche2] <- 0
        } else{
          #otherwise set fitness to product of the 2 loci
          for( locus1 in dimnames(Wniche)$locus1)
          {
            for( locus2 in dimnames(Wniche)$locus2)
            {    
              Wniche[locus1,locus2,niche1,niche2] <- Wloci[locus1,niche1,niche2] * Wloci[locus2,niche1,niche2]
            }
          }          
        }
      }
    }
```

Trying to get fitnesses from my array into the existing output object fbn at the end of runModel2

       -,- a,- A,- -,b -,B a,b A,B A,b a,B
SS1SS2   0   0   0   0   0   0   0   0   0
SS1RS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
SS1RR2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RS1SS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RS1RS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RS1RR2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RR1SS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RR1RS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RR1RR2  NA  NA  NA  NA  NA  NA  NA  NA  NA

STill not quite working ...

**1/5/15** 9-14.30 5.5hrs 3.15-6 2.75 8.25hrs

Fitness matrix different in my refactored version. But do I know which is correct ?

Browse[2]> fbn
       -,- a,- A,- -,b -,B a,b A,B A,b a,B
SS1SS2   1   0   0   0   0   0   0   0   0
SS1RS2   1   0   0   0   0   0   0   0   0
SS1RR2   1   0   0   0   0   0   0   0   0
RS1SS2   1   0   0   0   0   0   0   0   0
RS1RS2   1   0   0   0   0   0   1   0   0
RS1RR2   1   0   0   0   0   0   1   0   0
RR1SS2   1   0   0   0   0   0   0   0   0
RR1RS2   1   0   0   0   0   0   1   0   0
RR1RR2   1   0   0   0   0   0   1   0   0
Browse[2]> fbn2
       -,- a,- A,- -,b -,B a,b A,B A,b a,B
SS1SS2   0   0   0   0   0   0   0   0   0
SS1RS2   0   0   0   0   0   0   0   0   0
SS1RR2   0   0   0   0   0   0   0   0   0
RS1SS2   0   0   0   0   0   0   0   0   0
RS1RS2   0   0   0   0   0   0   0   0   0
RS1RR2   0   0   0   0   0   0   0   0   0
RR1SS2   0   0   0   0   0   0   0   0   0
RR1RS2   0   0   0   0   0   0   0   0   0
RR1RR2   0   0   0   0   0   0   0   0   0

Aha difference is because Wniche just has 0s in, may be a problem earlier in the program.
Aha issue is that Wloci is not read in yet.

Reading in of Wloci can also be refactored, but I first need to read in h,z,phi & s as arrays.

Then I can try to refactor this :
```
    ## Calculated fitnesses ####
    
    # absence of insecticide
    ## fitness of SS in absence of insecticide is entered above as a parameter
    W.RS1_00 <- 1 - (h.RS1_00 * z.RR1_00)
    W.RR1_00 <- 1 - z.RR1_00
    
    W.RS2_00 <- 1 - (h.RS2_00 * z.RR2_00)
    W.RR2_00 <- 1 - z.RR2_00
    
    # low levels of insecticide a
    W.SS1_a0 <- 1 - phi.SS1_a0
    W.RS1_a0 <- W.SS1_a0 + (h.RS1_a0 * s.RR1_a0)
    W.RR1_a0 <- W.SS1_a0 + s.RR1_a0
    
    # high levels of insecticide A
    W.SS1_A0 <- 1 - phi.SS1_A0
    W.RS1_A0 <- W.SS1_A0 + (h.RS1_A0 * s.RR1_A0)
    W.RR1_A0 <- W.SS1_A0 + s.RR1_A0
    
    # low levels of insecticide b
    W.SS2_0b <- 1 - phi.SS2_0b
    W.RS2_0b <- W.SS2_0b + (h.RS2_0b * s.RR2_0b)
    W.RR2_0b <- W.SS2_0b + s.RR2_0b
    
    # high levels of insecticide B
    W.SS2_0B <- 1 - phi.SS2_0B
    W.RS2_0B <- W.SS2_0B + (h.RS2_0B * s.RR2_0B)
    W.RR2_0B <- W.SS2_0B + s.RR2_0B
```

h only has 6 values 2*3. Because it's only for heterozygous loci & niches.

```
    # h = dominance coefficient
    h.RS1_00 <- input[32,i]
    h.RS1_a0 <- input[33,i]
    h.RS1_A0 <- input[34,i]
    
    h.RS2_00 <- input[35,i]
    h.RS2_0b <- input[36,i]
    h.RS2_0B <- input[37,i]
```

Perhaps best to store as
h[locusNum,niche1,niche2]

This nearly works, but includes un-needed homozygous niches aa etc 
createArray( loci=c('RS1','RS2'), niche1=c('0','a','A'), niche2=c('0','b','B') )

I could add a niches arg to createArray()
createArray( loci=c('RS1','RS2'), niches=c('00','a0','A0') )

BUT no even that doesn't work. 
Because it only needs to contain the insecticide niche associated with the locus.

Actually it just needs :
h[locusNum, nicheLevel] : where insecticide is no, lo, hi
... seeing how that works.

and what about z - fitness cost of resistance allele in no insecticide
z only has 2 values
    z.RR1_00 <- input[42,i]
    z.RR2_00 <- input[43,i]
z[locusNum]
Because by definition it is the fitness cost of one allele in absence of the corresponding insecticide. (so `RR1_00` is a little misleading `z.R1_0` more like it)

phi - fitness of SS in each insecticide/concentration 
just 4 values
    phi.SS1_a0 <- input[26,i]
    phi.SS1_A0 <- input[27,i]
    
    phi.SS2_0b <- input[28,i]
    phi.SS2_0B <- input[29,i]

phi[locusNum, nicheLevel] where nicheLevel is lo, hi

I have Wloci being 54
> str(Wloci)
 num [1:6, 1:3, 1:3] 0 0 0 0 0 0 0 0 0 0 ...
 - attr(*, "dimnames")=List of 3
  ..$ loci  : chr [1:6] "SS1" "RS1" "RR1" "SS2" ...
  ..$ niche1: chr [1:3] "0" "a" "A"
  ..$ niche2: chr [1:3] "0" "b" "B"
> length(Wloci)
[1] 54

but actually I think it only needs 1 niche that corresponds to the chosen locus.
for locus1 it only has as and locus2 it only has bs

look at how Wloci is used to think about how best to structure

This is where it is used 
`Wniche[locus1,locus2,niche1,niche2] <- Wloci[locus1,niche1,niche2] * Wloci[locus2,niche1,niche2]`
I think I need to change from 
`Wloci[locus1,niche1,niche2] * Wloci[locus2,niche1,niche2]`
to
`Wloci[locus1,niche1] * Wloci[locus2,niche2]`

I might want to change niche names to no,lo,hi throughout. Having them different for each locus (0aA & 0bB) causes the code difficulty.

But that is a bigger change making it more difficult to check that the code logic remains the same.

A temporray solution could be to have a lookup from nicheLevel to niche.
e.g. no,lo,hi to 0,a,A & 0,b,B for locus1 & 2 respectively

Later maybe
Wloci[(locusNum),allele=(SS,SR,RR),exposure=(no,lo,hi)]
for now
Wloci[loci=(SS1,SR1,RR1,SS2,SR2,RR2),exposure=(no,lo,hi)]

s selection coeeficient : just 4 values
    # s = selection coefficient
    s.RR1_a0 <- input[38,i]
    s.RR1_A0 <- input[39,i]
    
    s.RR2_0b <- input[40,i]
    s.RR2_0B <- input[41,i]

It's the selection coefficient of resistance in presence of the corresponding insecticide (at lo or hi)
s[locusNum,exposure]
but exposure is only lo or hi. Would a -ve value here for no be the same as z (the fitness cost of resistance in absence of insecticide ?)

cool I think I have generic createArray2 function working

I'm getting closer to fbn & fbn2 being the same

Differences in fbn2:
1) The 00 column has 0s, in fbn it is all 1s
2) Column aB has the values that should be in AB

e.g. SS1RR2
W.SS1RR2_00
Wniche['SS1','RR2','0','0']

Wloci[locus2,exposure2] == 0
but
W.RR2_00 == 1

aha think due to typo here :
Wloci[ paste0('RR',locusNum), 'no'] <- 1 - z[locusNum]

hurrah fixed first problem, now on to 2)

Wniche looks correct
So i think it may be due to ordering of columns in fbn
Which goes AB, Ab, aB, where I think I assume aB, Ab, AB

Wniche['RR1','RR2',,]
      niche2
niche1 0 b B
     0 1 0 0
     a 0 0 0
     A 0 0 1

as.vector( Wniche['RR1','RR2',,])
[1] 1 0 0 0 0 0 0 0 1

It may be tricky for me to get output in exactly the same order as Beths.
How can I use the niche names to name the output matrix ?

Difference in orders
colnames(fbn) <- c("-,-", "a,-", "A,-", "-,b", "-,B", "a,b", "A,B", "A,b", "a,B")
my order is         00     a0     A0     0b     ab     Ab     0B    aB     AB

which is 1,2,3,4,6,8,5,9,7

I might just need to allow that the fitness list is different in my refactor test ?

Still how can I name the output columns from the data ?
Fixed !!

Then I started to delete some of the old code, and found that it broke some things I didn't want to break yet.

Particularly this which suggests I need to get into the cis/trans issue soon:

W.bar.m <- Windiv['m']
W.bar.m <- (f.m.SS1SS2 * W.m.SS1SS2)

revert changes for now and come back to 
 
**5/5/15** 10.30 - 14 3.5hrs 2.45-6.15 7hrs
 
removed refactored fbn code 
 
sent update email to Ian & Beth 

starting to look at cis/trans
this is the first place they appear in the generation loop
```
# set genotype frequencies as variables
# from genotype frequency matrix generated above from initial value of P (freq. of R allele)
# f = frequency before selection

# male
f.m.SS1SS2 <- genotype.freq[1,]
f.m.SS1RS2 <- genotype.freq[2,]
f.m.SS1RR2 <- genotype.freq[3,]
f.m.RS1SS2 <- genotype.freq[4,]
f.m.RS1RS2_cis <- genotype.freq[5,]		### cis
f.m.RS1RS2_trans <- genotype.freq[6,]	### trans
``` 

How is genotype.freq created ?
```
# set up genotype matrix - records frequencies of each of the 9 two locus genotypes each generation
genotype <- matrix( nrow=max_gen, ncol=11 )
colnames(genotype) <- c("gen", "SS1SS2", "SS2RS2", "SS1RR2", 
                        "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                        "RR1SS2", "RR1RS2", "RR1RR2")
## make.genotypemat function will use this data and make a matrix of the genotype frequencies
## frequencies of genotypes before selection - in HW equilibrium and same in male and female
## needs name of matrix and takes corresponding frequency of resistant allele in function call
genotype.freq <- make.genotypemat ( P_1, P_2 )
```

Within make.genotypemat, currently seems that cis & trans are always set to the same ??
```
## two forms of RS1RS2
mat[5,] <- ( loc1[2,]*loc2[2,] ) * 0.5 #cis
mat[6,] <- ( loc1[2,]*loc2[2,] ) * 0.5 #trans

#this is what genotype.freq looks like
                      Freq.
SS1SS2       9.960060e-01
SS1RS2       1.994006e-03
SS1RR2       9.980010e-07
RS1SS2       1.994006e-03
RS1RS2_cis   1.996002e-06
RS1RS2_trans 1.996002e-06
RS1RR2       1.998000e-09
RR1SS2       9.980010e-07
RR1RS2       1.998000e-09
RR1RR2       1.000000e-12
```

aarg createArray2() fails if you pass named variables to it, e.g. sex=sex rather than sex=c('m','f')
seems like mget might solve, but :
mget('sex', inherits=TRUE)
Error: value for ‘sex’ not found

may be able to use this from SO
get_args <- function () {
as.list( match.call(
    def = sys.function( -1 ),
    call = sys.call(-1)) )[-1]

}

BUT from the commandLine it does seem to work ??
```
> sex2 = c('f','m')
> createArray2( sex = sex2 )
sex
f m 
0 0 
```
Ahhh so someohow it does work if the variable is in the Global environment.

eventually by trial & error found this solution :
  #fails
  #dimnames1 <- lapply(listArgs,function(x){eval.parent(x,n=1)})
  #works
  #eval.parent(listArgs$sex)
  #[1] "f" "m"
  #fails
  #dimnames1 <- lapply(listArgs,eval.parent)  
  #works!! n=3 but I'm not sure why !!!
  dimnames1 <- lapply(listArgs,function(x){eval.parent(x,n=3)}) 
 
 
**11/5/15** 1.30-5.45 4.25hrs   
checking current testthat difference
changing from `expect_identical` to `expect_equal` sorted it 
refactoring gamete calculations into createGametes() function 
 

**12/5/15** 9.45-14 4.25hrs 2.30-5.45 3.25 7.5hrs

sorting random mating & maybe getting to nub of cis/trans ?

Beths random mating code :
```
      f.m.SS1SS2 <- 0
      f.m.SS1RS2 <- 0
      f.m.SS1RR2 <- 0
      
      f.m.RS1SS2 <- 0
      f.m.RS1RS2_cis <- 0			#RS1RS2
      f.m.RS1RS2_trans <- 0		#RS1SR2
      f.m.RS1RR2 <- 0
      
      f.m.RR1SS2 <- 0
      f.m.RR1RS2 <- 0 
      f.m.RR1RR2 <- 0
      
      # SS male with SS female
      f.m.SS1SS2 <- f.m.SS1SS2 + ( G.m.S1.S2 * G.f.S1.S2 )
      # SS male with SR female
      f.m.SS1RS2 <- f.m.SS1RS2 + ( G.m.S1.S2 * G.f.S1.R2 )
      # SS male with RS female
      f.m.RS1SS2 <- f.m.RS1SS2 + ( G.m.S1.S2 * G.f.R1.S2 )
      # SS male with RR female
      f.m.RS1RS2_cis <- f.m.RS1RS2_cis + ( G.m.S1.S2 * G.f.R1.R2 )
      
      # SR male with SS female
      f.m.SS1RS2 <- f.m.SS1RS2 + ( G.m.S1.R2 * G.f.S1.S2 )
      # SR male with SR female
      f.m.SS1RR2 <- f.m.SS1RR2 + ( G.m.S1.R2 * G.f.S1.R2 )
      # SR male with RS female
      f.m.RS1RS2_trans <- f.m.RS1RS2_trans + ( G.m.S1.R2 * G.f.R1.S2 )
      # SR male with RR female
      f.m.RS1RR2 <- f.m.RS1RR2 + ( G.m.S1.R2 * G.f.R1.R2 )
      
      # RS male with SS female
      f.m.RS1SS2 <- f.m.RS1SS2 + ( G.m.R1.S2 * G.f.S1.S2 )
      # RS male with SR female
      f.m.RS1RS2_trans <- f.m.RS1RS2_trans + ( G.m.R1.S2 * G.f.S1.R2 )
      # RS male with RS female
      f.m.RR1SS2 <- f.m.RR1SS2 + ( G.m.R1.S2 * G.f.R1.S2 )
      # RS male with RR female
      f.m.RR1RS2 <- f.m.RR1RS2 + ( G.m.R1.S2 * G.f.R1.R2 )
      
      # RR male with SS female
      f.m.RS1RS2_cis <- f.m.RS1RS2_cis + ( G.m.R1.R2 * G.f.S1.S2 ) 
      # RR male with SR female
      f.m.RS1RR2 <- f.m.RS1RR2 + ( G.m.R1.R2 * G.f.S1.R2 )
      # RR male with RS female
      f.m.RR1RS2 <- f.m.RR1RS2 + ( G.m.R1.R2 * G.f.R1.S2 )
      # RR male with RR female
      f.m.RR1RR2 <- f.m.RR1RR2 + ( G.m.R1.R2 * G.f.R1.R2 )
```

Beth only has 10 combinations (including cis/trans), this is because RS is usually same as SR, and in all the names R goes first.
```
genotypes <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
                "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                "RR1SS2", "RR1RS2", "RR1RR2")
```                            


for randomMating this loop creates 16 combinations
including cis/trans

```
  counter <- 0
  
  for( m2 in c('S2','R2'))
  {
    for( m1 in c('S1','R1'))
    {
      for( f2 in c('S2','R2'))
      {
        for( f1 in c('S1','R1'))
        {
          counter <- counter+1
          cat(paste(counter, m1,f1,m2,f2,"\n"))
        }
      }
    }
  }
```
1 S1 S1 S2 S2 
2 S1 R1 S2 S2 
3 S1 S1 S2 R2 
4 S1 R1 S2 R2 
5 R1 S1 S2 S2 
6 R1 R1 S2 S2 
7 R1 S1 S2 R2 
8 R1 R1 S2 R2 
9 S1 S1 R2 S2 
10 S1 R1 R2 S2 
11 S1 S1 R2 R2 
12 S1 R1 R2 R2 
13 R1 S1 R2 S2 
14 R1 R1 R2 S2 
15 R1 S1 R2 R2 
16 R1 R1 R2 R2 

reformat slightly :
1 SS1 SS2
2 SR1 SS2
3 SS1 SR2
4 SR1 SR2
5 RS1 SS2
6 RR1 SS2
7 RS1 SR2
8 RR1 SR2
9 SS1 RS2
10 SR1 RS2
11 SS1 RR2
12 SR1 RR2
13 RS1 RS2
14 RR1 RS2
15 RS1 RR2
16 RR1 RR2

Perhaps I can do the calculations on these 16 & then aggregate to produce the 10 ?

```
#these are other ways of creating the 16

#as an array
arr <- createArray2(loc1all1=c('S','R'),loc1all2=c('S','R'),loc2all1=c('S','R'),loc2all2=c('S','R'))

#as a dataframe
dF <- expand.grid(loc1all1=c('S','R'),loc1all2=c('S','R'),loc2all1=c('S','R'),loc2all2=c('S','R'))
> dF
   loc1all1 loc1all2 loc2all1 loc2all2
1         S        S        S        S
2         R        S        S        S
3         S        R        S        S
4         R        R        S        S
5         S        S        R        S
6         R        S        R        S
7         S        R        R        S
8         R        R        R        S
9         S        S        S        R
10        R        S        S        R
11        S        R        S        R
12        R        R        S        R
13        S        S        R        R
14        R        S        R        R
15        S        R        R        R
16        R        R        R        R

genotypes <- paste0(dF[,1],dF[,2]," ",dF[,3],dF[,4])
#[1] "SS SS" "RS SS" "SR SS" "RR SS" "SS RS" "RS RS" "SR RS" "RR RS" "SS SR" "RS SR" "SR SR" "RR SR" "SS RR" "RS RR" "SR RR" "RR RR"

> length(genotypes)
[1] 16

#or could miss out the space

l1a1 <- substr(genotypes,1,1)
l1a2 <- substr(genotypes,2,2)
l2a1 <- substr(genotypes,4,4)
l2a2 <- substr(genotypes,5,5)

#these are homozygous at both loci and have 1:1 between expanded & contracted genotypes
which(l1a1==l1a2 & l2a1==l2a2)
[1]  1  4 13 16

#this gets the cis/trans
which(!(l1a1==l1a2 | l2a1==l2a2))
[1]  6  7 10 11

But I might not to be this fancy. If I do all the working in the expanded genotype storage, and then just have a single function to do the conversion ??

genotypesContract()
or genotypesLong2Short()

```
 
done : convert from the expanded genotype storage to the contracted one. 
 
Beware I do seem to have occasional small difference between runModel & runModel2. I set the random seed so that the difference is consistent.

It now seems only to turn up if I have >20 scenarios.
 
It's not to do with randomMating() or genotypesLong2Short because they are not fully implemented yet.

Failure(@testRefactoring.r#21): refactoring named variables with arrays doesn't change results 
runModel2(input, produce.plots = FALSE) not equal to runModel(input, produce.plots = FALSE)
Component "results": Component 25: Mean relative difference: 6.391493e-08

It does show some very small differences, but I suspect they are just rounding errors, e.g. this is how they start out. (and the M,F columns just contain 1 despite showing this difference here)

```
listOut2$results[[25]] - listOut$results[[25]]

       Gen          m.R1          m.R2          m.LD          f.R1          f.R2          f.LD             M             F
[1,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
  [2,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
  [3,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 -2.220446e-16 -2.220446e-16
  [4,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
  [5,]   0 -2.168404e-19 -2.168404e-19  0.000000e+00 -2.168404e-19 -2.168404e-19  0.000000e+00 -1.110223e-16 -1.110223e-16
```

tested that my new W.bar calc works :
W.bar
        m         f 
0.1000036 0.1000036 
Browse[2]> W.bar.m
   a.m_00 
0.1000036 
Browse[2]> W.bar.f
   a.f_00 
0.1000036 

**13/5/15** 10-14 2
 
dealing with cis/trans :
```
namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2cis','RS1RS2trans','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')
#or
namesLoci <- rownames( genotype.freq )
sex <- c("F","M")
f <- createArray2( sex=sex, loci=namesLoci )
#i then may be able to refactor more later
```
Plus see randomMating() which uses something like below, and then genotypesLong2Short.r to convert from the expanded genotype format.
```
  counter <- 0
  
  for( m2 in c('S2','R2'))
  {
    for( m1 in c('S1','R1'))
    {
      for( f2 in c('S2','R2'))
      {
        for( f1 in c('S1','R1'))
        {
          counter <- counter+1
          cat(paste(counter, m1,f1,m2,f2,"\n"))
        }
      }
    }
  }
```

deleting old code from runModel2() 

* done to f[m,loci] : sort storage of f.m.RS1RS2_cis
* done : refactor gamete calculations : G.m.R1.R2 <- G.m.R1.R2 + f.m.RS1RS2_cis ...
* done : sort calc of W.bar.m from Windiv
* done : remove reliance on W.m.SS1SS2 etc. (Windiv)
* done : remove reliance on a.m_00 etc.

So far down from 1590 lines to 685, with only a bit in new functions 

**20/5/15** 11-4 in liverpool meeting with Beth and train back

New bit from the manuscript by Ian about including sex-linked genes by a bit of trickery.

It is also possible to allow one of the loci to be sex-linked; here we assume that sex linkage is at Locus 1, and ‘conventional’ sex-determination i.e. that the females is hemozygous sex (XX), the male is XY, and that the gene exists only on the X chromosome. This option is important as Anopheles gambiae has only 2 autosomes and one sex chromosome; approximately ??% of the genes are on the X in this species.

The female genotype frequencies are as given on Table S1 and as described in Equation 4 and Equation 5  because females have two copies of the X chromosome. Locus 1 is homozygous in the male so heterozygotes are impossible at this locus; in this case the allele inherited by males at locus 1  is the maternal-derived one (because they get their X chromosome from their mother and the Y from the father). Hence the male genotypes can be derived by a process similar to that of females and autosomes as shown on Table S2. The males will be simulated  as “RR” or “SS” at the locus even though, in reality” they will be either “R-“ or “S-“. An inherent assumption is therefore that the hemizgous males “R-“ and “S-“ have the same fitness as the equivalent “RR” and “SS” females; this seems reasonable given dosage compensation and is an assumption commonly made.

The male “diploid” genotypes are derived by a similar process to that above and the genotypes are as given on Table S2. The calculation become slightly more complicated to those given above but the same process occurs i.e. Equation 4 becomes

Equation 6

The next genotype can be produced by two combinations so Equation 5 becomes

Equation 7 ...

IH thinks this short-cut will work because selection will be OK, and the males will also correctly pass on their maternal X-linked allele (Table 4). We need to discuss this.

It just means we need edit the code slightly. i.e. calculate the female genotypes then if locus 1 in autosomal just set the male frequencies equal to the females (as we do already), else re-calculate the male frequencies according to Table S2.

end of bit from manuscript
I think this is the important part for implementation :

we assume that sex linkage is at Locus 1, females XX, the male XY, 
and that the gene exists only on the X chromosome. 

If sex-linked Locus 1 is homozygous in the male so heterozygotes are impossible at this locus 
the allele inherited by males at locus 1 is the maternal-derived one 
(because they get their X chromosome from their mother and the Y from the father). 
males will be simulated  as RR or SS at the locus even though, in reality they will be either R- or S-


I think this change needs to be made in the randomMating() function.
Currently i think randomMating() doesn't differentiate between the sex of the offspring.

Does sex linkage have further implications in terms of the generation of gametes ? maybe not 
 
**22/5/15**
To stop things being tracked by git that you have just added to gitignore
```
git rm --cached `git ls-files -i -X .gitignore` 
``` 
Then I just did commit from RStudio.

1. somehow I've broken the check
Failure(@testRefactoring.r#12): refactoring named variables with arrays doesn't change results 
runModel2(input, produce.plots = FALSE) not equal to runModel(input, produce.plots = FALSE)
Component "results": Component 1: 'is.NA' value mismatch: 186 in current 0 in target
Its because of Warning messages like this:
1: In runModel2(input) : Male frequencies before selection total != 1 0.997966163681722

Perhaps I need to make the checks less sensitive ??
Is it because I re-enabled checks that had been disabled by Beth ?
I reduced the tolerance in checks, runs OK, but tests still fail.

Points to some difference having crept in ...

difference is in 
genotype outputs in generations after 1
all those starting with RS1

to do with object genotype ....
Aha! problems were due to my new sex-linked changes that were operating without me realising.
So I should probaly remove the tolerances I put in. and work out how to reassign the impossible genotype freqs.
done 
 
**4/6/15** thurs 12.15- 

checking on implementation of sexLinkage in randomMating() 
 
sexLinked probably needs to be included as an extra paramter in input (but should make it robust to cope if an input file doesn't have the sex-linked option)

be careful because this could break previous model runs

genotype.freq mostly removed it was the same as f['m',]. Creation of f & fs moved to before the generations loop. This was prompted because previously genotype.freq was the same for males and females and it adds an unecessary extra layer of potential confusion.
 
added calculating genotypes of both m&f in runModel2() to allow sex linkage.
 
bug in random mating with sex linkage genotypes aren't summing to 1

fGenotypeExpanded for non sex-linked looks like the top one
In the lower one the R1 homozgotes should sum to 1 too

    l1a2
l1a1     S1     R1
  S1 0.0625 0.0625
  R1 0.0625 0.0625

    l1a2
l1a1    S1     R1
  S1 0.125 0.0000
  R1 0.000 0.0625

Fixed now :
with this bit of code :
```
#heterozygotes at locus1 are impossible so add to the homozygotes instead
if(f1!=m1)
{
  #add to the homozygotes : f1,f1
  fGenotypeExpanded[f1,f1,f2,m2] <- fGenotypeExpanded[f1,f1,f2,m2] + fThisGenotype
} else #i.e. if not heterozygous at locus 1
{
  #just add to this genotype : f1,m1
  #have to add rather than set in case this is a homozygote that has already been added to by previous condition
  fGenotypeExpanded[f1,m1,f2,m2] <- fGenotypeExpanded[f1,m1,f2,m2] + fThisGenotype
}
```

Created a new shiny app in shinyCurtis2 and run from runUI2.R to look at graphs with & without sex linkage side by side. Initially these gave identical results, which was because resistSimple() called the old runModel() which doesn't allow sex linkage. Try it again now.

Cool, now I am getting a difference between locus1 & locus2. However males & females still have the same results, is that what would be expected ?

Maybe it's because of the way that results are output from Beths one sex system ?
Where does the graph get the data from ?

genplot <- plotallele.freq.andy( listOut$results[[1]] ) 

  # Males
  lines( mat[,1], mat[,2], col="darkblue", lwd=2, lty=13 )
  lines( mat[,1], mat[,3], col="green", lwd=2, lty=14 )
  # Females
  lines( mat[,1], mat[,5], col="red",lwd=2, lty=15 )
  lines( mat[,1], mat[,6], col="orange", lwd=2, lty=16 )

so its columns 2,3,5 & 6 I'm interested in

so where is listOut$results[[1]] filled ?
in runModel2()     listOut$results[[i]] <- results
& further back :
```
      ## frequency of resistance alleles
      m.R1 <- sum(f['m',grep("RR1",colnames(f))]) + ( 0.5 * sum(f['m',grep("RS1",colnames(f))]))
      m.R2 <- sum(f['m',grep("RR2",colnames(f))]) + ( 0.5 * sum(f['m',grep("RS2",colnames(f))]))
      f.R1 <- sum(f['f',grep("RR1",colnames(f))]) + ( 0.5 * sum(f['f',grep("RS1",colnames(f))]))
      f.R2 <- sum(f['f',grep("RR2",colnames(f))]) + ( 0.5 * sum(f['f',grep("RS2",colnames(f))]))   
      results[k,2] <- m.R1
      results[k,3] <- m.R2
      results[k,5] <- f.R1
      results[k,6] <- f.R2
```

Seems that this should be doing the right thing. Maybe the allele frequencies overall don't change, because for the males they just redistributed from heterozygotes to homozygotes.

**5/6/15**

Instructions emailed to Ian & Beth to run UI2
require(devtools)
install_github('AndySouth/resistance')
require(resistance)
runUI2()

Ians feedback on UI2 (sex-linkage)

The R script runs fine. The results are as I would have expected given that locus 1 is sex-linked and locus 2 is autosomal i.e. resistance evolves faster at the sex-linked locus.

The dynamics at locus 2 appear unchanged...I therefore assume the drugs were deployed individually i.e. not as a combination?

But I thought they were applied as a combination.
The exposure comes from here in the UI :

 h5("Exposure to each insecticide"),
 numericInput("a.m_AB", "same for both insecticides in Curtis", 0.5, min = 0.1, max = 0.9, step = 0.1)

and is set here in server :
a.m_AB = input$a.m_AB,
a.f_AB = input$a.m_AB, #set f to same as m

are these the correct parameters to set ??
yes i think so.


Thanks Ian,
In relation to your 2nd comment, from the UI, both a.m_AB and a.f_AB
are set from the box with "Exposure to each insecticide".

It's my understanding that should result in exposure to high levels of
both insecticides (i.e. in combination).

Is that correct Beth ?

Didn't get a reply from Beth yet.
From Ian :
OK, I was just speculating that if resistnce at locus 1 was spreading faster due to sex-linkage then so might resistnce at locus2 through genic linkage and hitch-hiking. I'll give its some more thought over the weekend.

be careful that runModel & runModel2 should no longer generate same results when sex linkage is set (but i don't think this should break current testthat tests)

**8/6/15** on train to liverpool 4.30-9.30
**The plan for running the sensitivity analysis for the first paper**

Edited from Ians text in the paper.

For each parameter combination 2 strategies : 
A) single ‘new’ insecticide alone
B) ‘new’ insecticide in combination with ‘older’ insecticide 

The simulation output was the ratio of number of generations for resistance  to the ‘new’ insecticide  to reach a given allele frequency when in combination compared to on its own.

A ratio >1 indicates that combination is better. 

Five threshold frequencies were investigated: 0.05, 0.1, 0.2, 0.5, 0.8.

But in the document :
"Suggested sensitivity analysis.docx" it talks about sequential and 2 mixture optoins.

Sequential use:, use one insecticide until resistance frequency reaches critical point, then switch insecticide until second resistance allele frequency reaches the critical point. Record time in total generations for the two insecticides.

Mixture option 1. Deploy mixture and record time in generations until EITHER frequency reaches the critical point

Mixture, option 2: deploy mixture until one resistance allele reaches the critical point, then stop using that insecticide in the mixture. Record time until both resistance allele frequencies reach critical point.

Easy by post-processing : Mixture 1 : set exposure to both insecticides
Trickier : Sequential
Trickiest : Mixture 2

Mixture 1 : set exposure to both insecticides and post process to get time until critical points for EITHER insecticide is reached.
Sequential : set callibration to 1013, add bit to runModel2() to switch from insecticide1 to insecticide2 when critical point for the first insecticide is reached. Post process results file to find the time that the 2nd critical point is reached.

Tried adding a sequential insecticide scenario as an example to runModel2(), doesn't seem to be working yet, or maybe I just haven't passed it the correct args.

aha, i think probably because the niche toggles not switched on. Can I set all of these to on by default so that this problem doesn't happen again ?? done

resistance to 2nd insecticide still seems not to rise.

It may be because I need to reset the fitnesses (which happens before gen loop):           

Windiv[sex,locus1,locus2] <- sum( a[sex,,] * Wniche[locus1,locus2,,])
and could put this into a function called setFitness() or the like.


** 9/6/15 ** 4-5

Chatting with Ian & working out that both sequential and mixture2 scenarios can also be done by post-processing of the output file.

Now looking at setting up sensiAnPaper1() to run the sensitivity analysis for the paper.
Realising that I might not be able to call sensiAn1 and setInputSensiScenarios() because of some of the dependencies between parameters.

Particularly that `P_2` needs to be set as a proprtion of `P_1`.

Actually sensiAnPaper1() can be much more straightforward if it has hardcoded args and doesn't need to be so flexible.


** 10/6/15 ** train back from liverpool

modifying sensiAnPaper1() to run the different insecticideUsed scenarios
started sensiPostProc() to do post-processing


** 24/6/15 ** in liverpool before talking to Ian

checking where I have got to.

I'm going to renam sensiAnPaper1 to sensiAnPaperPart()
then create a sensiAnPaperAll() to run everything (although in practice unlikely to run it all at once)

** 29/6/15 **
little on sensiPostProc()

** 30/6/15 **
moved sensiAnPaper1All() into an rmarkdown file - because in reality I'm not likely to run it all at once.

Beths work just uses male allele frequencies to assess criticalPoints m.R1 & f.R1 are the same for now, but it might be safer for me to use an average.

~ added the input matrix onto the listOut output object in runModel2() to help with post processing
Slight issue that this has now broken my testthat tests !!


How can m.R2 and f.R2 be greater than 1 (e.g. they start at 1.7455 in first run) ?
Is this because the frequency of R2 is set too high at start ?

m.R2 is calculated at line 801 in the older runModel.r and line 376 in runModel2

Frequency of resistance allele to the 2nd insecticide is starting above 1.

m.R2 <- ( f.m.SS1RR2 + f.m.RS1RR2 + f.m.RR1RR2 ) + ( 0.5 * (f.m.SS1RS2 + f.m.RS1RS2_cis + f.m.RS1RS2_trans + f.m.RR1RS2 ) )

#refactored
m.R2 <- sum(f['m',grep("RR2",colnames(f))]) + ( 0.5 * sum(f['m',grep("RS2",colnames(f))]))
 
Can I create a reproducible example ? or should I just ask first ?

What is the starting value of P_2 the starting frequency of resistance to insecticide2 ?
For the first dodgy run
listOutMix$input['P_2',1]
    P_2 
1.74551 

Aha yes, so maybe i should put something in to stop it going above 1 ??

I think problem was because I set the upper limit on P_2 to 100 rather than 1 (based upon Ians table).

FRom Ian :
A trick one that..

I think my first instinct was to not restrict one insecticide resistance frequency to be always  less than the other. But we would have to censor the inputs because if frequency at locus 1  is 0.1 then the frequency at locus 2 cannot be 100 times higher.

I think we should go with your suggestion i.e. recognise that frequency at one locus will always be less than at the other and the lower frequency occurs at Locus 2. I would maybe go from 0.0001 to 1 under these circumstances so we get a wide range of starting frequency ratios.

Cool findResistancePoints() seems to be working

> findResistancePoints(listOut, locus=1)
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19]
gen_cP0.1     9   10    5   36    4    3    5   19   48    77    12    10     3     2     4     3    31    36     5   
gen_cP0.25   20   20   27   70   15   16   11   41   72   120    19    16    17     5     8     9    54    71    20   
gen_cP0.5    27   29   39   89   28   25   18   56   97   146    25    23    36     9    12    16    73    95    39   
> findResistancePoints(listOut, locus=2)
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19] 
gen_cP0.1     4   31   21   76   25   14    7   30   23   301    59     8    34     2    25    91    19    26    97   
gen_cP0.25    8   48   45  128   56   26   16   40   35   334    79    12    42     5    30   118    29    53   134   
gen_cP0.5    13   61   62  154   77   36   24   51   48   357    90    17    51     9    34   133    39    81   166   
> findResistancePoints(listOut, locus='either')
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19]
gen_cP0.1     4   10    5   36    4    3    5   19   23    77    12     8     3     2     4     3    19    26     5 
gen_cP0.25    8   20   27   70   15   16   11   40   35   120    19    12    17     5     8     9    29    53    20   
gen_cP0.5    13   29   39   89   28   25   18   51   48   146    25    17    36     9    12    16    39    81    39   
> findResistancePoints(listOut, locus='both')
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19]
gen_cP0.1     9   31   21   76   25   14    7   30   48   301    59    10    34     2    25    91    31    36    97   
gen_cP0.25   20   48   45  128   56   26   16   41   72   334    79    16    42     5    30   118    54    71   134   
gen_cP0.5    27   61   62  154   77   36   24   56   97   357    90    23    51     9    34   133    73    95   166   

but the 2nd mixture scenario is the tricky bit because 

need to know which of the 2 insecticides reached the critical point first, and the resistance level of the other. I think I'll need to create a special function.

findResistancePointsMixResponsive()

** 31/6/15 **
Quick review of what needs to go into the final decision tree.

slight rethink : I want the input files to be mostly the same for I1,I2 & mixture, so that they can be added together in combined scenarios.

I added a set.seed into sensiAnPaperPart()
This shows the input files for diffeerent insecticide strategies are now the same.

tst <- listOutMix$input[c('P_1','P_2','a.f_AB','phi.SS1_A0','phi.SS2_0B','h.RS1_A0','h.RS2_0B','s.RR1_A0','s.RR2_0B'),]
#note a.f_A0 rather than a.f_AB
tst2 <- listOutI1$input[c('P_1','P_2','a.f_A0','phi.SS1_A0','phi.SS2_0B','h.RS1_A0','h.RS2_0B','s.RR1_A0','s.RR2_0B'),]
identical(as.numeric(tst),as.numeric(tst2))


** 6/7/15 **
`plotcurtis_f2_generic` : a function allowing mixture & individual scenarios to be compared, similar to Curtis Fig2. 

In creating a shiny app to display this, I could allow user to select parameters, but then the run might take a little while because it has to do 3 runs (I1,I2,mix) & post processing. Quicker may be to display my sensitivity analysis results & allow user to step through scenarios. + include printing out the input values used (& maybe even their relative position within ranges). 


** 7/7/15 **
resistanceMaster() does reproduce curtis fig2
so the parameters in system.file("extdata","input.parameters.csv", package="resistance")
must work 

what is different about the input params in the sensitivity analysis ?

resistanceMaster() still calls runModel() if I change it to call runModel2() does it still work ??

No I get this :
Error in runModel2(input, calibration) : 
  Error in male exposures: must total one:  1.9

but in the input files, looks like it is -- 0.1 & AB 0.9 for all. 
 
No not in the hardcoded, curtis fig2 input which comes from the program. It has a 0.9 for ab which slips in too. 
 
                                             B (HCH)   A (DDT) Combination
Calibration (100 default)                  1012.0000 1012.0000   1.012e+03
Number of generations                        70.0000   70.0000   1.600e+02
Collect fitness scores in matrix (1/0)        1.0000    1.0000   1.000e+00
Export fitness scores matrix to .csv (1/0)    0.0000    0.0000   0.000e+00
Frequency of R at locus 1                     0.0100    0.0100   1.000e-02
Frequency of R at locus 2                     0.0100    0.0100   1.000e-02
Recombination Rate                            0.5000    0.5000   5.000e-01
Exposure Males -,-                            0.1000    0.1000   1.000e-01
Exposure Males a,-                            0.0000    0.0000   0.000e+00
Exposure Males A,-                            0.0000    0.9000   0.000e+00
Exposure Males -,b                            0.0000    0.0000   0.000e+00
Exposure Males -,B                            0.9000    0.0000   0.000e+00
Exposure males a,b                            0.0000    0.0000   9.000e-01
Exposure Males A,B                            0.0000    0.0000   9.000e-01
...

exactly where is that input object created ?
createInputMatrix()
fixed bug, but it didn't effect the output


can I compare the inputs for the 3 scenarios by Beth to those in my first scenario ? in a csv.

This gets the file to create Curtis Fig2.
```
inputCurtis <- createInputMatrix(FALSE)

```

Can I paste the input for the first scenario onto it ?
Or might be better just to output all 100 scenarios.

see : 

seems that the dominance coefficient of locus2 in -B is much lower in Beths than in the sensi scenarios.

Dominance coefficient L2 in -,B	0.0016	0.0016	0.00016	0.944675269	0.497699242

First see if the curtis fig2 persists when dominance for the mixture is reduced.

Good news that this correction seemed to make the graph closer to Curtis. 
    #input[37,3] <- 0.00016	#Dominance coefficient in B
    #? should this be the same as the value in ,2 & ,1
    input[37,3] <- 0.0016	#Dominance coefficient in B

** 7/7/15 **
new version of UI comparing scenarios with greater differences in dominance.
Still doesn't produce something similar to curtis fig2 where mixture does much better.

created runcurtis_f2() to do runs to recreate curtis fig2. recreates it fine.

created shinyFig2Curtis, that calls runcurtis_f2()

for some reason it currently doesn't recreate the curtis plot.
what inputs are wrong in shinyFig2Curtis ?
aha! the difference was just `P_1 & P_2` which were set to 0.1 rather than 0.01
aha2, even after that fixed, something still not quite right

this was because phi.SS2_0B = 1 was set to 0 instead. This is the fitness of exposed susceptibles.
(1 suggests that the susceptible is completely fit in presence of the insecticide which doesn't seem right).

yes phi.SS2_0B seems key :
compare these 2:
```{r}
#curtis default phi.SS2_0B = 1 : mixture better than sequence
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )
#reduce phi.SS2_0B to 0.6 : mixture same as sequence
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 0.6 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

```

How does that relate to this comment from the paper ?
In particular, Curtis [1] allowed the resistant and sensitive homozygotes to survive DDT exposure with finesses of 50% and 27% respectively (caption to his Figure 2).
 
packageDescription('resistance')

GithubSHA1: 7858f1edc0022233c0cd26f84be64e1a9bd0cd04 
            7858f1edc0022233c0cd26f84be64e1a9bd0cd04
 
This shows that the installed package is the latest version from github.

problem was because I hadn't updated tghe NAMESPACE Ctrl Shift D

reran model with new ranges for phi vals (0.6 to 1)
(but still need to sort finite 'xlim' values errors)

e.g. scenario 2

Email from Ian, showing section vi of sensitivity doc.

We have to ensure the fitness of the RS and RR genotypes does not exceed 1.0 as this is the reference (maximum value of the SS genotype in the absence of insecticide. The fitness of the SS homozygote is determined by the ϕ values in Table 3 i.e. as 1- ϕ. Hence the values of s must be less than ϕ. I suggested we set s=x ϕ where x comes for a uniform distribution between 0.2 and 1.0”

This relates to this bit of code :

      for( exposure in c('lo','hi') )
      {
        Wloci[ paste0('SS',locusNum), exposure] <-  1 - phi[locusNum, exposure] 
        
        Wloci[ paste0('RS',locusNum), exposure] <- (1 - phi[locusNum, exposure]) + 
                                                   (h[locusNum, exposure] * s[locusNum, exposure])
        
        Wloci[ paste0('RR',locusNum), exposure] <- (1 - phi[locusNum, exposure]) + 
                                                   (s[locusNum, exposure])
      }
      

Wloci[RR] <- (1-phi) + s

so s must be < phi to keep Wloci[RR] below 1

added in checks for fitnesses >1 or <0 in runModel2


initial tree looks promising :
for mix1 better than sequential:
phi.SS2_0B < 0.66
phi.SS1_AO < 0.68
P_2 > 0.033
P_1


** 10/7/15 **
installed rpart.plot
for prp function used by susana

created loop to do all plots, with titles and save

repeated runs for 500 scenarios, took about an hour.
 
** 11/7/15 **
running 1000 scenarios
started at 10.15, check timestamp on outputs to see how long it took

** 13//7/15 **

done created findResistancePointsMixResponsive() for mix2

analysing 1000 runs

  #1000runs
  #betterMix1Seq_cP0.1 betterMix1Seq_cP0.25  betterMix1Seq_cP0.5 
  #              0.178                0.232                0.238 
  #betterMix2Seq_cP0.1 betterMix2Seq_cP0.25  betterMix2Seq_cP0.5 
  #              0.471                0.568                0.551 
  #betterMix3Seq_cP0.1 betterMix3Seq_cP0.25  betterMix3Seq_cP0.5 
  #              0.794                0.726                0.673 
rounded
  #betterMix1Seq_cP0.1 betterMix1Seq_cP0.25  betterMix1Seq_cP0.5 
  #              0.18                0.23                0.24 
  #betterMix2Seq_cP0.1 betterMix2Seq_cP0.25  betterMix2Seq_cP0.5 
  #              0.47                0.57                0.55 
  #betterMix3Seq_cP0.1 betterMix3Seq_cP0.25  betterMix3Seq_cP0.5 
  #              0.79                0.73                0.67 
  
essentially
mix1 : 0.2
mix2 : 0.5
mix3 : 0.7


alas the trees are more complicated, but that might be fixed by pruning them ?
maybe not ... 
does work well for :
betterMix1Seq_cP0.1
 
BUT wierdly seems that plots look better (ie much fewer branches) when done individually than when saved in the loop.

is it something to do with interaction between prp and savePlot() ?
 
Now I seem to be getting a different result each time ????? 
 
** 14/7/15 **
 
putting results document together. 
 
Just checking that the results file for the 1000 runs looks like the figures & UI for the 100 runs 
 
run3: sequential best (seq 28+29, mix3 50)
run4: mixture best (seq 93+32, mix3 172)
run5: resistance to one in mixture develops more slowly than both in sequence (seq 29+103, mix1 152)
 
resistPointsI1[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1    38  999   11   57    7
gen_cP0.25   48  999   21   82   22
gen_cP0.5    53  999   28   93   29
resistPointsI2[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1    64  210    7    4   91
gen_cP0.25   90  250   21   21   99
gen_cP0.5   102  268   29   32  103
resistPointsMix1[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1    75  999    9   10   34
gen_cP0.25   93  999   30   57  116
gen_cP0.5   102  999   41   86  152
resistPointsMix3[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1   168  999   21  129  191
gen_cP0.25  199  999   38  159  202
gen_cP0.5   213  999   50  172  207
 
 
Now can I do the 20% better results ? 
 betterMix1Seq20_cP0.1 betterMix1Seq20_cP0.25  betterMix1Seq20_cP0.5 
                 0.136                  0.175                  0.173 
 betterMix2Seq20_cP0.1 betterMix2Seq20_cP0.25  betterMix2Seq20_cP0.5 
                 0.241                  0.326                  0.313 
 betterMix3Seq20_cP0.1 betterMix3Seq20_cP0.25  betterMix3Seq20_cP0.5 
                 0.672                  0.560                  0.488  
 
rounded
 betterMix1Seq20_cP0.1 betterMix1Seq20_cP0.25  betterMix1Seq20_cP0.5 
                 0.14                  0.18                  0.17 
 betterMix2Seq20_cP0.1 betterMix2Seq20_cP0.25  betterMix2Seq20_cP0.5 
                 0.24                  0.33                  0.31 
 betterMix3Seq20_cP0.1 betterMix3Seq20_cP0.25  betterMix3Seq20_cP0.5 
                 0.67                  0.56                  0.49  
 
 
Low phi values lead to sequential being better.
e.g. https://andysouth.shinyapps.io/shinyMixSeqCompare1 
runs 
3: phi 0.53, 0.79
19: phi 0.65, 0.44
25: phi 0.60, 0.56
 
** 15/7/15 ** 
 
installed triangle package
this to generate random deviates
rtriangle(n, a=0, b=1, c=(a+b)/2)

changed label 'combination' to 'mixture' in my generic curtis fig2 plot

** 16/7/15 **
reran 1000 sensitivity scenarios with new agreed input ranges
& reran 100 scenarios and put them into shinyMixSeqCompare1

check that the example runs still support the args the way they did before

the 3 example plots that I pasted into the word doc seem to be identical to what they were before
ARSE is it possible I didn't reload the package so that it used the old function version with the old input ranges ?

yes I think that's what I did !

rereran 100 runs and put them into shinyMixSeqCompare1

sorted order of insecticides for sequential figures

run shinyMixSeqCompare1 in a window rather than the browser to enable copy and pasting figures into word doc

copied 3 figs into the word doc. Note that run 9 is interesting. Mixture does much better than sequential.


 
### TO DO  


1. set 1000 runs to do tonight

1. redo the word results doc



1. fix the testthat tests to avoid new input addition




Remember.
It is a 2 locus model. One locus for each insecticide. At each locus there are 2 alleles. This gives a total of 9 genotypes if the position of the alleles is ignored. 

Main question is : Tree1 is mixture better (0/1), Tree2 is mixture better by >20%.

Good simple test
```
input <- setInputOneScenario(max_gen=5)
tst2 <- runModel2(input)
tst <- runModel(input)
```

1. submit stackoverflow question about createArray2(), including about n=3 in eval.parent
1. what to call createArray2() ? maybe namedArray, or arrayNamed() ? could even put back in the option to pass the fill value ? submit to stackoverflow ?

1. is there a way I can easily run single & combined deployments for the same scenario ?
* is it `niche_00` and `niche_AB`
* or in `a.m_AB` insecticide exposure male hi1 hi2
* do I also need to change "Starting frequency of resistance" ?
* perhaps I can set up a simple example input to send to Beth ?
* or ask Beth to send me the input.csv used to produce tables 8 & 9 in 'Brief_full-genotype-tables.pdf'
* maybe comparing the files in full-dom_CD.csv & SD in C:\Dropbox\Levick et al IR manuscript will help me.



1. in susana's script the 'outcome' is 0/1 if used in our case that would seem to lose a lot of information about the outcomes ? Even timeToFifty would be better for us.

1. maybe add option in setInputSensiScenarios to save input object as a csv or rda.

1. add some defensive checks, e.g. for sensible input variables

1. create a vignette from first shiny app documenting how the model works

1. finish tidying documentation of reading in input vars in createInputMatrix()

1. the curtis fig2 plot is produced from >1 scenario, do I want to get it into 1 'reactive' scenario
1. submit tilde bug report

Want to do:
1. rename a to exposure
1. rename W to ?fitness : fitnessLocus, fitnessNiche and fitnessIndiv



### refactoring variables to arrays ?
Refactoring to arrays will reduce code volume, reduce risk of mistyped variable names and allow objects to be passed between functions.
cis/trans makes it slightly tricky, but not all variables have cis/trans
only used for genotype frequencies (f) and in output matrices fitness & genotype 
in the wild in Beths code :
```
fs.m.RS1RS2_cis <- f.m.RS1RS2_cis
fs.m.RS1RS2_trans <- f.m.RS1RS2_trans
```


### variable naming conventions
**SS1, RS1, RR1** homozygous susceptible, heterozygous or homozygous resistant genotypes at each locus.   
With 2 loci there are ten possible genotypes (including cis & trans forms of the double heterozygous).  

** A, B, a, b, -** 2 insecticides at 3 levels; low(ab), high(AB), absent(--) resulting in 9 niches.  

Locus1 relates to resistance for insecticide A, and locus2 for B. Locus2 does not effect exposure to A.  

```
colnames(fitness) <- c( "-,-", "a,-", "A,-", "b,-", "B,-", "a,b", "A,B", "A,b", "a,B" )
rownames(fitness) <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
                        "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                        "RR1SS2", "RR1RS2", "RR1RR2")
```
#### inputs
** a ** exposure to insecticide
** phi ** fitness of one locus (baseline)
** W.SS1_00 & W.SS2_00** fitness of SS in no insecticide
** h ** dominance coefficient
** s ** selection coefficient
** z ** fitness cost of resistance allele in no insecticide
** niche_ ** insecticide niche toggle

#### calculated
** W ** fitnesses by niche and sex
** f ** genotype frequency
** fs ** genotype frequency following selection
** G ** gametes

Different W fitness variables
`W.RR1_00 : [locus, niche1, niche2]` single locus (Wloci)
`W.RR1SS2_0b :   [locus1, locus2, niche1, niche2]` niche (Wniche)
`W.m.SS1SS2 : [sex, locus1, locus2]` (Windiv)

### refactoring, what I have done
Replaced the following collections of individual variables with arrays :
(To reduce code volume, reduce risk of mistyped variable names and allow objects to be passed between functions. Also potentially allows it to cope with more loci and/or niches later).

```
a.m_a0 ... *18 2*3*3  to: a[sex, niche1, niche2]
W.m.SS1SS2 ... *18 2*3*3 to:  Windiv[sex, locus1, locus2]
W.RR1SS2_0b ... *81 3*3*3*3 to: Wniche[locus1, locus2, niche1, niche2] 
W.RR1_00 etc. with Wloci[loci, niche1, niche2]

because of cis/trans:
f.m.SS1SS2 ... *19 2*3*3+1 to: f[sex, genotype]
```

### outputs ###
From runModel2()
A list of 3 lists of one or more scenarios: 
1. results : by generation freq of R allele at each loci in each sex plus linkage stuff
    columns include Gen, m.R1, m.R2, f.R1, f.R2
2. genotype : frequencies of each of the ten genotypes, per generation
3. fitness : fitness scores of each genotype/niche combination (table 4. of Main Document)
4. input
e.g. 
listOut$results[1] gives a results matrix for the first scenario
listOutMix$input gives the inputs for all scenarios
listOutMix$input[c('P_1','P_2'),] gives selected inputs

# current best UIs
https://andysouth.shinyapps.io/shinyMixSeqCompare1
https://andysouth.shinyapps.io/shinyFig2Curtis




