---
title: "resistanceModelNotesAndy"
output: html_document
---

Notes on insecticide resistance model for liverpool.
andy south 27/1/15

Timelines and deliverables
**Days 1 to 12:** 
Read and assimilate the Curtis paper cited above. Run the LSTM code written by Levick, replicate the results from the Curtis paper, and prepare the code for sensitivity analysis by firstly refactoring into an R package and creating a Github repository. Identify appropriate parameter ranges and any rules that should be applied while sampling (e.g. male exposure to insecticide within homes should not be higher than for the more endophilic females).  Run sensitivity analysis and present the results as classification trees showing what parameter combinations favour which deployment strategy.  

**Deliverables:**  the provision of sensitivity analyses and classification trees that will allow us to publish the model in a peer-reviewed journal.  

**Days 13 to 25.**  This component is to use the full model rather than the sub-model used to replicate the Curtis results. Liaise with IVCC to identify the policy question they consider the most pressing and the likely end users of the simulation package. Perform sensitivity analysis on the full model as above. Attempt to put a user-friendly interface onto the simulation package using the “Shiny R” package.  

**Deliverables:** sensitivity analysis of the complete model and a reasoned consideration (or delivery) of the feasibility of distributing the package with a graphical interface for use by users with no computing experience.  



**27/1/15**  7.5hrs

```
#working version of Beths code
#run it outside of this project because it changes wd
source("C:\\Dropbox\\Ian and Andy\\andy\\malaria\\Beth code\\malaria_code_beth.r")
```

Editing `malaria_code_andy.r`, indenting functions etc.

Much of the script from line 464 is setting up parameters for example runs.

input[a,b] : a=parameter number and b=scenario number

464         set input params
1058-2716   run model
2718        Actions needing full results.list (e.g. curtis plots use multiple scenarios) 

edited bookmarks to navigate around the file in RStudio.

**9/3/15** 4-5 1hr


**10/3/15** 4.15 - 7.15 3hrs

```{r}
listFunctions <- function(filename) {
  temp.env <- new.env()
  sys.source(filename, envir = temp.env)
  functions <- lsf.str(envir=temp.env)
  rm(temp.env)
  return(functions)
}

#listFunctions("C:\\Dropbox\\Ian and Andy\\andy\\malaria\\Beth code\\malaria_code_beth.r")
```

So there are only 11 functions.

allele.freq : function (mat)  
curtis_f1 : function (nrelaxmat, relaxmat, gencol, r1col)  
curtis_f2 : function (combmat, bmat, amat, gencol, r1col, r2col)  
curtis_ld : function (resultsmat, relaxedmat, gencol, ldcol)  
haplotype : function (mat)  
HW : function (P, mat)  
linkage : function (mat)  
make.genotypemat : function (P_1, P_2)  
make.matrix : function (mat, rnames)  
singlealleleFrequency : function (locus, max_gen, results.list, input)  
timetoFifty : function (locus, max_gen, results.list, input


I can create an initial github repo for the files from Beth (&this notes file)
That means I will have version control for my initial restructuring. Later I may want to create a new repo for the reformatted code as a package.

What to call the first repo ? resistance

1. created repo on github
1. created RStudio project from github repo
+ To get ssh push working.
+ RStudio Tools, Shell
+ git remote set-url origin git@github.com:AndySouth/resistance.git

**11/3/15** 9.30-12.30 3hrs 2.15-6.45 7.5hrs

moved my dropbox folder back to C.

Plan
1. move Beths functions out into a single file for each in an R folder
1. edit Beths commments into roxygen format
1. only later think about changing functions and documentation to follow best practice

Changed plan slightly by appending plot onto start of plot functions.

To source files in the R folder
`lapply(dir("R"),function(x) source(paste0("R//",x)))`

**12/3/15** 9.30-14 4.5hrs 3-5 6-7 7.5hrs

The model generates 3 matrices : results, genotype, fitness
1. Results : freq of R allele at each loci in each sex and linkage disequilibrium of R allele in each sex, per generation
2. Genotype : frequencies of each of the ten genotypes, per generation
3. Fitness : fitness scores of each genotype/niche combination (table 4. of Main Document)

When a file is input with multiple scenarios, each of the three matrices is stored in a list, where scenario number gives position of the matrix in the list.

results.list : 
fitness.list : 
genotype.list : 

Can I put these into a single list so that they can be returned from a single function ?

done ~ put input object creation into a function
But there is still a fair bit of code that reads parameter values out of the input matrix and into named variables.

I might be able to restructure these collections of single variables into arrays.
e.g. for the exposure levels of m&f

```{r}
#to create an array a[sex][locus1][locus2]
sex <- c("F","M")
locus1 <- c("0","A","a")
locus2 <- c("0","B","b")
dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','a','b']
#to check that male exposures total 1
sum(a['M',,])
```


```
  ## Exposure levels of males and females to each insecticide niche ##
  # males
  a.m_00 <- input[8,i]
  
  a.m_a0 <- input[9,i]
  a.m_A0 <- input[10,i]
  
  a.m_0b <- input[11,i]
  a.m_0B <- input[12,i]
  
  a.m_ab <- input[13,i]
  a.m_AB <- input[14,i]
  
  a.m_Ab <- input[15,i]
  a.m_aB <- input[16,i]
  
  #a.m <- sum(a.m_00, a.m_a0, a.m_A0, a.m_0b, a.m_0B, a.m_ab, a.m_AB, a.m_Ab, a.m_aB)
  #if ( a.m != 1 ){		 
  #	print( paste("Error in male exposures: must total one: ", a.m) )
  	
  #	}else{
  #		print( paste( "Male exposures total 1: ", a.m ))
  #		}
  		
  
  # females
  a.f_00 <- input[17,i]
  
  a.f_a0 <- input[18,i]
  a.f_A0 <- input[19,i]
  
  a.f_0b <- input[20,i]
  a.f_0B <- input[21,i]
  
  a.f_ab <- input[22,i]
  a.f_AB <- input[23,i]
  
  a.f_Ab <- input[24,i]
  a.f_aB <- input[25,i]
```

I should check that what we are doing fits in with what others are doing. We may be able to capitalise on existing tools.

[popgen on CRAN](http://cran.r-project.org/web/packages/popgen/popgen.pdf)
seems a bit old and doesn't do a huge amount, no vignette.

[gstudio](http://cran.r-project.org/web/packages/gstudio/) looks more promising. e.g. it defines locus objects.
[gstudio documentation](http://dyerlab.github.io/gstudio/)
A locus object is set like this, but I'm not sure if this would help us.
It does have some stuff on drift at end of help.
```
require(gstudio)
loc <- locus(c("C", "A"))
loc
```

pegas provides functions for the analysis of allelic data and of haplotype data from DNA sequences.
It requires and complements two other R-packages: ape and adegenet.
[pegas](http://ape-package.ird.fr/pegas.html)
[pegas data structures](http://ape-package.ird.fr/pegas/DefinitionDataClassesPegas.pdf)
includes a class called loci. : An object of class "loci" is a data frame where rows represent individuals
and columns are loci and optional additional variables.
Again I can't quite see how it would be useful for us.

There is a [HardyWeinberg](http://cran.r-project.org/web/packages/HardyWeinberg/vignettes/HardyWeinberg.pdf) package on CRAN. "The HardyWeinberg package consists of a set of tools for analyzing diallelic
genetic markers, and is particularly focused on the graphical representation of their (dis)equilibrium condition in various ways."

There is a CRAN [genetics task view](http://cran.r-project.org/web/views/Genetics.html).

The [genetics package](http://cran.r-project.org/web/packages/genetics/genetics.pdf) which also has a locus class, seems to be more detailed, about chromosomes etc. and probably not of use to us.

#### Structure of the results matrices

```{r}
max_gen <- 2 #just as example

# Set up results matrix - prints overall freq of R and S allele per locus per sex, LD and overall allele freq (i.e. 1)
results <- matrix ( nrow = max_gen, ncol = 11 )
colnames( results ) <- c( "Gen", "m.R1", "m.R2", "m.LD","f.R1", "f.R2", "f.LD", "M", "F", "dprime", "r2" )

# set up fitness by niche matrix - records fitness scores for each niche for each genotype
fitness <- matrix ( nrow = 10, ncol = 9, c(rep(0,90)))
colnames(fitness) <- c( "-,-", "a,-", "A,-", "b,-", "B,-", "a,b", "A,B", "A,b", "a,B" )
rownames(fitness) <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
						"RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
						"RR1SS2", "RR1RS2", "RR1RR2")

# set up genotype matrix - records frequencies of each of the 9 two locus genotypes each generation
genotype <- matrix( nrow=max_gen, ncol=11 )
colnames(genotype) <- c("gen", "SS1SS2", "SS2RS2", "SS1RR2", 
						"RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
						"RR1SS2", "RR1RS2", "RR1RR2")


```

Somehow it seems I've broken it, I'm getting NAs in results matrix.

The input files are different (but same structure) between Beths version that works and my new version ...
calibration is 100 in my failing version and 1012 in Beths, i think this is because I'm reading in the csv when I shouldn't.

I seem not to get to line 889 in my version
  	# male
  	f.m.SS1SS2 <- genotype.freq[1,]

Then suddenly it seemed to start working when I put a browser in.
It's something tricky around line 711, if it doesn't get to 889 it doesn't fill in the results matrix ??

Aha! It seems to have been this without the comments before the tilde's that stoppped it from working
```
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 		
  ### Loop to run the model from the initial conditions generated above ####
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

**13/3/15** 9.15-12.15 3hrs

A reproducible version of yesterdays bug.

```
x<-1
~
for(i in 1:10) x<-x+1
cat(x)
```
[R bug reporting](https://bugs.r-project.org/bugzilla3/enter_bug.cgi) 

1. can I create a list of the 3 lists for results ?
Yes
Replace this  
results.list <- list()			
fitness.list <- list()
genotype.list <- list()	
with  
listOut <- list( results=list(), fitness=list(), genotype=list() )

Then find & replace  
results.list with listOut$results  
fitness.list with listOut$fitness
genotype.list with listOut$genotype


**16/3/15** mon 18.30-19.30 1hr
cut out lines 92 to 1740 into a runModel() function, later can subdivide it up

**17/3/15** tues to liverpool 8.30-5.30 9hrs

Assuming that resistance to each insecticide was encoded at one locus respectively, and that these can either have a homozygous susceptible (SS), heterozygous (RS) or homozygous resistant (RR) genotypes at each locus, there are then ten possible genotypes (including cis & trans forms of the double heterozygous).

We consider two insecticides (A and B) at three possible levels; absent, low(ab) and high(AB), resulting in 9 niches;

Locus1 relates to resistance for insecticide A, and locus2 for B. Locus2 does not effect fitness under exposure to A.

Males and females can be exposed differently.

Curtis states that if resistance to one insecticide is present at very low levels in the population, then resistance will rise slower if this insecticide is used on its own, rather than in a combination with a more established insecticide. 

This is the important Curtis statement “the use of a mixture where the initial gene frequencies are unequal leads to more rapid increase in the frequency of the rarer of the genes”

A key aspect of Curtis’ prediction was that this relationship would be determined by an increasing association between the presence of the two resistance alleles, measured as high levels of linkage disequilibrium.

The results given above seem to be in contrast with Curtis’ assumption, and suggest that in fact using the insecticides in combination can slow the spread of resistance.

My summary :
Curtis suggests resistance spreads more quickly with 2 insecticides, due to  increasing association between the alleles. Beths work suggests that combination can slow spread of resistance. 

Good figure 1 in manuscript :
The rate of spread of IR depends on whether the resistance mutation is dominant, semi-dominant, or recessive.

**aha! I've only just understood this bit**
**At high concentrations only the RR individuals survive so resistance is recessive; as concentrations decline some RS mosquitoes survive making resistance semi-dominant; at low concentration both RR and RS mosquitoes survive, making resistance dominant.** 
**so main problems of resistance arise when insecticide concentrations decline**

From manuscript about scenarios to run:
**row in params csv**

Insecticide coverage for females: 0.1, 0.2, 0.6, 0.8, 0.9 (n=5) **9 & 15**  
Insecticide coverage for males: as for females, or half that of females (n=2) **18 & 24**  
‘New’ insecticide resistance starting freq: 0.0001, 0.001, 0.01 (n=3) **6**  
‘Old’ insecticide resistance starting freq: 0.05, 0.1, 0.2, 0.5, 0.8 (n=5) **7**  
Resistance heterozygous fitness: 0.5, 0.75. 1    (n=3 for each insecticide) **?28 & 30**  
Sensitive heterozygous fitness:0, 0.2, 0.4 (n=3 for each insecticide) **?27 & 29**  
Resistance allele dominance: 0, 0.2, 0.5, 0.9, 1 for each locus (n=5 for each locus) **32, 35 do to start**  
Fitness costs: 0, 0.05, 0.1 (n=3) **43&44**  
Fitness dominance: 0, 0.5, 1 (n=3) **33&36**  

BUT now we are going to do it by sampling from a uniform distribution before doing decision trees. Then you can progressively add results to a file.

I could start by creating a document with the results at the extremes of the parameter space.

Send Ian & Beth devtools installation instructions and a document.

Approx. 60,000 combinations. Too many???  

would latin cube hypersampling be better?  

The simulation output was the ratio of  the time (in generation) for resistance  to the ‘new’ insecticide  to reach a given allele threshold frequency when the ‘new’ insecticide was deployed as combination, compared to the time taken to reach the same threshold if deployed on its own; a ratio >1 indicates that combination is better. Three threshold frequencies were investigated: 0.1, 0.25, 0.5.

These ratios was then analysed by regression to find out how the input parameters alter the ratio, and hence under what circumstance combination would be best/worst. Classification tress can also be done.

The intention was not to consider particular plausible deployment situations, but was to systematically investigate a wide range of parameter space to determine whether CD was better than single-insecticide deployment at reducing the rate at which the minor resistance form spreads.
 
Show my progress :

1. github repo
1. move functions to their own files and structure documentation
1. move parts of main file into functions
1. combine 3 outputs to a single list so that can be returned from a function
1. nearly able to create a package with structured documentation etc.

Questions for Beth & Ian

1. Curtis Table1 vi that shows resistance spreading faster with mixture.
1. is coverage same as exposure ? yes
1. what subset of parameters would they like me to vary to start
1. locations of sensitivity params in params.csv
1. what outputs are wanted for sensitivity scenarios ? num generations until resistance reaches threshold (0.1, 0.25 & 0.5) Curtis just used 0.5.
 
Talking with Ian & Beth

Sequential use.
 
For some scenarios, we'll need to run with one insecticide and then stop and start other. 
 
When does mixture fail, until one resistance allele reaches critical point, then continue using 2nd insecticide until resistance for that one reaches critical point too. 
 
Could create a version of timetoFifty that accepts threshold param.
 
Ian has created list of defaults in most recent manuscript.
 
The Curtis version is going to be simpler. Because not all param variations are included.

**18/3/15** weds in liverpool 8.30-6 9.5hrs

Ian created `suggested_sensitivity_analysis.doc`
 
I created a variables table in there with Ians parameter ranges.

My poor paraphrasing of what I think Ian said about the sequential scenarios. The starting conditions from after one insecticide has 'failed' are the same as starting from scratch for the other insecticide because the insecticides to do effect selection on the other locus.
 
moving `malaria_code_andy.r` into a function resistanceMaster()
exposed a couple of variable scoping problems.

plus now the graphs look different and I get :
Warning message: In data.matrix(mat) : NAs introduced by coercion

this is from createInputMatrix()
input <- make.matrix(input, input$Input)
 
this might be something to do with the location of "input.parameters.csv" which I need to sort ...
 
devtools commands for setting up a package :
create requires that the directory doesn't exist yet; it will be created. 
setup assumes an existing directory from which it will infer the package name.
 
probably need to set rstudio rg to false because mine is already a rs project.

setup(rstudio=FALSE)
 
Then in RStudio I had to set project options, build tools to package
(so maybe I should have done just done devtools::setup())

but on build :
ERROR: The build directory does not contain a DESCRIPTION
file so cannot be built as a package.
Build directory: C:/rsprojects/resistance

but there is a DESCRIPTION there ???

devtools::load_all(".")
Error: Line starting 'person("Bethany", "L ...' is malformed!
sorted, just needed tab before 2ry authors
 
package does build now
 
 
potential refactoring of code, if i think carefully, using arrays I should be able to replace this :
```
    if(calibration==103){		## no selection calibration
        ## male
        # SS1
        fs.m.SS1SS2 <- f.m.SS1SS2
        fs.m.SS1RS2 <- f.m.SS1RS2
        fs.m.SS1RR2 <- f.m.SS1RR2
        # RS1
        fs.m.RS1SS2 <- f.m.RS1SS2 
        fs.m.RS1RS2_cis <- f.m.RS1RS2_cis
        fs.m.RS1RS2_trans <- f.m.RS1RS2_trans
        fs.m.RS1RR2 <- f.m.RS1RR2
        # RR2 
        fs.m.RR1SS2 <- f.m.RR1SS2
        fs.m.RR1RS2 <- f.m.RR1RS2
        fs.m.RR1RR2 <- f.m.RR1RR2
        
        ## female
        # SS1
        fs.f.SS1SS2 <- f.f.SS1SS2
        fs.f.SS1RS2 <- f.f.SS1RS2
        fs.f.SS1RR2 <- f.f.SS1RR2
        # RS1
        fs.f.RS1SS2 <- f.f.RS1SS2 
        fs.f.RS1RS2_cis <- f.f.RS1RS2_cis
        fs.f.RS1RS2_trans <- f.f.RS1RS2_trans
        fs.f.RS1RR2 <- f.f.RS1RR2
        # RR2 
        fs.f.RR1SS2 <- f.f.RR1SS2
        fs.f.RR1RS2 <- f.f.RR1RS2
        fs.f.RR1RR2 <- f.f.RR1RR2
```

with something like :
`fs <- f`

where arrays are set up as :
```{r}
#to create an array a[sex][locus1][locus2]
namesLoci <- c('SS','RS','RR')
locus1 <- paste0(namesLoci,'1')
locus2 <- paste0(namesLoci,'2')
sex <- c("F","M")

dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','SS1','RR2']
#to check that males total 1
sum(a['M',,])
```

Just need to work out how best to deal with cis & trans.
This would be a less pleasing alternative
```{r}

namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')

#adding cis & trans
namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2cis','RS1RS2trans','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')

sex <- c("F","M")

dimnames1 <- list( sex=sex, namesLoci=namesLoci )

dim1 <- sapply(dimnames1, function(x) length(x))
a <- array(0,dim=dim1, dimnames=dimnames1)
a
length(a)
#to access one element
a['M','SS1SS2']
#to check that males total 1
sum(a['M',])
str(a)
```

**19/3/15** fri 9.45-14 4.25hrs 3.30-7 3.5hrs 7.75hrs
 
checking whether I can run without the input file. Fixed now.

```{r}
#require(devtools)    
#install_github('AndySouth/resistance') 
library(resistance)
resistanceMaster(params.csv = FALSE)
``` 

This is what the fitness scores that can be saved for each scenario as *two-locus_fitness-scores.csv look like :
```
	-,-	a,-	A,-	-,b	-,B	a,b	A,B	A,b	a,B
SS1SS2	1	0	0	0	0	0	0	0	0
SS1RS2	1	0	0	0	0	0	1.86E-05	0	0
SS1RR2	1	0	0	0	0	0	0.1161	0	0
RS1SS2	1	0	0	0	0	0	0	0	0
RS1RS2	1	0	0	0	0	0	2.13E-05	0	0
RS1RR2	1	0	0	0	0	0	0.132913	0	0
RR1SS2	1	0	0	0	0	0	0	0	0
RR1RS2	1	0	0	0	0	0	3.44E-05	0	0
RR1RR2	1	0	0	0	0	0	0.215	0	0
```

Starting resistSimple() to be able to run single simple Scenarios.


**20/3/15** sat 10.15-2.15 4hrs

[recent popgen hack](https://github.com/NESCent/r-popgen-hackathon/wiki)
Doesn't seem that useful for this project.

Created setInputOneScenario to enable setting inputs for a single scenario
+ sensible defaults
+ later will allow customised scenarios e.g. setInputCurtisFig1
+ will reduce code volume and repetition in createInputMatrix

Good progress :
check out e.g.
`resistSimple(P_1=0.01,P_2=0.3)`


**23/3/15** mon 9.45-11.45 2hrs 5-8 5hrs 

created a modified plot allele freq function to be able to see overlapping lines

Default settings :
```{r}
resistSimple()
```

Reducing dominance coefficient of locus1 from 1 shifts curves to right & separates locus 1 & 2
```{r}
resistSimple(h.RS1_A0=0.17)
```

But then setting dominance coefficient of locus2 to anything less than 0.7 grounds the curves (actually by pushing them further to right).
```{r}
resistSimple(h.RS1_A0=0.17,h.RS2_0B=0.7)
```

Reducing the selection coefficient for locus1 in A (from 1) pushes curves to right, but strangely doesn't separate locus 1 & 2 ??
```{r}
resistSimple(s.RR1_A0=0.17)
```

Put variable names into the suggested sensitivity analysis table.
 
Thinking about creating a shiny app(s) within the resistance package.
[Advice from RStudio](http://rstudio.github.io/shiny/tutorial/#deployment-local)

Put your Shiny application directory under the package’s inst directory, then create and export a function that contains something like this:

`shiny::runApp(system.file('appdir', package='packagename'))`

I might be able to use something like [this](http://shiny.rstudio.com/gallery/plot-plus-three-columns.html) to put inputs in columns.

Or perhaps better to keep it to a single column to start ?

Starting frequency of resistance
Locus1
Locus2

Exposure to insecticide (same for M&F & in Curtis for both insecticides)

Fitness of susceptibles in presence of insecticide.
Insecticide1
Insecticide2

Dominance of resistance
Locus1
Locus2

Selective advantage of resistance
Locus1
Locus2

It could perhaps be 2 columns, Locus1 & Locus2

Or to fit better with widget titles, 2 rows (l1&2) with 5 columns.

**24/3/15** tue 9.45-2.15 4.5hrs

created function to run shiny UI from package, and add it to readme

there was a bug that dominance had no effect, because I was setting `h.RS1_00` rather than `h.RS1_A0 & h.RS1_0B` 

Cool! Seems to be working well now.

Sent message to Ian & Beth to test installing from Github and the shinyUI.

**25/3/15** weds 12.30-13.45 1.25 2.30-4 1.5 2.75

created a time record  

Barbosa paper methods: 
The analysis was performed using R & package lhs. It does not
allow for the specification of each variable distribution
beforehand, so sampling was performed assuming a uniform
distribution. Once the sample was generated, the
uniform sample from a column (variable) could be transformed
to the required distribution (Table 3) by using
quantile functions (using the qtriangle comand in R).

A data set of 3,000 replications was generated, with random
parameter sets. Ten replicates of this procedure were performed as suggested
in [24] to investigate the predictive precision of model
using LHS as the sampling method. This was achieved
by analysing each replicate separately and verifying that
results were consistent across ten replicates.

...

Results included a counter-intuitive outcome that the
inclusion of a synergist could lead to an increase in the
rate of the spread of resistance (i.e. y > 1). Further investigation of this result was pursed by performing a logistic regression with a binary dependent variable (1 if y > 1 and 0 if y < 1), therefore quantifying how changes in the
parameters values affect the odds of getting the unexpected
outcome y > 1.

...

Classification trees are used to predict membership of
cases in the classes of a categorical dependent variable
(1 if y > 1 or 0 if y < 1) from their input parameters
and were implemented using an algorithm that grows a
binary tree [27]. At each internal node in the tree, a test
is applied to the input parameters to identify the binary
distinction which gives the most information about the
class membership.

These are the references that refer to the classification trees :
27. Breiman L: Classification and Regression Trees. Belmont (Calif.): Wadsworth International Group; 1984: 358.
28. Therneau T, Atkinson EJ: An introduction to recursive partitioning using the RPART routines 2011:1–67. [r.789695.n4.nabble.com/attachment/3209029/0/zed.pdf].

seems that [recursive partitioning using rpart](http://r.789695.n4.nabble.com/attachment/3209029/0/zed.pdf) is the R software used.

[rpart package](http://cran.r-project.org/web/packages/rpart/index.html)
Recursive partitioning for classification, regression and survival trees. An implementation of most of the functionality of the 1984 book by Breiman et al.

[rpart intro vignette](http://cran.r-project.org/web/packages/rpart/vignettes/longintro.pdf) updated version of the reference.


First example from the vignette (p10). This works on a binary response yes/no. method = 'class' is for categorical.
```{r}
library(rpart)
#uses stagec, a dataframe
str(stagec)
#creates a factor for response var to improve labelling
progstat <- factor(stagec$pgstat, levels = 0:1, labels = c("No", "Prog"))
#fit the model by adding the columns
cfit <- rpart(progstat ~ age + eet + g2 + grade + gleason + ploidy, data = stagec, method = 'class')
plot(cfit) #plots tree
text(cfit) #labels tree
```

So that seems to suggest that I'll need to output a dataframe with one row per simulation, columns for outputs (generations until resistance reached) and inputs. Will need to run it by Ian exactly what the outputs are.

p25 in vignette has an example with multiple categories.
Seems there are methods that work on continuous data too, which might be our case ...

**30/3/15** mon 10.30 - 11.30 1hr

RSTMH digital methods conference.
KDR insecticide resistance gene.
Turning 'big' data about how resistance spreads into epidemiological models and then into DSS about control and eradication is very complex. 
Malaria genomics: tracking a diverse and evolving parasite population
Professor Dominic Kwiatkowski, University of Oxford and Wellcome Trust Sanger Institute
Paul Ben Allistere, Panoptes web app
https://www.malariagen.net/apps/ag1000g/phase1-AR2/index.html#genomebrowser

**2/4/15** 1hr

emails from Ian & Marlize Coleman about insecticide resistance gaming.

Ian : Marlize Coleman has funding from B&MGF to build computer games of insecticide deployment policies and wants to tie them to the outputs of various IR models.

Marlize : At the moment **we need someone to look at current mathematical models that have any relevance to insecticide resistance management and determine if there is any potential to use them, or just components of the model/s, to create scenarios for a IR game**. We planning to start development of the game in July so we need to get a game scope sorted pretty soon. It will also be important for this person to communicate and discuss these scenarios/simulations with the game development team to ensure that they have a good grasp of the backend mathematical detail.

We want to expand on this version by adding a bit more complexity in the backend and we’d like to create a bit more flexibility e.g. choice of vectors, disease prevalence/incidence, interventions, specific learning outcomes of interest etc.

If you are interested, we can discuss more detail. Probably 3 months full time or **6 months 2-3 days a week**. You might be better placed to tell me how much effort will be required. Perhaps more time in the first few months and less during game development for assistance to developers.

Me to Ian : Will be good to talk to you about potential IR models after I've spoken to Marlize. Do you imagine that a simplified version of Curtis might be useful for her ?

Ian : I definitely think the 2locus Curtis model should be included. I don't think it necessarily needs to be simple. We may restrict the user to certain parameter values then preload the results from all possible permutations. We can discuss later. BW Ian


**13/4/15** mon 9.30 - 13.30 4hrs 2.30-6 3.5hrs 7.5hrs

looking at new serious gaming project ... & can offset these hours on that

publish shinyCurtis1 to shinyapps.io to enable sharing with Marlize
using publish button from RStudio worked well
It gives this warning on installation, but then does work from web.
Warning message:
In FUN(c("shiny", "resistance")[[2L]], ...) :
  Package 'resistance' not available in repository or locally

https://andysouth.shinyapps.io/shinyCurtis1/

Options :
1. run R code from within the game
1. translate model modules into the game language
1. save model outputs as input files that can modify game behaviour

Points to discuss with Marlize :

* I am mostly committed until end of June, but might be able to shuffle. If we know what they want from the design, we could work on the model behind to produce this after the game design has been started.
* might they want to run actual model code or translate it ?
* I don't know how much info there is out there on the effect on resistance of different insecticide applications, but am happy to research in discussion with Ian, and produce working versions.
* I would be a good intermediary between the biologists and the coders, as I'm somewhere in between myself !

Imperial malaria models :
http://www1.imperial.ac.uk/resources/27F31FE6-4CA6-4C97-85BD-58B342BC2CDA/usingthemalariatoolssoftwareforesp.pdf
http://www1.imperial.ac.uk/malariamodelling/toolsdata/tools/
On download I got an error msg saying : "The program can't start because VCOMP120.DLL is missing from your computer. Try reinstalling the program to fix" 

http://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1000324
Reducing Plasmodium falciparum Malaria Transmission in Africa: A Model-Based Evaluation of Intervention Strategies

emailed them to see about the error


**14/4/15** tue 11.45 - 14.15 2.5hrs 3-5.30 2.5 5hrs  

can I create a plotting method that can accept multiple scenarios ?
initially it could just plot all scenarios as additional lines in the existing colours for locus & sex
use `plotallele.freq.andy`
might need to modify to accept more than one matrix of results outputs  
I thought tricky bit might be to set plot bounds based on multiple lines (but actually the y is already hardcoded from 0-1 for allele frequency). Might need to be a bit careful with x because different scenarios can have different numbers of generations.

done : fixed why `resistanceMaster()` currently fails. I suspect due to input scenario calibration combination, it used to work ??

Error in 1:ddt_cutoff : result would be too long a vector
In addition: Warning messages:
1: In data.matrix(mat) : NAs introduced by coercion
2: In min(which((amat[, r1col]) > 0.5)) :
  no non-missing arguments to min; returning Inf

the error comes from `plotcurtis_f2()`
I suspect it's because the input.parameters.csv isn't the one compatible with producing the curtis plot.
`plotcurtis_f2()` needs results matrices from 3 scenarios: 1 hch, 2 ddt, 3 combined. Yet the current csv I have in the package seems to have 7 scenarios
Input,Curtis Fig 2 Comb - calibration,DDT high,DDT high,DDT high,HCH high,HCH high,HCH high  
Why did it work before then ??
Aha! it's because in `createInputMatrix()` this sets up the 3 Curtis scenarios if an input file is not specified :
  `if( !params.csv && calibration == 1012 ){
    input <- matrix( ncol=3, nrow=52 )
    colnames(input) <- c("B (HCH)", "A (DDT)","Combination")`
So maybe I should have no input csv as the default.

setting params.csv=FALSE as the default for resistanceMaster() sorted it.

Back to `plotallele.freq.andy`, I might be able easily to modify it to accept a list of matrices for multiple scenarios, as well as it's current single scenario mode.

this gives me the number of generations per scenario
`sapply(listOut$results,nrow)`
70  70 160
`max_gen <- max( sapply(listOut$results,nrow) )`
 
done : create a plotting method that can accept multiple scenarios 
 
now try to use it to plot a parameter range.
To run multiple scenarios one way would be to call `setInputOneScenario()` repeatedly to create single scenarios & then rbind these together into an input object that I can pass to `runModel()`

```{r}

input <- NULL
vals <- seq(0,1,0.2) #create a range of input values
for(i in vals)
{
 inputOneScenario <- setInputOneScenario( h.RS1_A0=i )  
 input <- cbind(input, inputOneScenario)
}

listOut <- runModel(input)
plotallele.freq.andy(listOut)

```
This is getting close to allowing the user to specify an input parameter and what range of that parameter you want to plot. I could then potentially create a UI allowing the user to do that.
Would it be useful to put the above code into a function ? 
Actually the above code is very close to what I would need for the sensitivity analysis.

I could create a `setInputMultiScenario()` function that accepts a vector of values for a param, and creates an input object with all other params set to their default. It could maybe even accept multiple args, but then would need to return the n*n scenarios. Also this is going away from the way that Ian suggested doing it by smapling from a distribution, maybe I should try that route first ?

In that case do I want to randomly change all params at once ?
Also might be good to get it to output a file at the end of each 100 runs, so that if it does crash during a long run, everything is not lost. These could just be RDA files saved from the listOut object.
Do I want it to output listOut() each time or will I process so that the outputs are smaller ?
(Might be better to keep the raw outputs to save any confusion later.)
 
The sensitivity analysis function may be called a bit like this : 
 
```
#specify ranges for parameters, if a range is not specified the default will be used
#could offer option to setother params at diff fixed values
#this would only accomodate uniform distributions, OK for now
#think about random seeds later
#be careful about some params which must total one
input <- setInputSensiScenarios( n=100, h.RS1_A0=c(0,1) )
```
 
Am I reinventing the wheel here, can anything else help me with this ?
 
**15/4/15** wed 1.30-2.30 1hrs 3.45-6.15 2.5 3.5hrs

setInputSensiScenarios
function to create sensitivity analysis scenarios, probably by creating a large input object.

**16/4/15** thurs 10-14 4hrs 2.30-5 2.5 6.5hrs

To get input params into 1 row per scenario I can transpose, but the result is a matrix and has no names.
But I need to find a way of getting column names set.

```{r}
input <- setInputSensiScenarios( nScenarios=3, P_1=c(0,0.5), P_2=c(0,0.5) )
tinput <- t(input)
str(tinput)
#num [1:3, 1:52] 100 100 100 100 100 100 1 1 1 0 ...
```

putting names into `setInputOneScenario()` fixed this.

Having a problem with sensiAnalysis1()
Error in eval(..1) : ..1 used in an incorrect context, no ... to look in

This may be because setInputSensiScenarios() expects a range for each argument, whereas setInputOneScenario expects a single value for the same named args. (but I don't really think it is this)

Also stackoverflow pointed that using c for a varname (and there is one in resistance) is dangerous because R can expect the function.

replaced c with recomb_rate for "Recombination Rate" to avoid R conflicts with c() 
previous ... error still remained

Aha! i think problem is that I specify defaults in setInputSensiScenarios when I don't need to (and that stops them making it into the ...)

cool yes, getting there.

**22/4/15** jury service 11.30 - 12.30 1 hr

```
out <- sensiAn1( nScenarios = 10, h.RS1_A0=c(0.1,1))
```


**24/4/15** 4-5 1hr

**27/4/15** 9.15-13.15 4hrs 1.45-2.45 5hrs

trying to test rpart in sensiAn1
getting :
cfit <- rpart(res ~ h.RS1_A0 + h.RS2_0B, data = forCT, method = 'class')
Error in cbind(yval2, yprob, nodeprob) : 
  number of rows of matrices must match (see arg 2)

This was because there were some NAs in the results that can from where a resistance frequency of 0.5 was not reached.
so I temp converted them to 999

Now get :
Error in plot.rpart(cfit) : fit is not a tree, just a root

Is this because I haven't used enough samples or because it doesn't work with just 2 predictors ?

Trying with more samples, still get same error :

`inputAndResults <- sensiAn1( 100, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1) )`

more samples and more inputs I do get a tree
`inputAndResults <- sensiAn1(500, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))`


Some of susanas code for finessing the tree :
tree->rpart(outcome ~ h_n+ h_o+h_I+s_o+s_I+φ_o+φ_I+α_o+α_I+α_fn+α_mo+α_mI+α_mn+z+beta_f+β_m ,method='class',data=working_example,control=rpart.control(minsplit=50))

# control=rpart.control(minsplit=50)
# minsplit	the minimum number of observations that must exist in a node in order for a split to be attempted.
# then prune the tree to avoid overfitting the data
pruned_tree<- prune(tree, cp=tree$cptable[which.min(tree$cptable[,"xerror"]),"CP"])


**28/4/15** 10-14.30 4.5hrs 3-6 7.5hrs

reading Ians recent manuscript update.
I think this from p13 is key on what I need to do with the analysis : 

** The simulation output was the ratio of  the time (in generation) for resistance  to the ‘new’ insecticide  to reach a given allele threshold frequency when the ‘new’ insecticide was deployed as combination, compared to the time taken to reach the same threshold if deployed on its own; a ratio >1 indicates that combination is better. Five threshold frequencies were investigated: 0.05, 0.1, 0.2, 0.5, 0.8. **

In order to do that I'd need to repeat each scenario for a single & combined insecticide treatment. Single is just the new rarer insecticide, Combined also has an older insecticide to which initial resistance is more common.

Searching for the Curtis paper. Cited by 181 papers according to Google.
https://www.google.co.uk/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=%22theoretical%20models%20of%20the%20use%20of%20insecticide%20mixtures%22

https://books.google.co.uk/books?id=ooHjBwAAQBAJ&pg=PA177&lpg=PA177&dq=%22theoretical+models+of+the+use+of+insecticide+mixtures%22&source=bl&ots=b3CPTQc5LL&sig=FCB5MCvX1xUjtfgog1z_dq1pH1Y&hl=en&sa=X&ei=bXw_VbriFoWpsAGnvYG4Ag&ved=0CEEQ6AEwBA#v=onepage&q=Curtis&f=false

Pesticide Resistance in Arthropods (1990) edited by Richard Roush, Bruce E. Tabashnik

these 2 chapters look useful :
Roush, R. T., and J. C. Daly. 1990. The role of population genetics in resistance research
and management, pp. 97-152.In: Pesticide resistance in arthropods. R. T. Roush and
B. E. tabashnik (Eds.) Chapman and Hall, New York.

Tabashnik, B. E. 1990. Modeling and evaluation of resistance management tactics, pp.
153-182. In: Pesticide resistance in arthropods. R. T. Roush and B. E. Tabashnik
(Eds.) Chapman and Hall, New York.

page 131 talks about insecticide mixtures. Refers to some modelling by Roush (1989a)

Roush, R. T. 1989. Designing resistance management programs: How can you choose?
Pestic. Sci. 26:423-441.

p467 in Insect Resistance Management: Biology, Economics, and Prediction edited by David W. Onstad (2013)
seems to have some simulations addressing similar issues to us.

https://books.google.co.uk/books?id=6hp384ZH0_kC&pg=PA453&dq=%22Modeling+and+evaluation+of+resistance+management+tactics%22&source=gbs_toc_r&cad=4#v=onepage&q=%22Modeling%20and%20evaluation%20of%20resistance%20management%20tactics%22&f=false


Now have sorted passing user args to rpart in sensiAn1(). It does produce reasonable looking trees just by e.g.

`inputAndResults <- sensiAn1(100, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))`

**29/4/15** 10.30-12.30 2hrs 3.15-5.15 4hrs

e.g. W doesn't have cis/trans
I could refactor
```{r}
        # relaxed selection fitnesses
        ## Males
        W.m.SS1SS2 <- 0.1 
        W.m.SS1RS2 <- 0.1
        W.m.SS1RR2 <- 0.1
        
        W.m.RS1SS2 <- 0.1
        W.m.RS1RS2 <- 0.1  
        W.m.RS1RR2 <- 0.1  
        
        W.m.RR1SS2 <- 0.1 
        W.m.RR1RS2 <- 0.1 
        W.m.RR1RR2 <- 0.1  
        
        ## Female
        W.f.SS1SS2 <- 0.1
        W.f.SS1RS2 <- 0.1
        W.f.SS1RR2 <- 0.1
        
        W.f.RS1SS2 <- 0.1
        W.f.RS1RS2 <- 0.1 
        W.f.RS1RR2 <- 0.1 
        
        W.f.RR1SS2 <- 0.1
        W.f.RR1RS2 <- 0.1 
        W.f.RR1RR2 <- 0.1
# to :
#to create an array a[sex][locus1][locus2]
namesLoci <- c('SS','RS','RR')
locus1 <- paste0(namesLoci,'1')
locus2 <- paste0(namesLoci,'2')
sex <- c("f","m")

dimnames1 <- list( sex=sex, locus1=locus1, locus2=locus2 )
dim1 <- sapply(dimnames1, function(x) length(x))
W <- array(0,dim=dim1, dimnames=dimnames1)

#to fill al elements as above
W[] <- 0.1
#check length
length(W)
#to access one element
W['m','SS1','RR2']
#to check that males total 1
sum(W['m',,])
```        

working on refactoring variables to arrays, made flexible createArray() function.
good progress
started runModel2() to contain the refactoring

**30/4/15** 9.45-13.30 3.75hrs 2.15-5.30 3.25hrs 6-7 8hrs 

Need to be careful about different W fitness variables
`W.RR1_00 : [locus,niche]` single locus (Wlocus)
`W.RR1SS2_0b :   [locus1, locus2, niche]` niche (Wniche)
`W.m.SS1SS2 : [sex, locus1, locus2]` sex (Windiv)

Once I have array refactoring working by .
```{r}

    a <- createArray( sex=c('m','f'), niche1=c('0','a','A'), niche2=c('0','b','B') )
    Wniche <- createArray( locus1 = c('SS1','RS1','RR1'), locus2 = c('SS2','RS2','RR2'), niche1=c('0','a','A'), niche2=c('0','b','B') )    
    Windiv <- createArray( sex=c('m','f'), locus1 = c('SS1','RS1','RR1'), locus2 = c('SS2','RS2','RR2') )
```
I should be able to replace 
```
    W.m.SS1SS2 <- (a.m_00 * W.SS1SS2_00) + 
      (a.m_a0 * W.SS1SS2_a0) + (a.m_A0 * W.SS1SS2_A0) + 
      (a.m_0b * W.SS1SS2_0b) + (a.m_0B * W.SS1SS2_0B) + 
      (a.m_ab * W.SS1SS2_ab) + (a.m_AB * W.SS1SS2_AB) + 
      (a.m_Ab * W.SS1SS2_Ab) + (a.m_aB * W.SS1SS2_aB) 
```
with
```{r}
    Windiv['m','SS1','SS2'] <- sum( a['m',,] * Wniche['SS1','SS2',,])
```
To prove that the above works
```{r}
    #set constant
    a['m',,] <- 2 
    #set variable
    Wniche['SS1','SS2',,] <- 1:9
    Wniche['SS1','SS2',,]
    
    #showing that each element gets multiplied by 2
    a['m',,] * Wniche['SS1','SS2',,]
    
    #show result of sum
    sum( a['m',,] * Wniche['SS1','SS2',,])
    
    Windiv['m','SS1','SS2'] <- sum( a['m',,] * Wniche['SS1','SS2',,])
    
```
Gives a result of 90 for just mSS1SS2 which is correct
``` Windiv
, , locus2 = SS2

   locus1
sex SS1 RS1 RR1
  m  90   0   0
  f   0   0   0

, , locus2 = RS2

   locus1
sex SS1 RS1 RR1
  m   0   0   0
  f   0   0   0

, , locus2 = RR2

   locus1
sex SS1 RS1 RR1
  m   0   0   0
  f   0   0   0
```

Or indeed might be able to replace the whole 120 lines of code from 510-630 with :
No these give: 
```
#Error in a[, , ] * Wniche[, , , ] : non-conformable arrays
    Windiv['m',,] <- sum( a['m',,] * Wniche[,,,])
    Windiv['f',,] <- sum( a['f',,] * Wniche[,,,])
    #or even just
    Windiv[,,] <- sum( a[,,] * Wniche[,,,]) 
```
Just needs a little more thought ...
It's because a needs to be applied repeatedly to each genotype.
Maybe worry about that later ..

Want to develop a sequential testing process, so that I can compare runModel() and runModel2().
Maybe I could use testthat for it ?

Just need to do this to start :
`devtools::use_testthat()`

see testRefactoring() which does things like this :

```
input <- setInputOneScenario()
tst <- runModel(input)
tst2 <- runModel2(input)
input <- setInputSensiScenarios(20, h.RS1_A0=c(0.1,1), h.RS2_0B=c(0.1,1), s.RR1_A0=c(0.2,1), s.RR2_0B=c(0.2,1))
tst <- runModel2(input, produce.plots=FALSE)
tst2 <- runModel(input, produce.plots=FALSE)
identical(tst,tst2)
```

So now I can keep developing runModel2() and use Ctrl T to test I'm not breaking it. 

Think, does my Wniche contain too many elements at 81 ?
i.e. does it contain impossible combinations ? I think I'm OK Beths code fills 9*9 groups of individual variables.

`W.RR1SS2_0b` ... `*81 3*3*3*3` to: Wniche[locus1, locus2, niche1, niche2] 

I may need to add a genotype option to createArray() to cope with f that needs cis & trans


If I convert niche_0B to niche[niche1,niche2] I can simplify further the calculation of 
Two genotype fitnesses in two insecticide Niche
by having a loop that goes through each niche.

Coolio! reduced 250 lines of code to ~ 20 with this
```
    ## Two genotype fitnesses in two insecticide Niche ##

    #!r to replace 250+ lines below
    for( niche1 in dimnames(Wniche)$niche1)
    {
      for( niche2 in dimnames(Wniche)$niche2)
      {
        #if this niche toggled off set fitness to 0
        if (niche[niche1,niche2] == 0)
        {
          Wniche[,,niche1,niche2] <- 0
        } else{
          #otherwise set fitness to product of the 2 loci
          for( locus1 in dimnames(Wniche)$locus1)
          {
            for( locus2 in dimnames(Wniche)$locus2)
            {    
              Wniche[locus1,locus2,niche1,niche2] <- Wloci[locus1,niche1,niche2] * Wloci[locus2,niche1,niche2]
            }
          }          
        }
      }
    }
```

Trying to get fitnesses from my array into the existing output object fbn at the end of runModel2

       -,- a,- A,- -,b -,B a,b A,B A,b a,B
SS1SS2   0   0   0   0   0   0   0   0   0
SS1RS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
SS1RR2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RS1SS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RS1RS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RS1RR2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RR1SS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RR1RS2  NA  NA  NA  NA  NA  NA  NA  NA  NA
RR1RR2  NA  NA  NA  NA  NA  NA  NA  NA  NA

STill not quite working ...

**1/5/15** 9-14.30 5.5hrs 3.15-6 2.75 8.25hrs

Fitness matrix different in my refactored version. But do I know which is correct ?

Browse[2]> fbn
       -,- a,- A,- -,b -,B a,b A,B A,b a,B
SS1SS2   1   0   0   0   0   0   0   0   0
SS1RS2   1   0   0   0   0   0   0   0   0
SS1RR2   1   0   0   0   0   0   0   0   0
RS1SS2   1   0   0   0   0   0   0   0   0
RS1RS2   1   0   0   0   0   0   1   0   0
RS1RR2   1   0   0   0   0   0   1   0   0
RR1SS2   1   0   0   0   0   0   0   0   0
RR1RS2   1   0   0   0   0   0   1   0   0
RR1RR2   1   0   0   0   0   0   1   0   0
Browse[2]> fbn2
       -,- a,- A,- -,b -,B a,b A,B A,b a,B
SS1SS2   0   0   0   0   0   0   0   0   0
SS1RS2   0   0   0   0   0   0   0   0   0
SS1RR2   0   0   0   0   0   0   0   0   0
RS1SS2   0   0   0   0   0   0   0   0   0
RS1RS2   0   0   0   0   0   0   0   0   0
RS1RR2   0   0   0   0   0   0   0   0   0
RR1SS2   0   0   0   0   0   0   0   0   0
RR1RS2   0   0   0   0   0   0   0   0   0
RR1RR2   0   0   0   0   0   0   0   0   0

Aha difference is because Wniche just has 0s in, may be a problem earlier in the program.
Aha issue is that Wloci is not read in yet.

Reading in of Wloci can also be refactored, but I first need to read in h,z,phi & s as arrays.

Then I can try to refactor this :
```
    ## Calculated fitnesses ####
    
    # absence of insecticide
    ## fitness of SS in absence of insecticide is entered above as a parameter
    W.RS1_00 <- 1 - (h.RS1_00 * z.RR1_00)
    W.RR1_00 <- 1 - z.RR1_00
    
    W.RS2_00 <- 1 - (h.RS2_00 * z.RR2_00)
    W.RR2_00 <- 1 - z.RR2_00
    
    # low levels of insecticide a
    W.SS1_a0 <- 1 - phi.SS1_a0
    W.RS1_a0 <- W.SS1_a0 + (h.RS1_a0 * s.RR1_a0)
    W.RR1_a0 <- W.SS1_a0 + s.RR1_a0
    
    # high levels of insecticide A
    W.SS1_A0 <- 1 - phi.SS1_A0
    W.RS1_A0 <- W.SS1_A0 + (h.RS1_A0 * s.RR1_A0)
    W.RR1_A0 <- W.SS1_A0 + s.RR1_A0
    
    # low levels of insecticide b
    W.SS2_0b <- 1 - phi.SS2_0b
    W.RS2_0b <- W.SS2_0b + (h.RS2_0b * s.RR2_0b)
    W.RR2_0b <- W.SS2_0b + s.RR2_0b
    
    # high levels of insecticide B
    W.SS2_0B <- 1 - phi.SS2_0B
    W.RS2_0B <- W.SS2_0B + (h.RS2_0B * s.RR2_0B)
    W.RR2_0B <- W.SS2_0B + s.RR2_0B
```

h only has 6 values 2*3. Because it's only for heterozygous loci & niches.

```
    # h = dominance coefficient
    h.RS1_00 <- input[32,i]
    h.RS1_a0 <- input[33,i]
    h.RS1_A0 <- input[34,i]
    
    h.RS2_00 <- input[35,i]
    h.RS2_0b <- input[36,i]
    h.RS2_0B <- input[37,i]
```

Perhaps best to store as
h[locusNum,niche1,niche2]

This nearly works, but includes un-needed homozygous niches aa etc 
createArray( loci=c('RS1','RS2'), niche1=c('0','a','A'), niche2=c('0','b','B') )

I could add a niches arg to createArray()
createArray( loci=c('RS1','RS2'), niches=c('00','a0','A0') )

BUT no even that doesn't work. 
Because it only needs to contain the insecticide niche associated with the locus.

Actually it just needs :
h[locusNum, nicheLevel] : where insecticide is no, lo, hi
... seeing how that works.

and what about z - fitness cost of resistance allele in no insecticide
z only has 2 values
    z.RR1_00 <- input[42,i]
    z.RR2_00 <- input[43,i]
z[locusNum]
Because by definition it is the fitness cost of one allele in absence of the corresponding insecticide. (so `RR1_00` is a little misleading `z.R1_0` more like it)

phi - fitness of SS in each insecticide/concentration 
just 4 values
    phi.SS1_a0 <- input[26,i]
    phi.SS1_A0 <- input[27,i]
    
    phi.SS2_0b <- input[28,i]
    phi.SS2_0B <- input[29,i]

phi[locusNum, nicheLevel] where nicheLevel is lo, hi

I have Wloci being 54
> str(Wloci)
 num [1:6, 1:3, 1:3] 0 0 0 0 0 0 0 0 0 0 ...
 - attr(*, "dimnames")=List of 3
  ..$ loci  : chr [1:6] "SS1" "RS1" "RR1" "SS2" ...
  ..$ niche1: chr [1:3] "0" "a" "A"
  ..$ niche2: chr [1:3] "0" "b" "B"
> length(Wloci)
[1] 54

but actually I think it only needs 1 niche that corresponds to the chosen locus.
for locus1 it only has as and locus2 it only has bs

look at how Wloci is used to think about how best to structure

This is where it is used 
`Wniche[locus1,locus2,niche1,niche2] <- Wloci[locus1,niche1,niche2] * Wloci[locus2,niche1,niche2]`
I think I need to change from 
`Wloci[locus1,niche1,niche2] * Wloci[locus2,niche1,niche2]`
to
`Wloci[locus1,niche1] * Wloci[locus2,niche2]`

I might want to change niche names to no,lo,hi throughout. Having them different for each locus (0aA & 0bB) causes the code difficulty.

But that is a bigger change making it more difficult to check that the code logic remains the same.

A temporray solution could be to have a lookup from nicheLevel to niche.
e.g. no,lo,hi to 0,a,A & 0,b,B for locus1 & 2 respectively

Later maybe
Wloci[(locusNum),allele=(SS,SR,RR),exposure=(no,lo,hi)]
for now
Wloci[loci=(SS1,SR1,RR1,SS2,SR2,RR2),exposure=(no,lo,hi)]

s selection coeeficient : just 4 values
    # s = selection coefficient
    s.RR1_a0 <- input[38,i]
    s.RR1_A0 <- input[39,i]
    
    s.RR2_0b <- input[40,i]
    s.RR2_0B <- input[41,i]

It's the selection coefficient of resistance in presence of the corresponding insecticide (at lo or hi)
s[locusNum,exposure]
but exposure is only lo or hi. Would a -ve value here for no be the same as z (the fitness cost of resistance in absence of insecticide ?)

cool I think I have generic createArray2 function working

I'm getting closer to fbn & fbn2 being the same

Differences in fbn2:
1) The 00 column has 0s, in fbn it is all 1s
2) Column aB has the values that should be in AB

e.g. SS1RR2
W.SS1RR2_00
Wniche['SS1','RR2','0','0']

Wloci[locus2,exposure2] == 0
but
W.RR2_00 == 1

aha think due to typo here :
Wloci[ paste0('RR',locusNum), 'no'] <- 1 - z[locusNum]

hurrah fixed first problem, now on to 2)

Wniche looks correct
So i think it may be due to ordering of columns in fbn
Which goes AB, Ab, aB, where I think I assume aB, Ab, AB

Wniche['RR1','RR2',,]
      niche2
niche1 0 b B
     0 1 0 0
     a 0 0 0
     A 0 0 1

as.vector( Wniche['RR1','RR2',,])
[1] 1 0 0 0 0 0 0 0 1

It may be tricky for me to get output in exactly the same order as Beths.
How can I use the niche names to name the output matrix ?

Difference in orders
colnames(fbn) <- c("-,-", "a,-", "A,-", "-,b", "-,B", "a,b", "A,B", "A,b", "a,B")
my order is         00     a0     A0     0b     ab     Ab     0B    aB     AB

which is 1,2,3,4,6,8,5,9,7

I might just need to allow that the fitness list is different in my refactor test ?

Still how can I name the output columns from the data ?
Fixed !!

Then I started to delete some of the old code, and found that it broke some things I didn't want to break yet.

Particularly this which suggests I need to get into the cis/trans issue soon:

W.bar.m <- Windiv['m']
W.bar.m <- (f.m.SS1SS2 * W.m.SS1SS2)

revert changes for now and come back to 
 
**5/5/15** 10.30 - 14 3.5hrs 2.45-6.15 7hrs
 
removed refactored fbn code 
 
sent update email to Ian & Beth 

starting to look at cis/trans
this is the first place they appear in the generation loop
```
# set genotype frequencies as variables
# from genotype frequency matrix generated above from initial value of P (freq. of R allele)
# f = frequency before selection

# male
f.m.SS1SS2 <- genotype.freq[1,]
f.m.SS1RS2 <- genotype.freq[2,]
f.m.SS1RR2 <- genotype.freq[3,]
f.m.RS1SS2 <- genotype.freq[4,]
f.m.RS1RS2_cis <- genotype.freq[5,]		### cis
f.m.RS1RS2_trans <- genotype.freq[6,]	### trans
``` 

How is genotype.freq created ?
```
# set up genotype matrix - records frequencies of each of the 9 two locus genotypes each generation
genotype <- matrix( nrow=max_gen, ncol=11 )
colnames(genotype) <- c("gen", "SS1SS2", "SS2RS2", "SS1RR2", 
                        "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                        "RR1SS2", "RR1RS2", "RR1RR2")
## make.genotypemat function will use this data and make a matrix of the genotype frequencies
## frequencies of genotypes before selection - in HW equilibrium and same in male and female
## needs name of matrix and takes corresponding frequency of resistant allele in function call
genotype.freq <- make.genotypemat ( P_1, P_2 )
```

Within make.genotypemat, currently seems that cis & trans are always set to the same ??
```
## two forms of RS1RS2
mat[5,] <- ( loc1[2,]*loc2[2,] ) * 0.5 #cis
mat[6,] <- ( loc1[2,]*loc2[2,] ) * 0.5 #trans

#this is what genotype.freq looks like
                      Freq.
SS1SS2       9.960060e-01
SS1RS2       1.994006e-03
SS1RR2       9.980010e-07
RS1SS2       1.994006e-03
RS1RS2_cis   1.996002e-06
RS1RS2_trans 1.996002e-06
RS1RR2       1.998000e-09
RR1SS2       9.980010e-07
RR1RS2       1.998000e-09
RR1RR2       1.000000e-12
```

aarg createArray2() fails if you pass named variables to it, e.g. sex=sex rather than sex=c('m','f')
seems like mget might solve, but :
mget('sex', inherits=TRUE)
Error: value for ‘sex’ not found

may be able to use this from SO
get_args <- function () {
as.list( match.call(
    def = sys.function( -1 ),
    call = sys.call(-1)) )[-1]

}

BUT from the commandLine it does seem to work ??
```
> sex2 = c('f','m')
> createArray2( sex = sex2 )
sex
f m 
0 0 
```
Ahhh so someohow it does work if the variable is in the Global environment.

eventually by trial & error found this solution :
  #fails
  #dimnames1 <- lapply(listArgs,function(x){eval.parent(x,n=1)})
  #works
  #eval.parent(listArgs$sex)
  #[1] "f" "m"
  #fails
  #dimnames1 <- lapply(listArgs,eval.parent)  
  #works!! n=3 but I'm not sure why !!!
  dimnames1 <- lapply(listArgs,function(x){eval.parent(x,n=3)}) 
 
 
**11/5/15** 1.30-5.45 4.25hrs   
checking current testthat difference
changing from `expect_identical` to `expect_equal` sorted it 
refactoring gamete calculations into createGametes() function 
 

**12/5/15** 9.45-14 4.25hrs 2.30-5.45 3.25 7.5hrs

sorting random mating & maybe getting to nub of cis/trans ?

Beths random mating code :
```
      f.m.SS1SS2 <- 0
      f.m.SS1RS2 <- 0
      f.m.SS1RR2 <- 0
      
      f.m.RS1SS2 <- 0
      f.m.RS1RS2_cis <- 0			#RS1RS2
      f.m.RS1RS2_trans <- 0		#RS1SR2
      f.m.RS1RR2 <- 0
      
      f.m.RR1SS2 <- 0
      f.m.RR1RS2 <- 0 
      f.m.RR1RR2 <- 0
      
      # SS male with SS female
      f.m.SS1SS2 <- f.m.SS1SS2 + ( G.m.S1.S2 * G.f.S1.S2 )
      # SS male with SR female
      f.m.SS1RS2 <- f.m.SS1RS2 + ( G.m.S1.S2 * G.f.S1.R2 )
      # SS male with RS female
      f.m.RS1SS2 <- f.m.RS1SS2 + ( G.m.S1.S2 * G.f.R1.S2 )
      # SS male with RR female
      f.m.RS1RS2_cis <- f.m.RS1RS2_cis + ( G.m.S1.S2 * G.f.R1.R2 )
      
      # SR male with SS female
      f.m.SS1RS2 <- f.m.SS1RS2 + ( G.m.S1.R2 * G.f.S1.S2 )
      # SR male with SR female
      f.m.SS1RR2 <- f.m.SS1RR2 + ( G.m.S1.R2 * G.f.S1.R2 )
      # SR male with RS female
      f.m.RS1RS2_trans <- f.m.RS1RS2_trans + ( G.m.S1.R2 * G.f.R1.S2 )
      # SR male with RR female
      f.m.RS1RR2 <- f.m.RS1RR2 + ( G.m.S1.R2 * G.f.R1.R2 )
      
      # RS male with SS female
      f.m.RS1SS2 <- f.m.RS1SS2 + ( G.m.R1.S2 * G.f.S1.S2 )
      # RS male with SR female
      f.m.RS1RS2_trans <- f.m.RS1RS2_trans + ( G.m.R1.S2 * G.f.S1.R2 )
      # RS male with RS female
      f.m.RR1SS2 <- f.m.RR1SS2 + ( G.m.R1.S2 * G.f.R1.S2 )
      # RS male with RR female
      f.m.RR1RS2 <- f.m.RR1RS2 + ( G.m.R1.S2 * G.f.R1.R2 )
      
      # RR male with SS female
      f.m.RS1RS2_cis <- f.m.RS1RS2_cis + ( G.m.R1.R2 * G.f.S1.S2 ) 
      # RR male with SR female
      f.m.RS1RR2 <- f.m.RS1RR2 + ( G.m.R1.R2 * G.f.S1.R2 )
      # RR male with RS female
      f.m.RR1RS2 <- f.m.RR1RS2 + ( G.m.R1.R2 * G.f.R1.S2 )
      # RR male with RR female
      f.m.RR1RR2 <- f.m.RR1RR2 + ( G.m.R1.R2 * G.f.R1.R2 )
```

Beth only has 10 combinations (including cis/trans), this is because RS is usually same as SR, and in all the names R goes first.
```
genotypes <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
                "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                "RR1SS2", "RR1RS2", "RR1RR2")
```                            


for randomMating this loop creates 16 combinations
including cis/trans

```
  counter <- 0
  
  for( m2 in c('S2','R2'))
  {
    for( m1 in c('S1','R1'))
    {
      for( f2 in c('S2','R2'))
      {
        for( f1 in c('S1','R1'))
        {
          counter <- counter+1
          cat(paste(counter, m1,f1,m2,f2,"\n"))
        }
      }
    }
  }
```
1 S1 S1 S2 S2 
2 S1 R1 S2 S2 
3 S1 S1 S2 R2 
4 S1 R1 S2 R2 
5 R1 S1 S2 S2 
6 R1 R1 S2 S2 
7 R1 S1 S2 R2 
8 R1 R1 S2 R2 
9 S1 S1 R2 S2 
10 S1 R1 R2 S2 
11 S1 S1 R2 R2 
12 S1 R1 R2 R2 
13 R1 S1 R2 S2 
14 R1 R1 R2 S2 
15 R1 S1 R2 R2 
16 R1 R1 R2 R2 

reformat slightly :
1 SS1 SS2
2 SR1 SS2
3 SS1 SR2
4 SR1 SR2
5 RS1 SS2
6 RR1 SS2
7 RS1 SR2
8 RR1 SR2
9 SS1 RS2
10 SR1 RS2
11 SS1 RR2
12 SR1 RR2
13 RS1 RS2
14 RR1 RS2
15 RS1 RR2
16 RR1 RR2

Perhaps I can do the calculations on these 16 & then aggregate to produce the 10 ?

```
#these are other ways of creating the 16

#as an array
arr <- createArray2(loc1all1=c('S','R'),loc1all2=c('S','R'),loc2all1=c('S','R'),loc2all2=c('S','R'))

#as a dataframe
dF <- expand.grid(loc1all1=c('S','R'),loc1all2=c('S','R'),loc2all1=c('S','R'),loc2all2=c('S','R'))
> dF
   loc1all1 loc1all2 loc2all1 loc2all2
1         S        S        S        S
2         R        S        S        S
3         S        R        S        S
4         R        R        S        S
5         S        S        R        S
6         R        S        R        S
7         S        R        R        S
8         R        R        R        S
9         S        S        S        R
10        R        S        S        R
11        S        R        S        R
12        R        R        S        R
13        S        S        R        R
14        R        S        R        R
15        S        R        R        R
16        R        R        R        R

genotypes <- paste0(dF[,1],dF[,2]," ",dF[,3],dF[,4])
#[1] "SS SS" "RS SS" "SR SS" "RR SS" "SS RS" "RS RS" "SR RS" "RR RS" "SS SR" "RS SR" "SR SR" "RR SR" "SS RR" "RS RR" "SR RR" "RR RR"

> length(genotypes)
[1] 16

#or could miss out the space

l1a1 <- substr(genotypes,1,1)
l1a2 <- substr(genotypes,2,2)
l2a1 <- substr(genotypes,4,4)
l2a2 <- substr(genotypes,5,5)

#these are homozygous at both loci and have 1:1 between expanded & contracted genotypes
which(l1a1==l1a2 & l2a1==l2a2)
[1]  1  4 13 16

#this gets the cis/trans
which(!(l1a1==l1a2 | l2a1==l2a2))
[1]  6  7 10 11

But I might not to be this fancy. If I do all the working in the expanded genotype storage, and then just have a single function to do the conversion ??

genotypesContract()
or genotypesLong2Short()

```
 
done : convert from the expanded genotype storage to the contracted one. 
 
Beware I do seem to have occasional small difference between runModel & runModel2. I set the random seed so that the difference is consistent.

It now seems only to turn up if I have >20 scenarios.
 
It's not to do with randomMating() or genotypesLong2Short because they are not fully implemented yet.

Failure(@testRefactoring.r#21): refactoring named variables with arrays doesn't change results 
runModel2(input, produce.plots = FALSE) not equal to runModel(input, produce.plots = FALSE)
Component "results": Component 25: Mean relative difference: 6.391493e-08

It does show some very small differences, but I suspect they are just rounding errors, e.g. this is how they start out. (and the M,F columns just contain 1 despite showing this difference here)

```
listOut2$results[[25]] - listOut$results[[25]]

       Gen          m.R1          m.R2          m.LD          f.R1          f.R2          f.LD             M             F
[1,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
  [2,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
  [3,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 -2.220446e-16 -2.220446e-16
  [4,]   0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
  [5,]   0 -2.168404e-19 -2.168404e-19  0.000000e+00 -2.168404e-19 -2.168404e-19  0.000000e+00 -1.110223e-16 -1.110223e-16
```

tested that my new W.bar calc works :
W.bar
        m         f 
0.1000036 0.1000036 
Browse[2]> W.bar.m
   a.m_00 
0.1000036 
Browse[2]> W.bar.f
   a.f_00 
0.1000036 

**13/5/15** 10-14 2
 
dealing with cis/trans :
```
namesLoci <- c('SS1SS2','SS1RS2','SS1RR2',
               'RS1SS2','RS1RS2cis','RS1RS2trans','RS1RR2',
               'RR1SS2','RR1RS2','RR1RR2')
#or
namesLoci <- rownames( genotype.freq )
sex <- c("F","M")
f <- createArray2( sex=sex, loci=namesLoci )
#i then may be able to refactor more later
```
Plus see randomMating() which uses something like below, and then genotypesLong2Short.r to convert from the expanded genotype format.
```
  counter <- 0
  
  for( m2 in c('S2','R2'))
  {
    for( m1 in c('S1','R1'))
    {
      for( f2 in c('S2','R2'))
      {
        for( f1 in c('S1','R1'))
        {
          counter <- counter+1
          cat(paste(counter, m1,f1,m2,f2,"\n"))
        }
      }
    }
  }
```

deleting old code from runModel2() 

* done to f[m,loci] : sort storage of f.m.RS1RS2_cis
* done : refactor gamete calculations : G.m.R1.R2 <- G.m.R1.R2 + f.m.RS1RS2_cis ...
* done : sort calc of W.bar.m from Windiv
* done : remove reliance on W.m.SS1SS2 etc. (Windiv)
* done : remove reliance on a.m_00 etc.

So far down from 1590 lines to 685, with only a bit in new functions 

**20/5/15** 11-4 in liverpool meeting with Beth and train back

New bit from the manuscript by Ian about including sex-linked genes by a bit of trickery.

It is also possible to allow one of the loci to be sex-linked; here we assume that sex linkage is at Locus 1, and ‘conventional’ sex-determination i.e. that the females is hemozygous sex (XX), the male is XY, and that the gene exists only on the X chromosome. This option is important as Anopheles gambiae has only 2 autosomes and one sex chromosome; approximately ??% of the genes are on the X in this species.

The female genotype frequencies are as given on Table S1 and as described in Equation 4 and Equation 5  because females have two copies of the X chromosome. Locus 1 is homozygous in the male so heterozygotes are impossible at this locus; in this case the allele inherited by males at locus 1  is the maternal-derived one (because they get their X chromosome from their mother and the Y from the father). Hence the male genotypes can be derived by a process similar to that of females and autosomes as shown on Table S2. The males will be simulated  as “RR” or “SS” at the locus even though, in reality” they will be either “R-“ or “S-“. An inherent assumption is therefore that the hemizgous males “R-“ and “S-“ have the same fitness as the equivalent “RR” and “SS” females; this seems reasonable given dosage compensation and is an assumption commonly made.

The male “diploid” genotypes are derived by a similar process to that above and the genotypes are as given on Table S2. The calculation become slightly more complicated to those given above but the same process occurs i.e. Equation 4 becomes

Equation 6

The next genotype can be produced by two combinations so Equation 5 becomes

Equation 7 ...

IH thinks this short-cut will work because selection will be OK, and the males will also correctly pass on their maternal X-linked allele (Table 4). We need to discuss this.

It just means we need edit the code slightly. i.e. calculate the female genotypes then if locus 1 in autosomal just set the male frequencies equal to the females (as we do already), else re-calculate the male frequencies according to Table S2.

end of bit from manuscript
I think this is the important part for implementation :

we assume that sex linkage is at Locus 1, females XX, the male XY, 
and that the gene exists only on the X chromosome. 

If sex-linked Locus 1 is homozygous in the male so heterozygotes are impossible at this locus 
the allele inherited by males at locus 1 is the maternal-derived one 
(because they get their X chromosome from their mother and the Y from the father). 
males will be simulated  as RR or SS at the locus even though, in reality they will be either R- or S-


I think this change needs to be made in the randomMating() function.
Currently i think randomMating() doesn't differentiate between the sex of the offspring.

Does sex linkage have further implications in terms of the generation of gametes ? maybe not 
 
**22/5/15**
To stop things being tracked by git that you have just added to gitignore
```
git rm --cached `git ls-files -i -X .gitignore` 
``` 
Then I just did commit from RStudio.

1. somehow I've broken the check
Failure(@testRefactoring.r#12): refactoring named variables with arrays doesn't change results 
runModel2(input, produce.plots = FALSE) not equal to runModel(input, produce.plots = FALSE)
Component "results": Component 1: 'is.NA' value mismatch: 186 in current 0 in target
Its because of Warning messages like this:
1: In runModel2(input) : Male frequencies before selection total != 1 0.997966163681722

Perhaps I need to make the checks less sensitive ??
Is it because I re-enabled checks that had been disabled by Beth ?
I reduced the tolerance in checks, runs OK, but tests still fail.

Points to some difference having crept in ...

difference is in 
genotype outputs in generations after 1
all those starting with RS1

to do with object genotype ....
Aha! problems were due to my new sex-linked changes that were operating without me realising.
So I should probaly remove the tolerances I put in. and work out how to reassign the impossible genotype freqs.
done 
 
**4/6/15** thurs 12.15- 

checking on implementation of sexLinkage in randomMating() 
 
sexLinked probably needs to be included as an extra paramter in input (but should make it robust to cope if an input file doesn't have the sex-linked option)

be careful because this could break previous model runs

genotype.freq mostly removed it was the same as f['m',]. Creation of f & fs moved to before the generations loop. This was prompted because previously genotype.freq was the same for males and females and it adds an unecessary extra layer of potential confusion.
 
added calculating genotypes of both m&f in runModel2() to allow sex linkage.
 
bug in random mating with sex linkage genotypes aren't summing to 1

fGenotypeExpanded for non sex-linked looks like the top one
In the lower one the R1 homozgotes should sum to 1 too

    l1a2
l1a1     S1     R1
  S1 0.0625 0.0625
  R1 0.0625 0.0625

    l1a2
l1a1    S1     R1
  S1 0.125 0.0000
  R1 0.000 0.0625

Fixed now :
with this bit of code :
```
#heterozygotes at locus1 are impossible so add to the homozygotes instead
if(f1!=m1)
{
  #add to the homozygotes : f1,f1
  fGenotypeExpanded[f1,f1,f2,m2] <- fGenotypeExpanded[f1,f1,f2,m2] + fThisGenotype
} else #i.e. if not heterozygous at locus 1
{
  #just add to this genotype : f1,m1
  #have to add rather than set in case this is a homozygote that has already been added to by previous condition
  fGenotypeExpanded[f1,m1,f2,m2] <- fGenotypeExpanded[f1,m1,f2,m2] + fThisGenotype
}
```

Created a new shiny app in shinyCurtis2 and run from runUI2.R to look at graphs with & without sex linkage side by side. Initially these gave identical results, which was because resistSimple() called the old runModel() which doesn't allow sex linkage. Try it again now.

Cool, now I am getting a difference between locus1 & locus2. However males & females still have the same results, is that what would be expected ?

Maybe it's because of the way that results are output from Beths one sex system ?
Where does the graph get the data from ?

genplot <- plotallele.freq.andy( listOut$results[[1]] ) 

  # Males
  lines( mat[,1], mat[,2], col="darkblue", lwd=2, lty=13 )
  lines( mat[,1], mat[,3], col="green", lwd=2, lty=14 )
  # Females
  lines( mat[,1], mat[,5], col="red",lwd=2, lty=15 )
  lines( mat[,1], mat[,6], col="orange", lwd=2, lty=16 )

so its columns 2,3,5 & 6 I'm interested in

so where is listOut$results[[1]] filled ?
in runModel2()     listOut$results[[i]] <- results
& further back :
```
      ## frequency of resistance alleles
      m.R1 <- sum(f['m',grep("RR1",colnames(f))]) + ( 0.5 * sum(f['m',grep("RS1",colnames(f))]))
      m.R2 <- sum(f['m',grep("RR2",colnames(f))]) + ( 0.5 * sum(f['m',grep("RS2",colnames(f))]))
      f.R1 <- sum(f['f',grep("RR1",colnames(f))]) + ( 0.5 * sum(f['f',grep("RS1",colnames(f))]))
      f.R2 <- sum(f['f',grep("RR2",colnames(f))]) + ( 0.5 * sum(f['f',grep("RS2",colnames(f))]))   
      results[k,2] <- m.R1
      results[k,3] <- m.R2
      results[k,5] <- f.R1
      results[k,6] <- f.R2
```

Seems that this should be doing the right thing. Maybe the allele frequencies overall don't change, because for the males they just redistributed from heterozygotes to homozygotes.

**5/6/15**

Instructions emailed to Ian & Beth to run UI2
require(devtools)
install_github('AndySouth/resistance')
require(resistance)
runUI2()

Ians feedback on UI2 (sex-linkage)

The R script runs fine. The results are as I would have expected given that locus 1 is sex-linked and locus 2 is autosomal i.e. resistance evolves faster at the sex-linked locus.

The dynamics at locus 2 appear unchanged...I therefore assume the drugs were deployed individually i.e. not as a combination?

But I thought they were applied as a combination.
The exposure comes from here in the UI :

 h5("Exposure to each insecticide"),
 numericInput("a.m_AB", "same for both insecticides in Curtis", 0.5, min = 0.1, max = 0.9, step = 0.1)

and is set here in server :
a.m_AB = input$a.m_AB,
a.f_AB = input$a.m_AB, #set f to same as m

are these the correct parameters to set ??
yes i think so.


Thanks Ian,
In relation to your 2nd comment, from the UI, both a.m_AB and a.f_AB
are set from the box with "Exposure to each insecticide".

It's my understanding that should result in exposure to high levels of
both insecticides (i.e. in combination).

Is that correct Beth ?

Didn't get a reply from Beth yet.
From Ian :
OK, I was just speculating that if resistnce at locus 1 was spreading faster due to sex-linkage then so might resistnce at locus2 through genic linkage and hitch-hiking. I'll give its some more thought over the weekend.

be careful that runModel & runModel2 should no longer generate same results when sex linkage is set (but i don't think this should break current testthat tests)

**8/6/15** on train to liverpool 4.30-9.30
**The plan for running the sensitivity analysis for the first paper**

Edited from Ians text in the paper.

For each parameter combination 2 strategies : 
A) single ‘new’ insecticide alone
B) ‘new’ insecticide in combination with ‘older’ insecticide 

The simulation output was the ratio of number of generations for resistance  to the ‘new’ insecticide  to reach a given allele frequency when in combination compared to on its own.

A ratio >1 indicates that combination is better. 

Five threshold frequencies were investigated: 0.05, 0.1, 0.2, 0.5, 0.8.

But in the document :
"Suggested sensitivity analysis.docx" it talks about sequential and 2 mixture optoins.

Sequential use:, use one insecticide until resistance frequency reaches critical point, then switch insecticide until second resistance allele frequency reaches the critical point. Record time in total generations for the two insecticides.

Mixture option 1. Deploy mixture and record time in generations until EITHER frequency reaches the critical point

Mixture, option 2: deploy mixture until one resistance allele reaches the critical point, then stop using that insecticide in the mixture. Record time until both resistance allele frequencies reach critical point.

Easy by post-processing : Mixture 1 : set exposure to both insecticides
Trickier : Sequential
Trickiest : Mixture 2

Mixture 1 : set exposure to both insecticides and post process to get time until critical points for EITHER insecticide is reached.
Sequential : set callibration to 1013, add bit to runModel2() to switch from insecticide1 to insecticide2 when critical point for the first insecticide is reached. Post process results file to find the time that the 2nd critical point is reached.

Tried adding a sequential insecticide scenario as an example to runModel2(), doesn't seem to be working yet, or maybe I just haven't passed it the correct args.

aha, i think probably because the niche toggles not switched on. Can I set all of these to on by default so that this problem doesn't happen again ?? done

resistance to 2nd insecticide still seems not to rise.

It may be because I need to reset the fitnesses (which happens before gen loop):           

Windiv[sex,locus1,locus2] <- sum( a[sex,,] * Wniche[locus1,locus2,,])
and could put this into a function called setFitness() or the like.


** 9/6/15 ** 4-5

Chatting with Ian & working out that both sequential and mixture2 scenarios can also be done by post-processing of the output file.

Now looking at setting up sensiAnPaper1() to run the sensitivity analysis for the paper.
Realising that I might not be able to call sensiAn1 and setInputSensiScenarios() because of some of the dependencies between parameters.

Particularly that `P_2` needs to be set as a proprtion of `P_1`.

Actually sensiAnPaper1() can be much more straightforward if it has hardcoded args and doesn't need to be so flexible.


** 10/6/15 ** train back from liverpool

modifying sensiAnPaper1() to run the different insecticideUsed scenarios
started sensiPostProc() to do post-processing


** 24/6/15 ** in liverpool before talking to Ian

checking where I have got to.

I'm going to renam sensiAnPaper1 to sensiAnPaperPart()
then create a sensiAnPaperAll() to run everything (although in practice unlikely to run it all at once)

** 29/6/15 **
little on sensiPostProc()

** 30/6/15 **
moved sensiAnPaper1All() into an rmarkdown file - because in reality I'm not likely to run it all at once.

Beths work just uses male allele frequencies to assess criticalPoints m.R1 & f.R1 are the same for now, but it might be safer for me to use an average.

~ added the input matrix onto the listOut output object in runModel2() to help with post processing
Slight issue that this has now broken my testthat tests !!


How can m.R2 and f.R2 be greater than 1 (e.g. they start at 1.7455 in first run) ?
Is this because the frequency of R2 is set too high at start ?

m.R2 is calculated at line 801 in the older runModel.r and line 376 in runModel2

Frequency of resistance allele to the 2nd insecticide is starting above 1.

m.R2 <- ( f.m.SS1RR2 + f.m.RS1RR2 + f.m.RR1RR2 ) + ( 0.5 * (f.m.SS1RS2 + f.m.RS1RS2_cis + f.m.RS1RS2_trans + f.m.RR1RS2 ) )

#refactored
m.R2 <- sum(f['m',grep("RR2",colnames(f))]) + ( 0.5 * sum(f['m',grep("RS2",colnames(f))]))
 
Can I create a reproducible example ? or should I just ask first ?

What is the starting value of P_2 the starting frequency of resistance to insecticide2 ?
For the first dodgy run
listOutMix$input['P_2',1]
    P_2 
1.74551 

Aha yes, so maybe i should put something in to stop it going above 1 ??

I think problem was because I set the upper limit on P_2 to 100 rather than 1 (based upon Ians table).

FRom Ian :
A trick one that..

I think my first instinct was to not restrict one insecticide resistance frequency to be always  less than the other. But we would have to censor the inputs because if frequency at locus 1  is 0.1 then the frequency at locus 2 cannot be 100 times higher.

I think we should go with your suggestion i.e. recognise that frequency at one locus will always be less than at the other and the lower frequency occurs at Locus 2. I would maybe go from 0.0001 to 1 under these circumstances so we get a wide range of starting frequency ratios.

Cool findResistancePoints() seems to be working

> findResistancePoints(listOut, locus=1)
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19]
gen_cP0.1     9   10    5   36    4    3    5   19   48    77    12    10     3     2     4     3    31    36     5   
gen_cP0.25   20   20   27   70   15   16   11   41   72   120    19    16    17     5     8     9    54    71    20   
gen_cP0.5    27   29   39   89   28   25   18   56   97   146    25    23    36     9    12    16    73    95    39   
> findResistancePoints(listOut, locus=2)
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19] 
gen_cP0.1     4   31   21   76   25   14    7   30   23   301    59     8    34     2    25    91    19    26    97   
gen_cP0.25    8   48   45  128   56   26   16   40   35   334    79    12    42     5    30   118    29    53   134   
gen_cP0.5    13   61   62  154   77   36   24   51   48   357    90    17    51     9    34   133    39    81   166   
> findResistancePoints(listOut, locus='either')
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19]
gen_cP0.1     4   10    5   36    4    3    5   19   23    77    12     8     3     2     4     3    19    26     5 
gen_cP0.25    8   20   27   70   15   16   11   40   35   120    19    12    17     5     8     9    29    53    20   
gen_cP0.5    13   29   39   89   28   25   18   51   48   146    25    17    36     9    12    16    39    81    39   
> findResistancePoints(listOut, locus='both')
           [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14] [,15] [,16] [,17] [,18] [,19]
gen_cP0.1     9   31   21   76   25   14    7   30   48   301    59    10    34     2    25    91    31    36    97   
gen_cP0.25   20   48   45  128   56   26   16   41   72   334    79    16    42     5    30   118    54    71   134   
gen_cP0.5    27   61   62  154   77   36   24   56   97   357    90    23    51     9    34   133    73    95   166   

but the 2nd mixture scenario is the tricky bit because 

need to know which of the 2 insecticides reached the critical point first, and the resistance level of the other. I think I'll need to create a special function.

findResistancePointsMixResponsive()

** 31/6/15 **
Quick review of what needs to go into the final decision tree.

slight rethink : I want the input files to be mostly the same for I1,I2 & mixture, so that they can be added together in combined scenarios.

I added a set.seed into sensiAnPaperPart()
This shows the input files for diffeerent insecticide strategies are now the same.

tst <- listOutMix$input[c('P_1','P_2','a.f_AB','phi.SS1_A0','phi.SS2_0B','h.RS1_A0','h.RS2_0B','s.RR1_A0','s.RR2_0B'),]
#note a.f_A0 rather than a.f_AB
tst2 <- listOutI1$input[c('P_1','P_2','a.f_A0','phi.SS1_A0','phi.SS2_0B','h.RS1_A0','h.RS2_0B','s.RR1_A0','s.RR2_0B'),]
identical(as.numeric(tst),as.numeric(tst2))


** 6/7/15 **
`plotcurtis_f2_generic` : a function allowing mixture & individual scenarios to be compared, similar to Curtis Fig2. 

In creating a shiny app to display this, I could allow user to select parameters, but then the run might take a little while because it has to do 3 runs (I1,I2,mix) & post processing. Quicker may be to display my sensitivity analysis results & allow user to step through scenarios. + include printing out the input values used (& maybe even their relative position within ranges). 


** 7/7/15 **
resistanceMaster() does reproduce curtis fig2
so the parameters in system.file("extdata","input.parameters.csv", package="resistance")
must work 

what is different about the input params in the sensitivity analysis ?

resistanceMaster() still calls runModel() if I change it to call runModel2() does it still work ??

No I get this :
Error in runModel2(input, calibration) : 
  Error in male exposures: must total one:  1.9

but in the input files, looks like it is -- 0.1 & AB 0.9 for all. 
 
No not in the hardcoded, curtis fig2 input which comes from the program. It has a 0.9 for ab which slips in too. 
 
                                             B (HCH)   A (DDT) Combination
Calibration (100 default)                  1012.0000 1012.0000   1.012e+03
Number of generations                        70.0000   70.0000   1.600e+02
Collect fitness scores in matrix (1/0)        1.0000    1.0000   1.000e+00
Export fitness scores matrix to .csv (1/0)    0.0000    0.0000   0.000e+00
Frequency of R at locus 1                     0.0100    0.0100   1.000e-02
Frequency of R at locus 2                     0.0100    0.0100   1.000e-02
Recombination Rate                            0.5000    0.5000   5.000e-01
Exposure Males -,-                            0.1000    0.1000   1.000e-01
Exposure Males a,-                            0.0000    0.0000   0.000e+00
Exposure Males A,-                            0.0000    0.9000   0.000e+00
Exposure Males -,b                            0.0000    0.0000   0.000e+00
Exposure Males -,B                            0.9000    0.0000   0.000e+00
Exposure males a,b                            0.0000    0.0000   9.000e-01
Exposure Males A,B                            0.0000    0.0000   9.000e-01
...

exactly where is that input object created ?
createInputMatrix()
fixed bug, but it didn't effect the output


can I compare the inputs for the 3 scenarios by Beth to those in my first scenario ? in a csv.

This gets the file to create Curtis Fig2.
```
inputCurtis <- createInputMatrix(FALSE)

```

Can I paste the input for the first scenario onto it ?
Or might be better just to output all 100 scenarios.

see : 

seems that the dominance coefficient of locus2 in -B is much lower in Beths than in the sensi scenarios.

Dominance coefficient L2 in -,B	0.0016	0.0016	0.00016	0.944675269	0.497699242

First see if the curtis fig2 persists when dominance for the mixture is reduced.

Good news that this correction seemed to make the graph closer to Curtis. 
    #input[37,3] <- 0.00016	#Dominance coefficient in B
    #? should this be the same as the value in ,2 & ,1
    input[37,3] <- 0.0016	#Dominance coefficient in B

** 7/7/15 **
new version of UI comparing scenarios with greater differences in dominance.
Still doesn't produce something similar to curtis fig2 where mixture does much better.

created runcurtis_f2() to do runs to recreate curtis fig2. recreates it fine.

created shinyFig2Curtis, that calls runcurtis_f2()

for some reason it currently doesn't recreate the curtis plot.
what inputs are wrong in shinyFig2Curtis ?
aha! the difference was just `P_1 & P_2` which were set to 0.1 rather than 0.01
aha2, even after that fixed, something still not quite right

this was because phi.SS2_0B = 1 was set to 0 instead. This is the fitness of exposed susceptibles.
(1 suggests that the susceptible is completely fit in presence of the insecticide which doesn't seem right).

yes phi.SS2_0B seems key :
compare these 2:
```{r}
#curtis default phi.SS2_0B = 1 : mixture better than sequence
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )
#reduce phi.SS2_0B to 0.6 : mixture same as sequence
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 0.6 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

```

How does that relate to this comment from the paper ?
In particular, Curtis [1] allowed the resistant and sensitive homozygotes to survive DDT exposure with finesses of 50% and 27% respectively (caption to his Figure 2).
 
packageDescription('resistance')

GithubSHA1: 7858f1edc0022233c0cd26f84be64e1a9bd0cd04 
            7858f1edc0022233c0cd26f84be64e1a9bd0cd04
 
This shows that the installed package is the latest version from github.

problem was because I hadn't updated tghe NAMESPACE Ctrl Shift D

reran model with new ranges for phi vals (0.6 to 1)
(but still need to sort finite 'xlim' values errors)

e.g. scenario 2

Email from Ian, showing section vi of sensitivity doc.

We have to ensure the fitness of the RS and RR genotypes does not exceed 1.0 as this is the reference (maximum value of the SS genotype in the absence of insecticide. The fitness of the SS homozygote is determined by the ϕ values in Table 3 i.e. as 1- ϕ. Hence the values of s must be less than ϕ. I suggested we set s=x ϕ where x comes for a uniform distribution between 0.2 and 1.0”

This relates to this bit of code :

      for( exposure in c('lo','hi') )
      {
        Wloci[ paste0('SS',locusNum), exposure] <-  1 - phi[locusNum, exposure] 
        
        Wloci[ paste0('RS',locusNum), exposure] <- (1 - phi[locusNum, exposure]) + 
                                                   (h[locusNum, exposure] * s[locusNum, exposure])
        
        Wloci[ paste0('RR',locusNum), exposure] <- (1 - phi[locusNum, exposure]) + 
                                                   (s[locusNum, exposure])
      }
      

Wloci[RR] <- (1-phi) + s

so s must be < phi to keep Wloci[RR] below 1

added in checks for fitnesses >1 or <0 in runModel2


initial tree looks promising :
for mix1 better than sequential:
phi.SS2_0B < 0.66
phi.SS1_AO < 0.68
P_2 > 0.033
P_1


** 10/7/15 **
installed rpart.plot
for prp function used by susana

created loop to do all plots, with titles and save

repeated runs for 500 scenarios, took about an hour.
 
** 11/7/15 **
running 1000 scenarios
started at 10.15, check timestamp on outputs to see how long it took

** 13//7/15 **

done created findResistancePointsMixResponsive() for mix2

analysing 1000 runs

  #1000runs
  #betterMix1Seq_cP0.1 betterMix1Seq_cP0.25  betterMix1Seq_cP0.5 
  #              0.178                0.232                0.238 
  #betterMix2Seq_cP0.1 betterMix2Seq_cP0.25  betterMix2Seq_cP0.5 
  #              0.471                0.568                0.551 
  #betterMix3Seq_cP0.1 betterMix3Seq_cP0.25  betterMix3Seq_cP0.5 
  #              0.794                0.726                0.673 
rounded
  #betterMix1Seq_cP0.1 betterMix1Seq_cP0.25  betterMix1Seq_cP0.5 
  #              0.18                0.23                0.24 
  #betterMix2Seq_cP0.1 betterMix2Seq_cP0.25  betterMix2Seq_cP0.5 
  #              0.47                0.57                0.55 
  #betterMix3Seq_cP0.1 betterMix3Seq_cP0.25  betterMix3Seq_cP0.5 
  #              0.79                0.73                0.67 
  
essentially
mix1 : 0.2
mix2 : 0.5
mix3 : 0.7


alas the trees are more complicated, but that might be fixed by pruning them ?
maybe not ... 
does work well for :
betterMix1Seq_cP0.1
 
BUT wierdly seems that plots look better (ie much fewer branches) when done individually than when saved in the loop.

is it something to do with interaction between prp and savePlot() ?
 
Now I seem to be getting a different result each time ????? 
 
** 14/7/15 **
 
putting results document together. 
 
Just checking that the results file for the 1000 runs looks like the figures & UI for the 100 runs 
 
run3: sequential best (seq 28+29, mix3 50)
run4: mixture best (seq 93+32, mix3 172)
run5: resistance to one in mixture develops more slowly than both in sequence (seq 29+103, mix1 152)
 
resistPointsI1[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1    38  999   11   57    7
gen_cP0.25   48  999   21   82   22
gen_cP0.5    53  999   28   93   29
resistPointsI2[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1    64  210    7    4   91
gen_cP0.25   90  250   21   21   99
gen_cP0.5   102  268   29   32  103
resistPointsMix1[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1    75  999    9   10   34
gen_cP0.25   93  999   30   57  116
gen_cP0.5   102  999   41   86  152
resistPointsMix3[,1:5]
           [,1] [,2] [,3] [,4] [,5]
gen_cP0.1   168  999   21  129  191
gen_cP0.25  199  999   38  159  202
gen_cP0.5   213  999   50  172  207
 
 
Now can I do the 20% better results ? 
 betterMix1Seq20_cP0.1 betterMix1Seq20_cP0.25  betterMix1Seq20_cP0.5 
                 0.136                  0.175                  0.173 
 betterMix2Seq20_cP0.1 betterMix2Seq20_cP0.25  betterMix2Seq20_cP0.5 
                 0.241                  0.326                  0.313 
 betterMix3Seq20_cP0.1 betterMix3Seq20_cP0.25  betterMix3Seq20_cP0.5 
                 0.672                  0.560                  0.488  
 
rounded
 betterMix1Seq20_cP0.1 betterMix1Seq20_cP0.25  betterMix1Seq20_cP0.5 
                 0.14                  0.18                  0.17 
 betterMix2Seq20_cP0.1 betterMix2Seq20_cP0.25  betterMix2Seq20_cP0.5 
                 0.24                  0.33                  0.31 
 betterMix3Seq20_cP0.1 betterMix3Seq20_cP0.25  betterMix3Seq20_cP0.5 
                 0.67                  0.56                  0.49  
 
 
Low phi values lead to sequential being better.
e.g. https://andysouth.shinyapps.io/shinyMixSeqCompare1 
runs 
3: phi 0.53, 0.79
19: phi 0.65, 0.44
25: phi 0.60, 0.56
 
** 15/7/15 ** 
 
installed triangle package
this to generate random deviates
rtriangle(n, a=0, b=1, c=(a+b)/2)

changed label 'combination' to 'mixture' in my generic curtis fig2 plot

** 16/7/15 **
reran 1000 sensitivity scenarios with new agreed input ranges
& reran 100 scenarios and put them into shinyMixSeqCompare1


** 21/7/15 **
check that the example runs still support the args the way they did before

the 3 example plots that I pasted into the word doc seem to be identical to what they were before
ARSE is it possible I didn't reload the package so that it used the old function version with the old input ranges ?

yes I think that's what I did !

rereran 100 runs and put them into shinyMixSeqCompare1

sorted order of insecticides for sequential figures

run shinyMixSeqCompare1 in a window rather than the browser to enable copy and pasting figures into word doc

copied 3 figs into the word doc. Note that run 9 is interesting. Mixture does much better than sequential.

I'm a bit worried that the post-processed results for the new100 runs seem not to have changed even though
 
** 22/7/15 **
analysing 1000 rereruns 

betterMix1Seq_cP0.1 betterMix1Seq_cP0.25  betterMix1Seq_cP0.5 
               0.223                0.245                0.243 
betterMix2Seq_cP0.1 betterMix2Seq_cP0.25  betterMix2Seq_cP0.5 
               0.558                0.563                0.534 
betterMix3Seq_cP0.1 betterMix3Seq_cP0.25  betterMix3Seq_cP0.5 
               0.820                0.747                0.677 
               
rounded
betterMix1Seq_cP0.1 betterMix1Seq_cP0.25  betterMix1Seq_cP0.5 
               0.22                0.25                0.24
betterMix2Seq_cP0.1 betterMix2Seq_cP0.25  betterMix2Seq_cP0.5 
               0.56                0.56                0.53
betterMix3Seq_cP0.1 betterMix3Seq_cP0.25  betterMix3Seq_cP0.5 
               0.82                0.75                0.68
               
20% better :
 betterMix1Seq20_cP0.1 betterMix1Seq20_cP0.25  betterMix1Seq20_cP0.5 
                 0.163                  0.178                  0.182 
 betterMix2Seq20_cP0.1 betterMix2Seq20_cP0.25  betterMix2Seq20_cP0.5 
                 0.283                  0.302                  0.283 
 betterMix3Seq20_cP0.1 betterMix3Seq20_cP0.25  betterMix3Seq20_cP0.5 
                 0.622                  0.536                  0.458 
rounded               
 betterMix1Seq20_cP0.1 betterMix1Seq20_cP0.25  betterMix1Seq20_cP0.5 
                 0.16                  0.18                  0.18 
 betterMix2Seq20_cP0.1 betterMix2Seq20_cP0.25  betterMix2Seq20_cP0.5 
                 0.28                  0.30                  0.28 
 betterMix3Seq20_cP0.1 betterMix3Seq20_cP0.25  betterMix3Seq20_cP0.5 
                 0.62                  0.54                  0.46 
               

reran trees
 
Best ones :

 
to try to find an approximate general rule 

Favouring mixtures over sequential use.
insecticide effectiveness (phi) > 0.7
exposure < 0.7

e.g. see treebetterMix2Seq_cP0.5.jpg


I haven't done 20% better trees yet. Should I ?
done now

** 23/7/15 **
prettifying trees
changed response from 0/1 to mixture/sequence 

seems that it may have made the trees bigger, ARSE.

is it to do with pruning ?

tree$cptable[which.min(tree$cptable[,"xerror"]),"CP"]
 
** 30/7/15 **

checking on trees, and why they seem to change when I tweak formatting options
add a check into sensiAnPaper1All.Rmd

The tree is the same at the top (if reversed) but has more branches when 0/1 was changed to sequence/mixture.

treeResponse <- "betterMix2Seq_cP0.5"
    
tree$cptable[which.min(tree$cptable[,"xerror"]),"CP"]
#[1] 0.01

What was this error before I made the conversion to text ?

I tried converting back to numeric and I got the same big tree ...

but the value of the error is lightly different :
tree$cptable[which.min(tree$cptable[,"xerror"]),"CP"]
[1] 0.01072961

This is what the cptable looks like for the numeric version
numeric
> tree$cptable
          CP nsplit rel error    xerror       xstd
1 0.34549356      0 1.0000000 1.0000000 0.03385148
2 0.12875536      1 0.6545064 0.8047210 0.03285260
3 0.05150215      2 0.5257511 0.6931330 0.03173289
4 0.04184549      3 0.4742489 0.5107296 0.02889882
5 0.01824034      5 0.3905579 0.4527897 0.02768814
6 0.01287554      7 0.3540773 0.4163090 0.02683382
7 0.01180258      8 0.3412017 0.4098712 0.02667504
8 0.01072961     12 0.2939914 0.4012876 0.02645941
9 0.01000000     13 0.2832618 0.4098712 0.02667504

mixture/sequence
> tree$cptable
          CP nsplit rel error    xerror       xstd
1 0.34549356      0 1.0000000 1.0000000 0.03385148
2 0.12875536      1 0.6545064 0.7532189 0.03238844
3 0.05150215      2 0.5257511 0.6287554 0.03088571
4 0.04184549      3 0.4742489 0.4957082 0.02860113
5 0.01824034      5 0.3905579 0.4699571 0.02806478
6 0.01287554      7 0.3540773 0.4527897 0.02768814
7 0.01180258      8 0.3412017 0.4420601 0.02744467
8 0.01072961     12 0.2939914 0.4377682 0.02734549
9 0.01000000     13 0.2832618 0.4442060 0.02749387

Aha! when I repeat identical model, cptable is different 
> tree$cptable
          CP nsplit rel error    xerror       xstd
1 0.34549356      0 1.0000000 1.0000000 0.03385148
2 0.12875536      1 0.6545064 0.7618026 0.03247193
3 0.05150215      2 0.5257511 0.6351931 0.03097749
4 0.04184549      3 0.4742489 0.5085837 0.02885696
5 0.01824034      5 0.3905579 0.4635193 0.02792537
6 0.01287554      7 0.3540773 0.4248927 0.02704169
7 0.01180258      8 0.3412017 0.4163090 0.02683382
8 0.01072961     12 0.2939914 0.4098712 0.02667504
9 0.01000000     13 0.2832618 0.4098712 0.02667504


Can I get back to the nice smaller, neater tree ??


http://stackoverflow.com/questions/29197213/what-is-the-difference-between-rel-error-and-x-error-in-a-rpart-decision-tree

A rule of thumb is to choose the lowest level where the rel_error + xstd < xerror.

http://stackoverflow.com/questions/15318409/how-to-prune-a-tree-in-r

This does that :

nsplit <- tree$cptable[ tree$cptable[,"rel error"] + tree$cptable[,"xstd"]  < tree$cptable[,"xerror"],"nsplit"][1]


** 31/7/15 **
give more sensible error than "need finite 'xlim' values" from shiny curtis fig2 app
set starting freq of resistance to 0.0001 for each to generate the error.

this is when resistance is not reached within max generations.

runcurtis_f2()
plotcurtis_f2_generic()
Error in 1:ddt_cutoff : result would be too long a vector

fixed by adding in this code to plotcurtis_f2_generic() 
  # check if resistance has not been reached
  # and prevent an error, by seeting to max generations from input
  # beware this may have other unintended consequences
  if (is.infinite(hch_cutoff)) hch_cutoff <- nrow(amat)
  if (is.infinite(ddt_cutoff)) ddt_cutoff <- nrow(bmat)  

uploaded new version to shinyapps


** 26/8/15 **
submit invoice : from contract : maximum of 25 days at 217 per day
I spent 30.6 days (can claim some against tsetse work, or from future contract ...)


** 6/10/15 **
Chatting to Joe Lines at OpenMalaria TAG
~ liked the shiny front end
~ asked when he would be able to play with the model
~ would like to see plots of parameter space, 2D & 3D when mixture better than sequential 
~ suggested that Curtis does not say mixtures are bad (& that Ian may have slightly skewed impression)
~~ I said that sentence in GPIRM might suggest that mixtures not as good
~ he said with classification trees he would like to see an option to influence which branches are done first based upon the operational ability to alter them
# to do re Joe
~ send Joe link to the other shiny app
~ send him the sentence from GPIRM
~ ask Joe for a paper copy of GPIRM

Pre-existing resistance to one insecticide in the mixture could accelerate the development of double resistant vector populations because of linkage disequilibrium (Figure A9.1). 
 
 
GPIRM p121 "theoretical models suggest that mixtures might delay resistance longer than rotations
or broad mosaics (1)."

p122
"there is general agreement that LLIN mixture products, when developed, could
still contain pyrethroids, since the pyrethroid may still contribute
to the effectiveness of the net despite resistance (this needs to
be confirmed). Some agree that the rationale applies to IRS, and
that pyrethroid mixtures could be considered in areas of pyrethroid
resistance; others believe that this is inadvisable and likely to lead
to rapid evolution of double resistance."

p122 
Pre-existing resistance to one insecticide in the mixture could accelerate the development of double resistant vector populations because of linkage disequilibrium (Figure A9.1).

At current prices, rotations and mosaics are estimated to have the lowest cost effect among the IRS-based IRM interventions. They would increase the annual cost by 20–50%, whereas use of mixtures for IRS once developed, are projected to increase the cost by 30–100%, depending on the duration of the transmission
season. Because mixtures have not yet been developed or marketed, the final cost could be higher or lower than current projections.

1. World Health Organization. The technical basis for coordinated action against insecticide resistance: preserving the effectiveness of modern malaria vector control. Geneva, 2011.
 
Curtis is ref. 7. 
Perhaps I hadn't remembered very well what it said. Or maybe I was remembering more what is in the WHO 2011 publication ? Curtis paper not cited in WHO 2011.


** 7/12/15 **

email from Ian :

I had a major session on the 2-locus insecticide-resistance manuscript over the weekend (in between DIY; what a dull life I live!). Anyway I am now fairly happy with it . I plan to submit a IR-modelling grant application on 12th January so would like to have the paper submitted by then. I am not sure if I will manage but at least it is a target... I assume you can see the current version in Dropbox.

Andy,  can you commit a couple to days to it ,or however long it takes, charging to the malaria grant (i.e. the one working with Swiss TPH). Here is a shopping list, but I think #3 is the only one that may require a significant amount of coding


As I note in the cover page:  The big change I made was that calculating fitness (equations 1 and 2) was previously described as occurring within the generations. In fact it does not depend on current genotype frequencies so only needs to be done once prior to each deployment scenario  i.e. before the simulation is iterated over the generations. Check the code...if it is  still calculated each generation, it can be moved out, calculated once prior to the iterations,  and computational speed may increase substantially. Andy, you might like to check that it does not currently depend on current genotype frequencies in the code.

The classification trees are currently based on 1,000 samples. Can you increase them to 5,000. The tree for "combinations only better if last >20% longer" needs to be re-drawn with more transparent parameter names. I have updated Table 5 with some suggested variable names for the trees... feel free to change and/or shorten them if you wish. I assume it is trivial to rename the variables in the tree graphing package. I think we need the pruned trees (i.e. within 1 s.e.) for the main manuscript and the full, unpruned trees for the supplementary material.

At the moment we simply re-create the results from Curtis. I think a slight extension will greatly improve the ms by demonstrating the flexibility of the methodology.   I have proposed two small extensions: as described on  Page 14/15. One is for biological realism i.e. male Anopheles mosquitoes are less likely to enter the house so will typically have lower exposure to the insecticide than females. The second is operational: even if we propose to deploy a mixture then various factors (theft, sloth, stock-outs etc) mean that one insecticide will often be absent from the mixture so the other insecticide is sprayed alone. As before, a sensitivity analysis with 5,000 runs. I think we just use these as examples then wrap up the ms by arguing that many other examples could be explored but we lack time/space so have focussed on describing the basic methodology to be used in future more policy-orientated publications

That should keep you (or more accurately your computer) busy over Christmas. We can phone and discuss if you want, just nominate a time. I am away in Geneva Wed to Friday this week inclusive so could not talk then. Sorry I can't be here on Thursday for a pint but I'm sure you will manage.

Best wishes,
Ian

Other bits from the manuscript itself :

Jo Lines/Andy conversation at TAG. Suggest a few graphs where just change one parameter at a time. Need to resurrect our default parameters and easy to do….include them on Table 5 along with Curtis default parameters in other columns.

Andy.
Can we include the synergism/antagonism/cross-resistance factors Λ in the code; see calibration section of methods, page 6 and Table S1

I think the classification trees each  need 10,000 rather than 1,000 runs. Don’t start clogging up your computer until we decide the final format.

Recreate the “mixtures are only better if last 20% longer” classification tree for Figure 3. I know you have already done it but the nodes all had technical symbols that are difficult to interpret. They need simple names like in Figure 3A. I assume its just a question of changing the parameter names in a “R” plotting script.


Run a new set of simulations using the “Curtis extended scenario” described on Page 16 to produce new classification trees again using the two criteria used previously i.e. (a) mixtures better if last >1 generation more than sequential (b) mixtures only better if last >20% longer

page 8 :
Our simulations will assume that the effects of insecticides in a mixture are multiplicative e.g. if the probability of surviving exposure to insecticide A alone is 0.3 and of surviving insecticide B alone is 0.2, then the probability of surviving expose to a  mixture of A and B is 0.3*0.2=0.06. In other words there is no synergy or antagonistic interactions between the insecticides. We do build flexibility into our methodology by scaling the fitnesses in niches where mosquitoes encounter both insecticides by a factor ΛAB, ΛAb, ΛaB or Λab (Table S1).  If insecticides are synergistic then Λ<1, but if insecticides are antagonistic, for example because they share the same target site or share a similar mechanism of (cross)resistance, then  Λ>1. This is a flexible approach but other strategies are also possible, for example if the insecticides are highly antagonistic then the fitness may be defined as the locus with the higher individual fitness factor, w.

Where would this 'synergism/antagonism/cross-resistance' bit need to go in code ?
scaling the fitnesses in niches where mosquitoes encounter both insecticides by a factor ΛAB, ΛAb, ΛaB or Λab (Table S1).  If insecticides are synergistic then Λ<1, but if insecticides are antagonistic, for example because they share the same target site or share a similar mechanism of (cross)resistance, then  Λ>1

in runModel2.r, probably Wniche[locus1, locus2, niche1, niche2]
Around line 300, just after this :
 ## Two locus fitnesses in two insecticide Niche

    for( nicheNum1 in 1:3 ) #todo get this 1:3 from somewhere
    {
      for( nicheNum2 in 1:3 ) #todo get this 1:3 from somewhere
      { 
        #temporary solution
        #to get both niche (one of 0aAbB)
        #and exposure (one of no,lo,hi)
        niche1 <- dimnames(Wniche)$niche1[ nicheNum1 ]
        niche2 <- dimnames(Wniche)$niche2[ nicheNum2 ]
        exposure1 <- dimnames(Wloci)$exposure[ nicheNum1 ]
        exposure2 <- dimnames(Wloci)$exposure[ nicheNum2 ]        
        
        #if this niche toggled off set fitness to 0
        if (niche[niche1,niche2] == 0)
        {
          Wniche[,,niche1,niche2] <- 0
        } else{
          #otherwise set fitness to product of the 2 loci
          for( locus1 in dimnames(Wniche)$locus1)
          {
            for( locus2 in dimnames(Wniche)$locus2)
            {   
              #6/1/16 i think ians new insecticide interaction parameter can just go here
              #does in need to be just one param or 4 ?
              #ΛAB, ΛAb, ΛaB or Λab 
              #Wniche[locus1,locus2,niche1,niche2] <- interaction * Wloci[locus1,exposure1] * Wloci[locus2,exposure2]
              Wniche[locus1,locus2,niche1,niche2] <- Wloci[locus1,exposure1] * Wloci[locus2,exposure2]
            }
          }          
        }
      }
    }


** 18/12/15 **
~ what do I need to do to get labels correct in all trees ?

Looking at the requests from Ian.
I think it would be better if he & I sat down to go through it before I spend time doing 1000s more runs.


** 6/1/16 ** on train to liverpool

Questions for Ian :
~ I recomend a few days, maybe a week, finishing the restructurng of the model. It still doesn't conform well to good practice.
done ~ we could go through the model itself, checking Ians code points.
hat ~ What can we call the insecticide interaction parameter see 290 in runModel2() ?
4 ~ Is it just 1 param or do we need 4 ?
no ~ should I remove W.bar arounf line 460 in runModel2()
~ can we go over the curtis extended scenario p16 ?

Linkage dis is bad in the context of mixtures because mosquitos are resistant to both insecticides more often than would be expected by chance. So problem is when the mixture is in use. LD decreases by half each generation under free recombination (recomb rate 0.5) . As genes get closer on chromosome recomb rate decreases.
 
Extended Curtis requests from Ian
1. 'Male exposure' as a proportion of female (0.5 to 1)
Exposure is held in array a, read from input file by sex, line 90.
    a['m','0','0'] <- input[8,i]
I can add a parameter to the setExposure() function which is called from sensiAnPaperPart.r


1. 'correct deployment' for mixtures (0.5 to 1) the proportion correctly exposed to the mixture, the rest are divided equally by being exposed to high concentrations of each individual each e.g. exposure 0.8 and correct deployment 0.7 : 0.8x0.7=0.56 are exposed to niche AB, while [0.8x0.3]/2=0.12 are exposed to niche A and 0.12 are exposed to niche B.
Again this can be added as a parameter to setExposure()


Main actions required from Ians last email :

1. Can we include the synergism/antagonism/cross-resistance factors Λ in the code; see calibration section of methods and Table S1. This is about how effect of >1 insecticide are combined. see my notes above for where this is in R code.
1. Run a new set of simulations using the “Curtis extended scenario” described on Page 16 to produce new classification trees again using the two criteria used previously i.e. (a) mixtures better if last >1 generation more than sequential (b) mixtures only better if last >20% longer
1. Graphs where just change one param at a time. Jo Lines/Andy conversation at TAG.
1. Recreate the “mixtures are only better if last 20% longer” classification tree for Figure 3. Replace technical symbols with simple names like in Figure 3A. (see sensiAnPaper1All.Rmd rownames(treeInput)[rownames(treeInput)=="a.f_AB"] <- "exposure")
1. once we have decided what to do increase sensitivity analysis to 5000 or 10000 runs & run new trees


We the run a “Curtis extended scenario ” to further illustrate the processes. The simulations have two additional parameters in addition to those described above and on Table 5 i.e.

•	A “male exposure” parameter which defines the proportion of males exposed to an insecticide niche as a percentage of the females; it is assumed to be the same for each insecticide niche and varies uniformly from 50% to 100%. So, for example if females are 20% unexposed, 56% exposed to ‘AB’ and 12% exposed to each of niches’ A’ and ‘B’.  If male exposure in 60% of the females their expose to niches ‘AB’,  ‘A’, and ‘B’ are  0.56*0.6= 33.6%, 0.12*0.6=7.2% and 0.12*0.6=7.2% respectively and the proportion unexposed rises to 0.2+(0.8*0.40)= 52%.

•	A “correct deployment” parameter in the simulations examining the use of mixtures. Correct deployment is often not achieved for a variety of operational reasons and exposure to insecticides may not be as anticipated. In particular, mixtures may be mandated but not achieved for various reasons (such as stock-outs) such that the ‘mixtures’ are often replaced by the single insecticide that is available.  We vary this factor uniformly between 50% and 100% and the incorrectly exposed mosquitoes are equally divided into exposure of the single insecticides. So, for example, if exposure is 80% and correct deployment” is 70% then within that 80% exposure to insecticides 0.8x0.7= 56% are exposed to niche AB, while [0.8x0.3]/2= 12%are exposed to niche A and [0.8x0.3]/2= 12% are exposed to niche B. In reality, poor deployment will result in exposure to lots of other niches i.e. a, b, Ab, Ab, ab but we only examine high levels of insecticide here to avoid having to additionally define survival probabilities to low insecticide levels ‘a’ and ‘b’.
 

** 9/1/16 ** on train back from liverpool game rollout workshop

modifying setExposure() function to cope with Ians request for lower male exposure
I can add it in as a param with a default of 1.

** 13/1/16 **
starting on the new 1 parameter at a time sensi analysis in sensiAnPaperPartOneAtATime.r
 
** 14/1/16 **
first see if I can make any informative plots with the existing sensitivity analysis data
 
These outputs are the frequencies by generation of the resistance alleles for the different insecticides (1&2) 
Gen, m.R1, m.R2, f.R1, f.R2
 
I want to get to a single file that has inputs & time_to_resistance thresholds for each scenario.

I must have something nearly like this already from sensiAnPaper1All.Rmd

Yes : resistPointsI1,2,Mix etc.
For each scenario these have gen_cP0.1,0.25,0.5 which is the number of generations taken to reach those resistance thresholds.

The next chunk works out which strategy is better than which.

Maybe I want to bind all the resistPoints together into one file and bind onto the inputs, so that we can look at how the inputs effect the time to resistance for the different strategies.

I want input on the x axis & time to resistance on the y. (but there will be the issue that the y values will be affected by the changing values of other parameters too) Cross that bridge when I come to it.

how to get data in format for ggplot2
check the diamonds dataset in ggplot2 
 
so I want to get as a dataframe with scenarios in rows, & columns for inputs & outputs

so I can do something like :
ggplot(diamonds, aes(x=carat, y=price, color=cut)) + geom_point()

If I have a column that is strategy (insecticide1, insectide2, sequential, mixture)
Then I would be able to facet by these later.

ggplot(diamonds, aes(x=carat, y=price, color=cut)) + geom_point() + facet_wrap( ~ cut)

This could be good as I can see the different strategies side by side.

So I want columns :
input1, input2 ... strategy, outputs : gen_cP0.1, 0.25, 0.5
 
so the dataframe will have length=num strategies*num scenarios

I'll need to rbind the strategies on top of each other.

Some success with ggplot plots.

** 15/1/16 **
sent first version of ggplot doc (effectOfIndividualInputs.pdf) to Ian & Beth

prettifying parameter names

** 20/1/16 ** on train to liverpool

removed from plots runs where resistance threshold not reached

** 22/1/16 ** with Ian in Liverpool
I said realistically it would be a month before I could get started on the

conventional wisdom of poor spraying increasing resistance is because poor spray thought not to kill heterozygotes. we see some o this effect in the dominannce param. Later we can loook at by the low exposure niches e.g. a vs A.
We can look at in the next paper.

Ian plans to submit to PLOS computationally biology with the aim that it's a flexible model that can address a bunch of situations. Do we want to be slightly carful to protect our IP ? e.g. in the first version w might not want to promote use by others (we'd need to do some more wrk to make it properly useable). I can discuss this with Ian when we are closer to publication.

~ I did think that correctDeploymentProp only effects mixtureswhich would constrain tree analyses for whether mix better than seq, but actually analysis can still be done, you would just expect mix to be closer to seq.

~ added maleExposureProp & correctMixDeployProp into tree & plot analyses, by adding on to the end of the input object. inputs are got from   treeInput <- listOutMix$input which is created by runModel2() just from the input object that is passed to it.

~ added as an extra argument to sensiAnPaperPart() that would make less code repetition



** 25/1/16 **
added to effectOfIndividualInputs.Rmd
~~ new predictors for males & deploy
~~ calc of new ratio, copy from sensiAnPaperAll 

13.40 started 1000 reps of the 'extended' experiment 
< 50 mins to run

This is what Susanan did for her 2012 model :

Simulations to understand the influence of each parameter
on the outcome variables (mean fitness and change
in resistance allele frequency) were performed using latin
hypercube sampling (LHS) to generate a data set and
partial rank correlation coefficients (PRCC) calculated to
provide a quantitative measure of the impact of each
parameter [24]. LHS techniques were first developed to
explore the behavior of complex models in economics,
engineering, chemistry and physics and have been used in
models predicting the impact of insecticide-treated nets
on malaria transmission [25].

The analysis was performed using R software [26] and
implementation of LHS using package lhs. It does not
allow for the specification of each variable distribution
beforehand, so sampling was performed assuming a uniform
distribution. Once the sample was generated, the
uniform sample from a column (variable) could be transformed
to the required distribution (Table 3) by using
quantile functions (using the qtriangle comand in R).

A data set of 3,000 replications was generated, with random
parameter sets and the corresponding values of the
outcome variables using equations 8, 9, 12 and 13. Ten
replicates of this procedure were performed as suggested
in [24] to investigate the predictive precision of model
using LHS as the sampling method. This was achieved
by analysing each replicate separately and verifying that
results were consistent across ten replicates.
 
** 26/1/16 **
looking into prcc analysis of model outputs &  whether I might want tio use lhs.

~ what is difference between lhs and what I have done
~ check lhs package 
~ lhs would ensure a more uniform distributions of samples across parameter space, i would think results are unlikely to be different to our monte-carlo approach.
~ it might be tricky for the params where we haven't used a uniform distribution (although we probably could get around)
~ some advice on using lhs package
http://r.789695.n4.nabble.com/latin-hypercube-sampling-td4659028.html

~ add PRCC code
This provides some (dated) advice on running a sensitivity anlaysis including LHS & prcc
https://cran.r-project.org/web/packages/pse/vignettes/pse_tutorial.pdf

pcc function in package sensitivity may do what we want :
https://cran.r-project.org/web/packages/sensitivity/sensitivity.pdf

pcc computes the Partial Correlation Coefficients (PCC), or Partial Rank Correlation Coefficients
(PRCC), which are sensitivity indices based on linear (resp. monotonic) assumptions, in the case of
(linearly) correlated factors.

~ or could use epiR package
epi.prcc(dat) # where final column in dat is the output, all other columns are inputs


I got sensitivity::pcc() working for a single strategy.


** 28/1/16 **
~ fixed diagonal axis labels.
~ get sensitivity::pcc() working for all strategies & putting data into format for faceted plots. 
 
** 29/1/16 **

~ why does correctMixDeployProp seem to have an effect in the PRCC of insecticide1 when it shouldn't have any effect outside of mixtures ?

In the plots againts time_to_resistance there is only a bit of a pattern for Mixture1 (time_to_R increaeses between 0.9 & 1 correct deployment).
Points to it being a problem in the PRCC itself.

~~ in setExposure() looks fine, its not applied to insecticide1 or 2

other things it could be :
~~ switched with another var in the call to setExposure
~~ something about how runs are done in sensiAnPaperPart.r 
~~ rows getting reordered within effectOfIndividualInputs.Rmd ?
 
it is +vely correlated with time to resistance, so better application leads to longer times for resistance to be reached (not sure whether that is waht I would expect ?) 
 
aha PRCC results are influenced by the presence of the other vars, if just do for exposure & deployProp deployProp goes down to 0.2. 
x2 <- x[c('correctMixDeployProp','exposure')]
pcc_res2 <- pcc(x2, y, rank=TRUE)
pcc_res2$PRCC
                       original
correctMixDeployProp  0.2167706
exposure             -0.5248086 
 
I still don't quite understand why an essentially random var is correlated with the output ??

This shows that there's very little correlation between the in & out, for insecticide1
> cor(xy$time_to_resistance0.25,xy$correctMixDeployProp)
[1] 0.003849606

** 1/2/16 **

Checking again on whats happening with correctMixDeployProp

is the correlation similar for mixture1 as it is for insecticide1 ?

could problem be that the inputs are got from a mix run ? shouldn't matter
  treeInput <- listOutMix$input
 
could it be something to do with the random number seed being disrupted by the extra params ? shouldn't happen

i think the inputs should be identical. 

I think it may be a correlation in the analysis.

What I have done is generate inputs for 'male proportion' and 'correct mix deploy'.
Correct mix deployment is not used in the individual insecticide runs, but it is still included as a potential predictor in the PRCC analysis.

In the mixture scenarios 'correct mix deploy' effects exposure ...

exposure is not a single variable in the simulation, but is it used as a single value as a predictor in the sensitivity & prcc analyses ?

aha the exposure predictor is taken from the female AB (actually I think that is OK ??)
although maybe not because it will be reduced by correct deploy and maybe by male (or maybe not ???)

  rownames(treeInput)[rownames(treeInput)=="a.f_AB"] <- "exposure"

correctDeploy has a strong interaction with exposure in mixtures.

When correct deploy is low in mixture runs exposure to AB is much lower because of corresponding exposure to A0 and 0B.

But i think this shouldn't effect single insecticide runs.

Am I sure I'm not mixing up rows somewhere in effectOfIndividualInputs.Rmd ??

I could try setting the values of results for I1 and see if it propogates through.

Instead I saved the original random value of exposure and used that in PRCC.

I think what was happening was that correctDeployProp influenced the AB exposure value for mixtures and that was the predictor I took for exposure. Instead I want the intended exposure value.

I still don't quite understand how this effects the results in the single insecticide runs.

but with my 100 test runs what I'm seeing in the PRCC is no effect of anything for insect1 & 2. and a similar pattern as before for mixture1.

This may point me towards another problem that is causing this.
I may need to exclude corectDeploy as a predictor from those runs in which it wasn't used ??


** 2/2/16 **

~ ran 10000 runs over night
~ run effectOfIndividualInputs
~ now it's fine !!! PRCC looking good.

New trees from 10,000 : 
Predictors :
effectiveness_insecticide1 :9
effectiveness_insecticide2 :9
exposure :                  7

dominance_allele2 :         3
dominance_allele1 :         4
resist_start_1_div_2 :      1
start_freq_allele2 :        1
start_freq_allele1 :        2 
 
Best tree to show (if indeed trees are the best) : 
Can't find a best one ! I don't like these trees !!
Try to find a better method for Ian. Could it be PRCC.
 
Violin plots. look good. They are very similar for time to resistance 0.1, 0.25 or 0.5. 
~ need a plot of proportion of runs in which mixture or sequential better. Maybe I could do a violin plot of the time to threshold, with diff strategies on x axis. 

~ faceting them by exposure is cool
~ faceting by effectivity shows very little
 
 
** 5/2/16 **

trying to get plots working

## GOOD RESULTS ! :

###1 effectiveness < 0.6 sequential best. > 0.6 mix best. 
Plot of time_to_resistance against effectiveness  with 4 strategies on the same plot is key.

It shows the line for mix2 crossing that of sequential. Effectiveness 0.4-0.6 resistance arises faster for mix2. > 0.6 resistance arises faster for sequential.

I should adapt this plot perhaps just to show mix2 and sequential.

###2 selection coeff < 0.2 sequential best. > 0.2 mix best.
selection coefficient curves also cross for mix2 and sequential. < ~0.2 sequential is better. > 0.2 mixture is better. 
But check that this is multiplied as shown below. So beware could this be caused by the effectiveness result ?
    ##effectiveness
    phi.SS1_A0 <- runif(1, min=0.4, max=1)
    phi.SS2_0B <- runif(1, min=0.4, max=1)    
      
    # Ian suggested this should be dependent on effectiveness to ensure fitness of RR stays below 1 
    s.RR1_A0 <- runif(1, min=0.2, max=1) * phi.SS1_A0
    s.RR2_0B <- runif(1, min=0.2, max=1) * phi.SS2_0B

I think perhaps I should remove multiplication of selection coefficient by effectiveness ?


###3
Effect of exposure is greater overall on times to resistance, but it has a more similar effect on mixtures & sequential thus it is of less importance when considering the differences between the two strategies.


In majority of cases the point where resistance to just one of the insecticides in a mixture is reached comes before that for both insecticides in sequential use. [this seems obvious]
Also it always takes longer to reach resistance for both insecticides when starting with both in a mixture. Is this obvious, or a meaningfull result or an error in our methodology ?

I think for the paper we should focus on mix2 versus sequential. We could have that as what we call mixture, and then refer in smallere analyses to the other mix scenarios.
Maybe I should have a quick look at curtis paper to remind myself ?

Tried removing multiplication of selection coefficient by effectiveness.
But got warnings about fitnesses > 1.

 
6/2/16

problem is with locus fitness values going over 1

This is the problem that causes it :

Wloci[ paste0('RR',locusNum), exposure] <- (1 - phi[locusNum, exposure]) + 
                                                   (s[locusNum, exposure])

so it just occurs when s > phi, selection coeff > effectiveness
 
current range :
effectiveness : 0.4 - 1
selection coeff : 0.2 - 0.5
 
Options : 
1) discard sims in which selection > effectiveness
2) set selection range to 0.1-0.45 and effectiveness from 0.45-1

Doing 2) might be good to check whether the line crossing still goes on ... 
 
Yes saving 2) as listOutMix_ex2_1000.rda etc.
 
 Done in effectOfIndividualInputs



** 7/2/16 **
with new runs 

Selection coefficient curves for mix2 & sequential no longer cross. Seems likely that this previous pattern was generated by correlation with effectiveness.


** 9/2/16 **
sent new figs of PRCC of diff between seq & mix2 to Ian

** 10/2/16 **

~ reran trees they show that effectiveness > ~ 0.7 & exposure < 0.7 favours mixture2 over sequential
~ this is consistent with my graphs of difference vs effect & exposure (therefore are the trees really necessary ?)

mix2 can already be displayed in plotcurtis_f2_generic() using addCombinedStrategy
I had just set default to F because the plot gets even more confusing.

Maybe add option to just plot mix2 ?

Also offer option to plot from : runcurtis_f2()

This which simply reduces effectiveness of insecticide2 from 1 to 0.85, makes seq better than mix (where mix was better for curtis).

runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 0.85 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

** 11/2/16 **
~ added default curtis values to UI labels in server.r
~ fixed plotting of combined strategy, i think I may not be doing it right 
 
** 12/2/16 **
~ added position of Curtis inputs to the figures

** 18/2/206 **
Response email from Ian :

Anyway I think the plots all look very nice, and give intuitive answers (but see point #1 below). My comments are
 
(1)    It is very strange that “correct deployment” parameter has no impact on time until resistance in mixtures.

(2)    Re your question at the end of the document i.e. “In mixtures what causes resistance to the 2nd insecticide to arise more slowly when the first insecticide is removed from the mixture ?”. Isn’t it the opposite i.e. if we compare Mixture 2 (where the first insecticide is removed) then doesn’t resistance spread more rapidly compared to when the first insecticide is retained (as in mixture 3)??

(3)    This question stimulated me to look again at your definitions of mixtures etc on the front page of the document. I know its only a draft and we both know what it means but I took the liberty of expanding the definition as in the attachment. Can you check they are correct and hence that I have correctly understood your question addressed in the previous point (and, if so, you may want to cut-and-paste them into the document). I also suggested a potential change (i.e. that mixture 1 and 3 be combined because they only differ in their output measure). This would clarify the captions on the plots but it just depends how easily you could cut-and-paste such changes.
 
One other thing that has just occurred to me. One option not considered by Curtis, but commonly proposed now, is to deploy  “mosaics” where insecticides are deployed singly but at the same time…so some fields/huts may have insecticide “A” while some will have insecticide “B”. Could we simulate these mosaics simply  by setting the “correct deployment” parameter to zero?? It would be  another good selling point for the manuscript. If it is straightforward to set up, I would do it, If not we could just note how we could do it in the ms. It really follows on from point #1 above… if correct deployment has no effect even when it reaches zero then it implies mosaics are as good as mixtures.

Ians recomendations for captions :
Single use of either insecticide. The outcome measure is time for resistance allele frequency to reach the user-defined threshold value (0.25 in these case)  

Sequential : sole use of one insecticide, switch to other when threshold reached. Outcome measure is the time until frequency of resistance to the second insecticide reaches the threshold value.

Mixture 1: Insecticides always used as a mixture. Outcome measure is the time until threshold reached for either insecticide i.e. the frequency of resistance alleles exceeds 0.25 at either locus

Mixture 2 : insecticides initially deployed as mixtures until the threshold is reached for either insecticide, then switch to sole use of the other insecticide. Outcome measure is time until resistance allele frequencies exceed the threshold value for the second insecticide
 
Mixture 3 : Insecticides always used as a mixture. The outcome measure is the time until the threshold allele frequency exceeds the threshold value at both loci.

Actually, it would be much better if we combined mixture 1 and 3 and just note the difference in outcome measure. Mixture 3 could be renamed “Adaptive mixture”. In the figures the key could then be

Single
Sequential
Mixture (either AI)*
Mixture (both AIs)
Adaptive mixture

* I assume you know this but AI is short for “active ingredient” which the commercial developers much prefer to “insecticide” for reasons that elude me


Which would be much easier for readers (and reviewers) to remember


**26/2/16 **

look at Ians point :
Re your question at the end of the document i.e. “In mixtures what causes resistance to the 2nd insecticide to arise more slowly when the first insecticide is removed from the mixture ?”. Isn’t it the opposite i.e. if we compare Mixture 2 (where the first insecticide is removed) then doesn’t resistance spread more rapidly compared to when the first insecticide is retained (as in mixture 3)??

Resistance spreads more rapidly when the 2nd insecticide is removed from a mixture. See Fig 1.

Perhaps I should rename all references to mix 1,2 & 3 to avoid confusion of getting tangled up with this again.

Sequential                                     or seq
Mixture (either AI)   [mixture1] or 1st in mix or mix1
Mixture (both AIs)    [mixture3] or 2nd in mix or mix2
Adaptive mixture      [mixture2] or Adaptive mix or comb 
 
 
Hi Ian,

Firstly, yes you are right. Resistance spreads more rapidly in the responsive mixture strategy than when the first insecticide is retained.

I agree that we need to rationalise the names to try to minimise this confusion. I would prefer to not use AI in labels if possible, but we can refer to in text. (but we can discuss).

If we could agree consistent long, medium and short names as started below that would be very helpful. I can then go through my scripts and rename. I'll have to be very careful because I will need to rename mix2 to mixA & mix3 to mix2. So if I miss any of the first mix2s we'll get in a right pickle.

Also can we agree on what to call the strategy that changes from mix to sole ? My fault that I've been variously calling it adaptive, responsive, combination. Is there a precedent ? Adaptive seems potentially confusing as it could be confused with being selected for. I have a slight preference for responsive. 

Feel free to edit these below. Once we are agreed I'll print out and bluetac to my wall so that I stick to it !!

Long Names (in text where space not an issue)
Mixture (1st insecticide)   
Mixture (2nd insecticide)
Adaptive mixture (2nd insecticide)
Sequential (2nd insecticide)

Medium Names (plot labels where medium space)
mix 1st              
mix 2nd             
mix adaptive
sequential 2nd           

Short Names (plot labels where v little space)
mix1
mix2
mixA
seq

cheers,
Andy 
 
** 29/2/2016 **

checking on correctMixDeployProp after Ian commented :

(1)    It is very strange that “correct deployment” parameter has no impact on time until resistance in mixtures.

In the plots the only effect was on mix1 that seemed to rise slightly.

in setExposure() this is how it is done :

    if ( correctMixDeployProp < 1 )
    {
      a['f','A','0'] <- (1 - correctMixDeployProp)/2 * exposure
      
      a['m','A','0'] <- a['f','A','0'] * maleExposureProp
      
      #m&f together
      a[,'0','B'] <- a[,'A','0']    
      a[,'0','0'] <- 1 - (a[,'A','B'] + a[,'A','0'] + a[,'0','B'])  
      
    }
    
Code seems as expected.

Could anything about how the sequential & adaptive mixture scenarios are patched together influence this ?
I don't think so. 

#' @param correctMixDeployProp proportion of times that mixture is deployed correctly, 
#'    assumes that when not deployed correctly the single insecticides are used instead

~ Thinking about correctMixDeployProp, how do we choose which of the 2 insecticides is used singly ? It is divided between them. Maybe this is the issue ? Perhps splitting between the 2 single applications in the same simulation has a similar effect as the 'mixture' ?

~ maybe write a function to produce a graphical output from setExposure() could put it into a UI to send to Ian.
~~ effectively it just needs to visualise the array a.
~~ I could firstly vis as a table rather than a graph
~~ outputs : 2 tables side by side, a["f",,] & a["m",,]
https://andysouth.shinyapps.io/shinyNiche/ 
 
~ checking on correctMixDeployProp
~ added correctMixDeployProp into curtis shiny UI to enable checking mechanism
~~ first added correctMixDeployProp into runcurtis_f2 
 
 
~ carefully implement strategy name changes
Find & replace todo : in sensiAnPaper1All.Rmd & paper1_results_figs.Rmd

resistPointsMix2    resistPointsMix_A
resistPointsMix3    resistPointsMix_2
resistBetterMix2SeqBoolean  resistBetterMix_ASeqBoolean
resistBetterMix2Seq         resistBetterMix_ASeq

#### replacements done :
Mix1 - Mix_1
Mix2 - Mix_A
Mix3 - Mix_2
"Mixture 1" - "Mix either"
"Mixture 2" - "Mix adaptive"
"Mixture 3" - "Mix both"
done mix2_minus_seq0.25          mixA_minus_seq0.25 
 
** 2/3/2016 ** 
 
done ~ record time spent & when was the last time I claimed on the previous contract ?
Previous contract claimed 15 Sept 2015.
submit invoice : from contract : maximum of 25 days at 217 per day
I spent 30.6 days (can claim some against tsetse work, or from future contract ...)
So time I spent in January on toggl should be on new contract/
To the end of Feb :
Hours		105:45:00
Hours rounded		105.00
Days (/7.5)		14.00
Days remaining (from 85)		71.00 
 
~ modifying PRCC mutliplots to get just one axis 

** 3/3/2016 **  

msg from Ian :
Aha! I've understood what he means by setting correct_deploy to 0, to study mosaics.
~ changed order of strategies in legend
~ created new killer figure with added effectivenesses on the y & exposure on x.  
 
** 8/3/2016 **
~ add Curtis fig2 params to killer figure
~ tricky adding to legend or labelling
~ can I get legend for killer figure beneath ?
~ can I derive a line for the border between when seq & mix better by fitting a line to the 0 values ?
~ get the classification tree that matches my killer figure

** 9/3/2016 **
~ creating 20% tree & plot
~ got the 0 line onto the first facet of the 20% plot

** 10/3/2016 **
~ looking at the manuscript

I think the manuscript can be about describing development and results of a model that is a more generic version of Curtis. But we do not need to make the model useable by others at this stage. 

** 11/3/2016 **
manuscript

current wordcount 9,300

Details from Plos Computational Biol :
Software Submissions
PLOS Computational Biology publishes articles describing outstanding open source software of exceptional importance that has been shown to provide new biological insights, either as a part of the software article, or published elsewhere.

The software must already be widely adopted, or have the promise of wide adoption by a broad community of users. Enhancements to existing published open source software will only be considered if those enhancements bring exceptional new capabilities. The software must be downloadable anonymously in source code form and licensed under an Open Source Initiative (OSI) compliant license. The source code must be accompanied with documentation on building and installing the software from source, as well as for using the software, including instructions on how a user can test the software on supplied test data. Software articles require a presubmission inquiry that includes explanations on how the above criteria are met.

Format
Articles should be concise (less than 3500 words, not including supplementary material) 

Sentence on how the decisoin tree level decided.

A rule of thumb is to choose the lowest level where the rel_error + xstd < xerror
xerror is the cross-validation error

To avoid overfitting the number of levels was restricted to the minimum where the relative error plus the standard error was less than the cross-validation error.

added mix both to paper1_results_figs.Rmd (because of my slight uncertainty in mix adaptive method)
repeated all figs for 50% resistance in paper1_results_figs_slimmed_50.Rmd

** 14/3/16 **

Curtis contentious statement p261 :
"Example vishows the, at first sight unexpected, point that the use of a mixture where the initial gene frequencies are unequal leads to a more rapid increase in the rarer of the genes."

vi params ?
exposure 0.9
strt freq loc1 0.01
strt freq loc2 0.001

modify fig2 by setting freq1 from 0.01 to 0.001

runcurtis_f2( P_1 = 0.001 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 , correctMixDeployProp = 1 )

when I changejust freq loc1 to 0.001 I get an error :
Warning in min(which((combmat[, r1col]) > max(criticalPoints))) :
  no non-missing arguments to min; returning Inf

Trying to work it out with this run :
runcurtis_f2( P_1 = 0.01 , P_2 = 0.003 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 , correctMixDeployProp = 1 )

in plotcurtis_f2_generic problem is here :

  cutoff_i1_mix <- min(which((combmat[,r1col])>max(criticalPoints)))
  cutoff_i2_mix <- min(which((combmat[,r2col])>max(criticalPoints)))  
  
  if (cutoff_i1_mix == Inf | cutoff_i2_mix == Inf)
  {
    warning("critical point not reached for i1 or i2: cutoff_i1_mix=",cutoff_i1_mix," cutoff_i2_mix=",cutoff_i2_mix)
    #don't add the combined(adaptive) strategy to the graph
    addCombinedStrategy <- FALSE
  }


in plotcurtis_f2_generic this calculation at the start seems to produce diff results than what they appear in the plot :
  maxGensMix <- findResistancePoints(combmat, locus='both', criticalPoints = criticalPoints)

maxGensMix
           [,1]
gen_cP0.1   294
gen_cP0.25  999
gen_cP0.5   999

problem partly caused by max_gen default in run_curtis_f2 being 300

runcurtis_f2 <- function( max_gen = 300,
whereas in sensiAnPaperPart max_gen is set to 500

probably I should set max_gen in run_curtis_f2 to 500 too so it is the same.

yes, now this works, but P_2 = 0.001 still fails to reach threshold
runcurtis_f2( P_1 = 0.01 , P_2 = 0.002 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 , correctMixDeployProp = 1 )


** 15/3/16 ** on train to liverpool

Looking at the form of the results from modified curtis fig2 plots :
~ in mixtures resistance in the final insecticide only rises after the 'first' one has become ineffective ~50%

~ in curtis scenario reducing start freq of I2 from 0.01 to 0.002 makes it take longer but doesn't really alter the form of curves
~ but then changing effectiveness of I2 from 1 to 0.95 makes a big difference.
~ aha! seems that an effectiveness of 1 in one insecticide stops resistance rising in the other one until ~50% resistance. Reducing effectiveness to even 0.95 (or just 0.99) seems to remove this effect. At effectiveness I2 0.99 resistance to I1 rises slowly until 50% is reached then it accelerates.

~ increasing dominance from 0.0016 to 0.02 doesn't radically alter curtis2 pattern


** 16/3/16 ** meeting Ian in liverpool

paper todo
~ rename fig x1 y axis to "time in generations" or time (generations)
~ rename y axes in fig x3 "time in generations" or time (generations)
~ rename maleExposureProp to male_exposure
~ rename correctMixDeployProp to correct_mix_deploy
~ set variable names to be the same as above in table 5 of manuscript
~ check again that correctMixDeploy is doing what we expect. Bit strange that in the plot against time-to-r the curves are completely level when in my curtis simulation it shows that decreasing the correct deploy does decrease time to resistance e.g. for mix2 from  
done ~ change fig x3 to just exposure, effectiveness and dominance
~ add diff between seq and standard mixture to fig x4
~ Ian will send me the params to rerun the line from table vi (which is just for 1 generation)
~ find the equation line of zero difference for the 0.5 threshold (and maybe for mix2 vs seq too ? )

~ for the 20% diff plot add the line that exactly 20% diff and keep colour scheme from killer plot
~ see df_20pc
~ how to make sure that the same points apear in each plot ?? maybe remove the facet column.

Qs for Ian
~ is effectiveness of 1 possible/likely ? The pattern in the curtis plot relies quite heavily on it. Reducing it to just 0.98 changes pattern considerably. 
~ are we confident enough about the mix adaptive strategy to focus the paper on it ? Ian said yes. He has said before that it just takes a couple of generations for frequencies to return to Hardy Weinberg.

Remember that effectiveness is proportion of the SS genotypes that die when exposed.

Ian said, science is unsure what the effectiveness rates ofr insecticides are under field conditions. all sensitive mosquitos ar expected to die in a bottle assay. BUt under filed conditions where they briefly land on a net this may not be the case. experimental huts.

IVCC are apparently asking if we could model 3 insecticides.

Ian keen to keep sex-linkage & niches in. Say we have developed a general model and say we will explore more behaviour in future ...

~ asked Ian where could I submit a paper explaining this resistance thing in simpler terms for dummies. If I want to get into anything sciency a first author paper would really help me out. It could concentrate just on the mechanism of how the model works rather than on the sensitivity outputs. It could have plot combinations, showing how the curves change in response to 1 param changing. I could start writing it now. It could go to ecological modelling or similar epidemiology journal (maybe ask Jo Lines).
~~ concise
~~ focus on mechanisms
~~ for non expert audience
~~ dominance, exposure
~~ draft something
~~ ian suggested malaria journal
~~ paper2_resistance_mechanisms_mixtures.Rmd


** 22/3/16 **
comments back from Ian on paper1

One other thing for your "to do" list. I have made the point that Fig 4 (your x2) analyses the data summarised in Fig 3 (your x1).. which helps the flow of the paper. But we only present PRCC for 4 of the data shown in Figure 3 so it may worthwhile doing PRCC for the missing one i.e. time to resistance to first insecticide in sequential use. I assume that would be relatively easy to do and automate in the file?


Fix x1
I did Time to resistance (gens)~ Y axis label change to  "Time (gens)"
done ~ X axis labels as in the figure legend i.e.  "Mix 1st"  "Mix both",  "Seq 1st" and "Seq both". I think "both" is better than "second" as it avoids any potential confusion between insecticide labelled 1 and 2.
done ~ I think we decided to re-labelled "sole use" to "Seq 1st" for clarity
done ~ You might also like to change "Adaptive mix" to just "Adaptive" to try and save some space which I think will become important later as graphs are scaled down to their published size.


done ~ Fig x2.. can we make the labels identical to those above i.e., in order  "Seq both"  "Mix 1st"  "Adaptive both "   "Mix both".... A reviewer will wonder why we don't do all 5 outputs described in Fig x1 i.e. whey we omit "Seq 1st"
We can make it clear we are looking at the most likely application strategies so we only show those when resistance reached for both. (largely it's a space thing and the others don't show anything different).


Fig x3...Y axis label change to "Time (gens)".

Fig x5 and x6....
Can we change the Y axis label to just "Effectiveness of insecticides"... drop the word "added" as it is a bit ugly, makes the label longer and we now say it the sum of effectiveness in the figure legend.
Can we change the X axis label to just "exposure"... do we need to say it is for "both insecticides"????


** 23/3/16 **
email from Ian

We extensively look at how best to deploy 2 new insecticides, but not how to deploy one new insecticide when several others are in use and there is resistance to them….
 
This Curtis argument was not about sequential vs mixtures (i.e. where we start off with 2 insecticides and decide how best to deploy them). It is simply this questions: “If I have a single new insecticide where resistance is presumably present at a very low frequency, should I deploy it on its own (sole use), or in combination with an existing insecticide whose resistance levels are high”; 

Curtis defines allele B as the one encoding resistance to the new insecticide and which therefore has a lower starting frequency and asserted that, under these circumstances, deploying the new insecticide as a mixture would be counterproductive for allele B i.e. would drive resistance more quickly than when insecticide B was deployed alone.

Can I show this by modifying the fig2 plot ?
Although probably we would need to rerun a sensitivity type anlysis if we wanted to address this properly ...

curtis f2
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

curtis f2 with i1 set to high starting freq (0.5 breaks plot slightly, but still seems to show that resistance for i2 takes longer in mixture)
runcurtis_f2( P_1 = 0.5 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

curtis_f2 with P_1 changed from 0.01 to 0.45 : mix slower than seq
runcurtis_f2( P_1 = 0.45 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43, addCombinedStrategy=FALSE )

curtis_f2 with P_2 changed from 0.01 to 0.45 : mix slower than seq
runcurtis_f2( P_1 = 0.01 , P_2 = 0.45 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43, addCombinedStrategy=FALSE )

similar with P_2 0.3
runcurtis_f2( P_1 = 0.01 , P_2 = 0.3 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43, addCombinedStrategy=FALSE )

but what about the params for curtis_tab1_ex_iv ?
Beths analysis table uses the same params I think.

Hi Ian, 

As a first step I've added an extra fig x8 onto the end of the figures document attached and in dropbox (I started to address figure tweaks, but have held fire on that while we sort this). Fig x8 addresses whether to keep using an old insecticide when a new one is introduced. It just increases the starting frequency of each resistance in turn to just below the threshold (0.49). I think this accomplishes the same as Beths table 4.

Would I be able to alter the parameters to recreate Beths table3 and Curtis table 1 example vi in a similar way ? I couldn't quite work out which inputs to change.

How would we investigate this in a sensitivity analysis ? Essentially we could repeat what we've done already but with unequal starting frequencies. We had the issue before of not wanting one frequency to be dependent on the other. So we could simply set P_1 to a low range as in current analysis, and then have P_2 at a higher range (e.g. 0.3-0.5). Or perhaps we don't need a range for the 'old' insecticide we could just set at 0.5 ?

I'm slightly nervous about running more analyses but I could investigate this with a couple of days work. Then we would need to decide how to squeeze into the paper.

Happy to chat,

Andy

### previous current state of play 23/3/16
1. sensiAnPaper1All.Rmd : code to run whole sensitivity analysis & produce plots 
1. sensiAnPaperPart.r   : sets up param values and runs for one treatment (e.g. mixture, insecticide1)
1. runModel2.r          : runs model for the passed scenarios
1. paper1_results_figs_slimmed50.Rmd : final figures for paper

What would I need to do ?

1. sensiAnPaperPart.r : set new param ranges for P_2, perhaps add an argument to enable running them 
1. sensiAnPaper1All.Rmd : rerun calling new version of sensiAnPaperPart, save results to something like : listOutMix_unequal_10000.rda
1. paper1_results_figs_slimmed50.Rmd : save with new name, just rerun using the new results files

perhaps do for 500 runs first ?

reran new unequal scenarios with start_freq set to constant 0.5.
first plots OK, but then get a weird thing where mixa_minus_seq0.5 is always -2. So mix adapative is always taking 2 generations less than sequential. I suspect this may be an artefact of detecting resistance thresholds when the threshold is already reached for i2.

~ created pap1_figs_slimmed_50_unequal.Rmd
~ all changes marked with #4unequal

Seems that resistance is slower for mixtures in all replicates.
For the 500 runs at least.

** 23/3/16 **
email from Ian

I have been racking my brain to think if we can address this question using  data that have already been generated (and hence avoiding having to explain how we generated new data) . Rather than trying to explain it informally in an Email, then write it up, I went directly to step 2 and wrote the following as a draft paragraph for the ms…. Does it make sense???
 
“We can analyse a subset of the sensitivity analysis to investigate this specific question of whether single  use of a new insecticide is better than deploying it as a mixture with a pre-existing insecticide which already has some resistance present in the population.  We selected runs where the starting frequency of resistance to insecticide 2 was > 10 fold higher than frequency of resistance to insecticide 1. We then compared the time for insecticide 1 resistance to reach 50% when deployed on its own, compared to time for resistance to insecticide 1 to reach 50% when deployed as a mixture with the pre-existing insecticide. As in the full analysis, we compute the ratio mixture/single for each run and generate a classification tree to see what circumstances favour what policy and, in particular whether the ratio of resistance starting frequencies has a large impact”
 
I am assuming insecticide #1 is always deployed first in the sequential runs so the time to its resistance reaching 50% is already recorded. I am also assuming that in the mixture runs you have recorded  the time until resistance to each insecticide reaches 50% so you always have the data #1.

this is the important bit from Ian :
We selected runs where the starting frequency of resistance to insecticide 2 was > 10 fold higher than frequency of resistance to insecticide 1. We then compared the time for insecticide 1 resistance to reach 50% when deployed on its own, compared to time for resistance to insecticide 1 to reach 50% when deployed as a mixture with the pre-existing insecticide.

I think it would be easier to do this by new runs, which is close to what what I've done in pap1_figs_slimmed_50_unequal. (although I might want to change the freq of the 'old' insecticide from a constant 0.5 to a range e.g. 0.2 - 0.5, but then would).

Could I practically do what Ian has suggested ? 
How many runs are there where the start freq of i2 is > 10 * i1 ?
Where resist_start_1_div_2 < 0.1

see in paper1_results_figs.rmd

note my hi_div_lo was actually lo_div_hi, corrected in the figs_slimmed_50.rmd

resist_start_hi_div_lo <- ifelse( treeInput['P_1',] < treeInput['P_2',], treeInput['P_1',]/treeInput['P_2',], treeInput['P_2',]/treeInput['P_1',])

aha! had a thought. Can I highlight those runs in the killer plot where resist_start_1_div_2 < 0.1 ?
probably need to do it for the mix2 graph rather than mixadaptive.

I tried and it didn't show the split I expected perhaps because the higher start resist isn't high enough.

** 29/3/16 **

spoke to Ian on the phone. He's still keener on reanalysing the current sensi results.  Because he said its more relevant where the existing insecticide has a resistance level of 5-10% (or even less).

this is the important bit from Ian :
We selected runs where the starting frequency of resistance to insecticide 2 was > 10 fold higher than frequency of resistance to insecticide 1. We then compared the time for insecticide 1 resistance to reach 50% when deployed on its own, compared to time for resistance to insecticide 1 to reach 50% when deployed as a mixture with the pre-existing insecticide.

probably similar to :
findResistancePointsMixResponsive()

Effectively I think this is just comparing mixAdaptive with mixBoth for the subset of scenarios.

Aha no actually isn't it comparing sole use with mixture 'new' insecticide, the different bit will be that its not quite the same as the current mix1 & mix2. Because I don't think we care about the resistance level of the 'old' insecticide.

Maybe
1 create a function to calc when resistance is reached for new+old
2 create a 2nd func to calc when resistance reached for new on its own

1) findResistancePointsOldPlusNew <- function( listOutMix,
                                               listOutI1,
                                               listOutI2 )
                                               
~ find whether I1 or I2 has lower starting freq this is the 'new' insecticide
~ if new==I1 get from mix time til resistance for I1
~ if new==I2 get from mix time til resistance for I2                                               

2) findResistancePointsNewOnly <- function( listOutI1,
                                            listOutI2 )
                                               
~ find whether I1 or I2 has lower starting freq this is the 'new' insecticide
~ if new==I1 get from I1 time til resistance for I1
~ if new==I2 get from I2 time til resistance for I2  

For each get the outputs as 

then create a dataframe : dif_mixnew_sole


** 30/3/16 **

~ rather than doing as above I can use the existing files for I1 & I2
  resistPointsI1 <- findResistancePoints(listOutI1, locus=1)
  resistPointsI2 <- findResistancePoints(listOutI2, locus=2)
~ & then calc
  resistPointsMixI1 <- findResistancePoints(listOutMix, locus=1)
  resistPointsMixI2 <- findResistancePoints(listOutMix, locus=2)

Then post-process from these files for when starting freq of I2 >10*I1

A future thought. If I bit the bullet and allowed the simulation to represent adaptive strategies things might be much easier ? Also in future could I allow the effectiveness of insecticides to change over time, similar to how we have done in the game ?

doing this in paper1_results_figs.Rmd

now have :
dif_oldnew_newonly["mix_minus_sole0.5"]
 
BUT this is still for all runs (i.e. not just the ones where start freq of the old is > 10 * the new) 
 
** 11/4/2016 **

I think I want to plot this against each input variable to check whats going on, i.e. where is the mix better than sole use.

dif_oldnew_newonly["mix_minus_sole0.5"]

see Fig xx3

In most (70%) runs mixture are better, but that does leave 30% where sole better.

Eyeballing the plots suggests that exposure is still the most influential input. At high exposures the benefit of the mixture is less.

May be something weird going on with start_freq_allele2, noticeable lack of runs where mix-sole < 0 as start_freq_allele2 gets higher ... 

Are the inputs relating to allele1 & 2 still meaningfull with the post-processing that I do ?

No.
aha! problem I was having is that I1 & I2 inputs are not swapped around.

  #might be better just to restrict to the runs in which I2 < I1
  resistPointsNewOnly <- resistPointsI2_T[indicesI2_lessthan_I1, ]
  resistPointsOldPlusNew <- resistPointsMixI2_T[indicesI2_lessthan_I1, ] 

when I decided to restrict to runs where startfreqI2 < I1 mix seems to be better all the time ??

? am I doing something wrong ?
seems not

Eyeballing fig xx3

when I1 is the old :
Most important inputs & their effect on the benefit of mixture are :
exposure : reduces benefit of mixture
effectiveness i1,old : increases benefit of mixture
dominance i2,new : reduces benefit of the mixture
selection coefficient i2 : reducces benefit of the mixture 

My thoughts on mechanisms.
The mixture does better because the old insecticide kills some mosquitoes that have developed resistance to the new, thus reducing rate of increase in resistance to the new. 
Exposure. Low exposure reduces the benefits of mixtures. This could be because the beneficial effect of the old insecticide is reduced.
Effectiveness of the old insecticide. High effectiveness increases the benefit of the mixture because more of the mosquitoes that have developed resistance to the new insecticide are killed.
Dominance. High dominance of the new insecticide allele decreases the benefit of the mixture. High dominance in resistance to the new insecticide should allow more mosquitoes to escape the new insecticide therefore I would expect the benefit of the mixture to be greater. But its the other way around.

I probably need to get straighter in my head how dominance works.
Does dominance effect the relative performance of the 2 insecticides where mosquitos are exposed to both. i.e. Does high dominance of the new allele mean that a mosquito is more likely to survive the presence of the old insecticide ?

To understand the mechanisms more I could look at the effects of the inputs on each strategy on its own.

These results are consistent with previous ones.

~ ran PRCC on the new results (confirmed same important params)

copied new bits into keep_old_insecticide.Rmd and sent the pdf to Ian


** 12/04/2016 **

Ian was surprised that mixtures better in all runs.

He suggested that The Curtis eg where sole better might be strange and not fall within our parameter space.

~ check whether I can recreate some results from the new analysis by putting the params into the UI, or better by calling runcurtis_f2()

e.g. first run in dif_oldnew_newonly
mix : 180
sole : 48
difference : 132

#a previous example
runcurtis_f2( P_1 = 0.01 , P_2 = 0.3 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43, addCombinedStrategy=FALSE )

#putting params of 1st run in
runcurtis_f2( P_1 = 0.0159762192 , P_2 = 0.0076494164 , h.RS1_A0 = 0.660797792 , h.RS2_0B = 0.629114044 , exposure = 0.5582827 , phi.SS1_A0 = 0.9441143 , phi.SS2_0B = 0.9695714 , s.RR1_A0 = 0.1216252 , s.RR2_0B = 0.1720911, addCombinedStrategy=FALSE )

gives warning :
critical point not reached for i1 or i2: cutoff_i1_mix=Inf cutoff_i2_mix=Inf

is this because ?? :
1. I've pasted wrong
no
2. an indexing issue means that inputs and results in the table are misaligned

3. what i'm running in these new runs is different from whats done in the plot
don't think so
4. extra defaults no passed to the plot are different from those used in sensi runs
***aha! could be maleExposureProp and correctMixDeployProp


Looking at the plot sole use does look like it takes about 48 as in the results, but for I2 in the mix looks like it should take ~600 rather than 180 as suggested in the results.


New corrected version :

runcurtis_f2( P_1 = 0.0159762192 , P_2 = 0.0076494164 , h.RS1_A0 = 0.660797792 , h.RS2_0B = 0.629114044 , exposure = 0.5582827 , phi.SS1_A0 = 0.9441143 , phi.SS2_0B = 0.9695714 , s.RR1_A0 = 0.1216252 , s.RR2_0B = 0.1720911, addCombinedStrategy=FALSE, maleExposureProp = 0.90820779, correctMixDeployProp = 0.6008410 )

YES this does now match. Hurrah!
(also shows that maleExposureProp and/or correctMixDeployProp in isolation can have a big effect that I could explore in the mechanisms paper).

Should I do for another result too, just to confirm ?


Hi Ian,
I would like to see if I can recreate Curtis Table 1, (vi).

It makes my head hurt trying to work out what our inputs should be based on his table.

Could you help me out by filling in the below * ? (or is it not possible and I'm missing something?).
# start frequencies of each resistance allele
P_1 = *, 
P_2 = * , 
# dominances
h.RS1_A0 =  *, 
h.RS2_0B =  *, 
# exposures
exposure =  *,
# effectivenesses
phi.SS1_A0 =  *, 
phi.SS2_0B =  *, 
# selection coefficients
s.RR1_A0 =  *, 
s.RR2_0B =  *  


Here are the equivalent values used to recreate his figure 2.
# start frequencies of each resistance allele
P_1 = 0.01 , 
P_2 = 0.01 , 
# dominances
h.RS1_A0 = 0.17 , 
h.RS2_0B = 0.0016 , 
# exposures
exposure = 0.9 ,
# effectivenesses
phi.SS1_A0 = 0.73 , 
phi.SS2_0B = 1 , 
# selection coefficients
s.RR1_A0 = 0.23 , 
s.RR2_0B = 0.43 

cheers,
Andy

Back from Ian :

P_1 = 0.01,
P_2 = 0.001,
# dominances
h.RS1_A0 =  1,
h.RS2_0B =  1,
# exposures
exposure =  0.9,
# effectivenesses
phi.SS1_A0 =  1,
phi.SS2_0B =  1,
# selection coefficients
s.RR1_A0 =  1,
s.RR2_0B =  1 )

runcurtis_f2( P_1 = 0.01,
P_2 = 0.001,
# dominances
h.RS1_A0 =  1,
h.RS2_0B =  1,
# exposures
exposure =  0.9,
# effectivenesses
phi.SS1_A0 =  1,
phi.SS2_0B =  1,
# selection coefficients
s.RR1_A0 =  1,
s.RR2_0B =  1,
addCombinedStrategy=FALSE )
 
This seems to generate a very rapid rise in resistance.

and generrates this error in the graph :
Error in xy.coords(x, y) : 'x' and 'y' lengths differ
addCombinedStrategy=FALSE fixes the error

i think I should switch i1 & 2 frequencies

runcurtis_f2( P_1 = 0.001,
P_2 = 0.01,
# dominances
h.RS1_A0 =  1,
h.RS2_0B =  1,
# exposures
exposure =  0.9,
# effectivenesses
phi.SS1_A0 =  1,
phi.SS2_0B =  1,
# selection coefficients
s.RR1_A0 =  1,
s.RR2_0B =  1,
addCombinedStrategy=FALSE,
addStrategyLabels = FALSE )
 
 
This seems to show that resistance to the new insecticide (red) rises faster when alone (dotted) than when in combination (solid) with an old insecticide at a higher frequency. This is contrary to what Curtis said.

Also note that resistance arises much faster than in our sensitivity runs. I think  this is because the selection coefficients (1) are much higher than the range we use (0.1-0.45).


** 15/4/16 **

from Ian :
I have re-read Curtis again and its that paragraph on page 261 that is causing all the problems i.e.
 
“Example vi shows the, at first unexpected……”
 
I guess the key phrase  is what he means by “more rapid”…I assume he means more rapid than using it on its own. But that doesn’t tally with your figure. His example is only a 1-genertion argument so I should be able to do some simple arithmetic. *Can you tell me what the allele frequencies were at generation 2 of your plot* and I’ll check against my arithmetic.


to get allele freqs at gen 2 :


tst$results[[2]][1:8,]
 
** 18/4/16 **

email from Ian
Morning Andy,
 
here is my logic, honed by long walks along the beach
 
(1)    According to your analysis of a subset of the sensitivity analysis i.e. where the “new” insecticide is present at lower frequency…….it always lasts longer under a mixture (presumably this time is defined as that taken for the resistance allele to reach 50%)

 
(2)    But is some of the runs in the sensitivity analysis, sequential use is favoured as it maximises the time for both resistance alleles to reach 50%. Given the result in (1) this can only occur if , in some runs, the “old” insecticide with higher frequency of resistance spreads faster under a mixture.

 
Does this make sense? Can you check if it is true?
 
Ian


~ in runs where resistance arises slowest for sequential does resistance to the the old (higher freq) insecticide arise faster in the presence of the new insecticide than without it ?

~ look at an example of a run where sequential is 'best'

Hi Ian,
Good question. I'm somewhat jet-lagged and a little hungover in a seminar room In Seattle so not thinking at my best but I'll try a quick shot which may be wrong.

The example below is the first one I could find where resistance arises slower for sequential. (it happens to be where the starting frequencies are equal).

In this case it seems that sequential is slower because resistance to insecticide2 (blue) only starts increasing when resistance to the first insecticide has been reached. The curve itself is steeper in the absence of the other insecticide (compare blue dashed to solid) but because it starts later it gets to the threshold later.

I'll try to look for another example.

Andy

#sequential best
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.63 , phi.SS2_0B = 0.85 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )
#relative to curtis fig2
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

#sequential best. effectivenesses bothg reduced by 0.2 from curtis fig2
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.53 , phi.SS2_0B = 0.8 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43, addCombinedStrategy=FALSE )


~ looking at my killer plot gave me a clue how to set the parameter ranges for when sequential should be better. Low effectiveness and high exposure.

#trying to make sequential better, high exposure, low phi

runcurtis_f2( P_1 = 0.05 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 1 , phi.SS1_A0 = 0.1 , phi.SS2_0B = 0.1 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

FRom Ian web version gives an error with certain param combinations

runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 1 , h.RS2_0B = 1 , exposure = 0.9 , phi.SS1_A0 = 1 , phi.SS2_0B = 1 , s.RR1_A0 = 1 , s.RR2_0B = 1, addCombinedStrategy=TRUE )
Error in xy.coords(x, y) : 'x' and 'y' lengths differ

no error when I set addCombinedStrategy=FALSE, but it doesn't plot a curve for i1 mixture ?? or maybe that i1 mix line is behind i2 mix line. Setting a transparency could solve that.

runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 1 , h.RS2_0B = 1 , exposure = 0.9 , phi.SS1_A0 = 1 , phi.SS2_0B = 1 , s.RR1_A0 = 1 , s.RR2_0B = 1, addCombinedStrategy=FALSE )

where is the error ?

in plotcurtis_f2_generic.r

lines( gens, comb, col="red", lty=3 )

because comb contains nothing.





** 27/4/16 **

email from Ian :
I agree entirely with your analysis of why sequential sometimes lasts longer than mixtures (see below in case you have forgotten!). I updated the ms to reflect this and tried to get a coherent discussion of the Curtis results. See new text on pages 16 (paragraph starting “We next investigated the statement by Curtis  “) and on page 19 (paragraph starting “The secondary objective was to apply”).
 
As you will see I put my head on the block (page 16 when I said)
“Finally, for future use, we analysed the other sub-set of the sensitivity data i.e. when the first insecticide to be deployed in the sequence was the one with the higher frequency of resistance and determined whether resistance to that insecticide spread faster when it was deployed on its own or as part of a mixture. Again, in all cases, resistance spread faster when the insecticide was deployed on its own. In summary then, resistance to an insecticide always spreads faster when that insecticide is deployed alone rather than as part of a mixture, irrespective of the relative frequencies of resistance to the two insecticides.“

Can you check if that is true.
 
Secondly, a minor point, but it may become important when determining number of generation for resistance to reach 50%... do you number the initial generation zero so that subsequent generation numbers reflect the elapsed number of generation of selection? e.g. Figure 10  looks like the initial generation is #1. Fig 12 looks like its starts from 0 but the scale is small
 
I had immense problem understanding Table 3 which came from a very early version. I have checked back in my archives but have no previous instnaces of it. It makes no sense to me to and it is not compatible with the dynamics shown on Figure 10 (see note on ms). For example I cannot understand why time to resistance under a mix is not affected by the starting frequency at locus A. If you cannot understand it either, can you re-derive the results and correct it. Thanks.
 
Table 4 looks sensible to me but if you are rechecking table 3, perhaps you could check a couple of values in table 4 just to make sure nothing unexpected has happened.
 
Like I said, happy to Skype if you want to discuss further. Once we get this bit finalised, we can circulate to the Curtis-fans. Like I explain on page 19, I think they may have misinterpreted the comparison he was trying to make.

~~ I checked tables 3 & 4, they were Beths so I sent an email to Ian asking him to ask her 


** 28/4/16 **

https://www.bayer.co.za/en/bayer-develops-the-first-two-way-insecticide-mixture-for-indoor-residual-spraying.php

Consistent with its mission of delivering ‘Science For A Better Life’, Bayer has recently submitted a dossier to the World Health Organization Pesticide Evaluation Scheme (WHOPES) for the evaluation of a new two-way insecticide mixture which includes a new mode-of-action for indoor residual spraying (IRS) against disease vectors. Named Fludora™ Fusion, this first IRS based on two active ingredients is intended to provide an effective solution to help African disease control programs address the challenge of insecticide resistance in malaria-transmitting mosquitoes. Field testing of the product has shown excellent results against many different kinds of resistant mosquitoes and strong performance across a wide range of surfaces. Bayer foresees the WHOPES evaluation and testing process to take about 2 years and anticipates market availability of the product by the end of 2017.
 
This has some good stuff about IRM strategies : 
http://www.slideshare.net/srinivasnaik52643/insecticide-resistance-management-strategy 
 
Curtis et al 1993, potential IRM strategies :
~ moderation : preserve susceptible genes by limiting selection pressure
~ saturation : high dose so that heterozugous resistants are killed
~ multiple attack : independently acting pressures nneither of which is strong enough to lead to resistance
 
Mani GS. Evolution of resistance in the presence of two insecticides. Genetics 1985;109(4):761–83.
 
Abstract
A two-locus model is used to analyze the effectiveness of a mixture of insecticides in delaying resistance, compared to the use of the insecticides singly. The effects of factors such as recombination, effective dominance, initial value of allele frequencies and initial value of linkage disequilibrium are considered. It is shown that the use of mixtures is always more effective in delaying the onset of resistance, often by many orders of magnitude. It is shown that there exists a threshold value of recombination fraction, above which the evolution of resistance is extremely slow. Resistance evolves very rapidly for values of recombination fraction below the threshold. Finally, the relevance of these results on resistance management is discussed.

~ is mixture always better than sole ?
Yes see this in paper1_results_figs.Rmd :
  mix_minus_soleI1 <- resistPointsMixI1_T["gen_cP0.5"] - resistPointsI1_T["gen_cP0.5"]
  which( mix_minus_soleI1 < 0 ) 
  #0
  mix_minus_soleI2 <- resistPointsMixI2_T["gen_cP0.5"] - resistPointsI2_T["gen_cP0.5"]
  which( mix_minus_soleI2 < 0 ) 
  


 
### TO DO  

~ set up a new UI allowing comparison of 2 model runs side by side (like my curtis one but be able to modify params for both plots)


~ confirm whether Ians statement is true on all sensi runs : does resistance always spread faster when an insecticide is alone to when it is in a mixture ?
"resistance to an insecticide always spreads faster when that insecticide is deployed alone rather than as part of a mixture, irrespective of the relative frequencies of resistance to the two insecticides.""

~ do I start figs at gen 0 or 1 ? First generation is 1.


~ read through Ians comments on new version of paper




~ create a user interface to represent Killeen graphical models


~ how to recreate Curtis Table 1, (vi)'the use of a mixture when the initial gene frequencies are unequal leads to a more rapid increase in the frequency of the rarer of the genes'.





~ remember this is still for all runs (i.e. not just the ones where start freq of the old is > 10 * the new) 

~ **check this from above**
~ I probably need to get straighter in my head how dominance works.
Does dominance effect the relative performance of the 2 insecticides where mosquitos are exposed to both. i.e. Does high dominance of the new allele mean that a mosquito is more likely to survive the presence of the old insecticide ?

To understand the mechanisms more I could look at the effects of the inputs on each strategy on its own.



# quick thought check, sequential is calculated by adding together the times for I1 & I2 in isolation, are we sure that is OK ? e.g. do we expect it to take same time for resistance to 2nd insecticide to reach threshold irrespective of the level of resistance to I1 even when I1 is not present.




1. tidy these notes


I've done below for the mix-both plot, repeat it for mix-adaptive.
Ian suggested adding in the maleProp effect to the x axis to see if that improves the split.
To get the exposure of both M&F : x=exposure+exposure*maleExposureProp





~ would be a good learning tool to have a UI similar to my current curtis fig2 one, but starting with the params equal for each insecticide

Curtis contentious statement p261 :
"... the use of a mixture where the initial gene frequencies are unequal leads to a more rapid increase in the rarer of the genes."
Not supported by our results. We need to nail which results to use to show this.



bumped into colin bannister on the train to london. he said he really ejoyed my talk. he said he would talk to me sometime about keeping the work going in a consultancy, difficult to do. He was on his way to london for a shellfish assessment.

In fig x1 sequential & adaptive mixture look very similar indeed ???

Could I write a very short paper describing the shiny app that allows users to take the curtis figure & modify the code ? Where might it go ? It wouldn't need any of the sensitivity analysis etc. ? It could cite the paper with the snesitivity analysis.

Note that in most functions (e.g. runcurtis_f2) there is a single exposure param for both insecticides. I could relatively easily allow different exposure for each insecticide. It is probably already possible by passing array:a directly.


Tables 3 & 4 in the manuscript- can we do these as a figure ?
These address Curtis statement

Do we want to go with a 50% threshold across all analyses to keep clearer ?



An influential assertion is that mixtures may be counter-productive (made in the abstract and  repeated on page 261) where Curtis refers to his Example (vi) on his Table 1. Unfortunately, Curtis Figure 1 only shows the spread of resistance using a mixture and does not quantify the extent to which it is inferior to sequential use. However, the calibration used to replicate Curtis Figure 1 allows us to extend the one-generation argument presented in Curtis Table 1(vi) over the numerous generations required to simulate the spread of resistance (see later discussion of our Table 3).


reread Curtis to check the below ...

From Ian
~ I think the Curtis result (the one that said sequential was better) assumed effectiveness of both =1 and that exposure =0.9. I think that fits right at the top RH corner of the figure.


~ check with Ian what the best resistance threshold to focus on is. In Curtis Fig2 it is 50%. I have been using the 25% in my figs (but I could change). WHO thresholds are :

Potential for me to look at in a 2nd paper :
Differential exposure of males & females. (I have shown exposure is important so this could have a big effect).


~ I just had a thought. Is the effectiveness effect additive ? i.e. do 2 insecticides of effectiveness 0.7 have the same effect e.g. as one of 1 and the other of 1.4 ? 

~ maybe communicate my concern to Ian that figures I've spent quite a long time on just go into supplementary material where few will see them. Can we think of a way of getting these into a paper themselves ? e.g. if this first paper is targeted at policy people can we get a 2nd paper that is targetted at biologists and explores mechanisms ?






Long Names (in text where space not an issue)
Mixture (either insecticide)   
Mixture (both insecticides)
Adaptive mixture (both insecticides)
Sequential (both insecticides)

Medium Names (plot labels where medium space)
mix either              
mix both             
mix adaptive
sequential           

Short Names (plot labels where v little space)
mix1
mix2
adapt
seq



effectI2 from 1 to 0.85, & resistance arises faster for mix1 than seq
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 0.85 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )

effectI2 from 1 to 0.5, & resistance arises faster for mix2 than seq
runcurtis_f2( P_1 = 0.01 , P_2 = 0.01 , h.RS1_A0 = 0.17 , h.RS2_0B = 0.0016 , exposure = 0.9 , phi.SS1_A0 = 0.73 , phi.SS2_0B = 0.5 , s.RR1_A0 = 0.23 , s.RR2_0B = 0.43 )


~ check code in findResistancePointsMixResponsive()


~ can I get plots to show just a subset of points ?

~ produce a graphich of the 3 diff mixture scenarios & sequential
~~ may be able to use the seriousGame plot method
~~ I could even show resistance curves next to it ?


# would it be useful for Ian for me to create an interactive shiny app to view the 1000 sim data. E.g. could just have the ggplot code with options to set the x value & colour by different inputs.
# maybe try faceting by parameter ranges ?


1. create a vignette that shows how best to run current model
1. fix the testthat tests to avoid new input addition
1. create functions fitnessLocus(), fitnessNiche(), fitnessIndiv()

Remember.
It is a 2 locus model. One locus for each insecticide. At each locus there are 2 alleles. This gives a total of 9 genotypes if the position of the alleles is ignored. 


### current state of play 23/3/16
1. sensiAnPaper1All.Rmd : code to run whole sensitivity analysis & produce plots 
1. sensiAnPaperPart.r   : sets up param values and runs for one treatment (e.g. mixture, insecticide1)
1. runModel2.r          : runs model for the passed scenarios
1. paper1_results_figs_slimmed50.Rmd : final figures for paper


Files :
listOut* I1,I2, Mix : raw results for insecticide1,2 & mixture
                    : results are in $results[[scenarioNum]]
                    : e.g. listOutI1$results[[1]] has columns
                    : colnames(listOutI1$results[[1]])
                    : "Gen" "m.R1" "m.R2"  "m.LD" "f.R1"  "f.R2" "f.LD" "M"  "F"  "dprime" "r2"
treeInput : a matrix with scenarios in columns, inputs & outputs in rows  

Good simple test
```
input <- setInputOneScenario(max_gen=5)
tst2 <- runModel2(input)
tst <- runModel(input)
```

1. submit stackoverflow question about createArray2(), including about n=3 in eval.parent
1. what to call createArray2() ? maybe namedArray, or arrayNamed() ? could even put back in the option to pass the fill value ? submit to stackoverflow 
1. maybe add option in setInputSensiScenarios to save input object as a csv or rda.
1. add some defensive checks, e.g. for sensible input variables


Want to do:
1. rename a to exposure
1. rename W to ?fitness : fitnessLocus, fitnessNiche and fitnessIndiv


### variable naming conventions
**SS1, RS1, RR1** homozygous susceptible, heterozygous or homozygous resistant genotypes at each locus.   
With 2 loci there are ten possible genotypes (including cis & trans forms of the double heterozygous).  

** A, B, a, b, -** 2 insecticides at 3 levels; low(ab), high(AB), absent(--) resulting in 9 niches.  
Locus1 relates to resistance for insecticide A, and locus2 for B. Locus2 does not effect exposure to A.  

```
colnames(fitness) <- c( "-,-", "a,-", "A,-", "b,-", "B,-", "a,b", "A,B", "A,b", "a,B" )
rownames(fitness) <- c( "SS1SS2", "SS2RS2", "SS1RR2", 
                        "RS1SS2", "RS1RS2_cis", "RS1RS2_trans", "RS1RR2",
                        "RR1SS2", "RR1RS2", "RR1RR2")
```
#### inputs
** a ** exposure to insecticide
** W.SS1_00 & W.SS2_00** fitness of SS in no insecticide
** s ** selection coefficient
** phi ** fitness of one locus (baseline)
** h ** dominance coefficient
** z ** fitness cost of resistance allele in no insecticide
** niche_ ** insecticide niche toggle

#### calculated
** W ** fitnesses by niche and sex
** f ** genotype frequency
** fs ** genotype frequency following selection
** G ** gametes

Different W fitness variables
`W.RR1_00 : [locus, niche1, niche2]` single locus (Wloci)
`W.RR1SS2_0b :   [locus1, locus2, niche1, niche2]` niche (Wniche)
`W.m.SS1SS2 : [sex, locus1, locus2]` (Windiv)

```
a.m_a0 ... *18 2*3*3  to: a[sex, niche1, niche2]
W.m.SS1SS2 ... *18 2*3*3 to:  Windiv[sex, locus1, locus2]
W.RR1SS2_0b ... *81 3*3*3*3 to: Wniche[locus1, locus2, niche1, niche2] 
W.RR1_00 etc. with Wloci[loci, niche1, niche2]

because of cis/trans:
f.m.SS1SS2 ... *19 2*3*3+1 to: f[sex, genotype]
```

### outputs ###
From runModel2()
A list of 3 lists of one or more scenarios: 
1. results : by generation freq of R allele at each loci in each sex plus linkage stuff
    columns include Gen, m.R1, m.R2, f.R1, f.R2
2. genotype : frequencies of each of the ten genotypes, per generation
3. fitness : fitness scores of each genotype/niche combination (table 4. of Main Document)
4. input
e.g. 
listOut$results[1] gives a results matrix for the first scenario
listOutMix$input gives the inputs for all scenarios
listOutMix$input[c('P_1','P_2'),] gives selected inputs

# current best UIs
https://andysouth.shinyapps.io/shinyMixSeqCompare1
https://andysouth.shinyapps.io/shinyFig2Curtis




